{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dash_bootstrap_components as dbc\n",
    "import dash_daq as daq\n",
    "import heapq\n",
    "import numpy as np\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import polars as pl\n",
    "import pyarrow\n",
    "\n",
    "from dash import Dash, dcc, html, dash_table, Input, Output, State, ctx\n",
    "from jupyter_dash import JupyterDash\n",
    "from gtfparse import read_gtf # gtfparse==1.3.0\n",
    "\n",
    "\n",
    "app = JupyterDash(__name__, \n",
    "           external_stylesheets = [dbc.themes.LUX, dbc.icons.BOOTSTRAP],\n",
    "           title = \"GWAS SVatalog\")\n",
    "\n",
    "app._favicon = \"assets/favicon.ico\"\n",
    "\n",
    "server = app.server\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (471_113, 34)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>DATE ADDED TO CATALOG</th><th>PUBMEDID</th><th>FIRST AUTHOR</th><th>DATE</th><th>JOURNAL</th><th>LINK</th><th>STUDY</th><th>DISEASE/TRAIT</th><th>INITIAL SAMPLE SIZE</th><th>REPLICATION SAMPLE SIZE</th><th>REGION</th><th>CHR_ID</th><th>CHR_POS</th><th>REPORTED GENE(S)</th><th>MAPPED_GENE</th><th>UPSTREAM_GENE_ID</th><th>DOWNSTREAM_GENE_ID</th><th>SNP_GENE_IDS</th><th>UPSTREAM_GENE_DISTANCE</th><th>DOWNSTREAM_GENE_DISTANCE</th><th>STRONGEST SNP-RISK ALLELE</th><th>SNPS</th><th>MERGED</th><th>SNP_ID_CURRENT</th><th>CONTEXT</th><th>INTERGENIC</th><th>RISK ALLELE FREQUENCY</th><th>P-VALUE</th><th>PVALUE_MLOG</th><th>P-VALUE (TEXT)</th><th>OR or BETA</th><th>95% CI (TEXT)</th><th>PLATFORM [SNPS PASSING QC]</th><th>CNV</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;2q24.2&quot;</td><td>&quot;2&quot;</td><td>160772472</td><td>&quot;NR&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs146071273-A&quot;</td><td>&quot;rs146071273&quot;</td><td>0</td><td>&quot;146071273&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.0867&quot;</td><td>0.000007</td><td>5.154902</td><td>&quot;(EA)&quot;</td><td>&quot;0.052&quot;</td><td>&quot;[0.029-0.075] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;3p21.31&quot;</td><td>&quot;3&quot;</td><td>47364189</td><td>&quot;NR&quot;</td><td>&quot;KLHL18 - PTPN2…</td><td>&quot;ENSG0000011464…</td><td>&quot;ENSG0000007620…</td><td>null</td><td>17373</td><td>16822</td><td>&quot;rs146331761-A&quot;</td><td>&quot;rs146331761&quot;</td><td>0</td><td>&quot;146331761&quot;</td><td>&quot;regulatory_reg…</td><td>1</td><td>&quot;0.0367&quot;</td><td>0.0000005</td><td>6.30103</td><td>&quot;(EA)&quot;</td><td>&quot;0.0875&quot;</td><td>&quot;[0.053-0.122] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;5q14.2&quot;</td><td>&quot;5&quot;</td><td>82217129</td><td>&quot;NR&quot;</td><td>&quot;ATG10&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000015234…</td><td>null</td><td>null</td><td>&quot;rs146817970-T&quot;</td><td>&quot;rs146817970&quot;</td><td>0</td><td>&quot;146817970&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.2459&quot;</td><td>4.0000e-13</td><td>12.39794</td><td>&quot;(EA)&quot;</td><td>&quot;0.0532&quot;</td><td>&quot;[0.039-0.068] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;14q21.1&quot;</td><td>&quot;14&quot;</td><td>38968140</td><td>&quot;NR&quot;</td><td>&quot;LINC00639 - Y_…</td><td>&quot;ENSG0000025907…</td><td>&quot;ENSG0000020723…</td><td>null</td><td>19867</td><td>19141</td><td>&quot;rs147032157-A&quot;</td><td>&quot;rs147032157&quot;</td><td>0</td><td>&quot;147032157&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.1018&quot;</td><td>0.000004</td><td>5.39794</td><td>&quot;(EA)&quot;</td><td>&quot;0.0483&quot;</td><td>&quot;[0.028-0.069] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;13q33.3&quot;</td><td>&quot;13&quot;</td><td>107031171</td><td>&quot;NR&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs148731571-GC…</td><td>&quot;rs148731571&quot;</td><td>0</td><td>&quot;148731571&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.0639&quot;</td><td>0.000006</td><td>5.221849</td><td>&quot;(EA)&quot;</td><td>&quot;0.0712&quot;</td><td>&quot;[0.04-0.102] u…</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;1p11.2&quot;</td><td>&quot;1&quot;</td><td>121695219</td><td>&quot;NR&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs149253773-T&quot;</td><td>&quot;rs149253773&quot;</td><td>1</td><td>&quot;78533661&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.5359&quot;</td><td>6.0000e-40</td><td>39.221849</td><td>&quot;(EA)&quot;</td><td>&quot;0.0944&quot;</td><td>&quot;[0.08-0.108] u…</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;22q12.1&quot;</td><td>&quot;22&quot;</td><td>28532870</td><td>&quot;NR&quot;</td><td>&quot;TTC28&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000010015…</td><td>null</td><td>null</td><td>&quot;rs149936356-C&quot;</td><td>&quot;rs149936356&quot;</td><td>0</td><td>&quot;149936356&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.0246&quot;</td><td>1.0000e-18</td><td>18.0</td><td>&quot;(EA)&quot;</td><td>&quot;0.176&quot;</td><td>&quot;[0.14-0.22] un…</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;17q21.2&quot;</td><td>&quot;17&quot;</td><td>41094871</td><td>&quot;NR&quot;</td><td>&quot;KRTAP4-7 - KRT…</td><td>&quot;ENSG0000024087…</td><td>&quot;ENSG0000026196…</td><td>null</td><td>9730</td><td>2110</td><td>&quot;rs150537328-C&quot;</td><td>&quot;rs150537328&quot;</td><td>0</td><td>&quot;150537328&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.0647&quot;</td><td>0.000002</td><td>5.69897</td><td>&quot;(EA)&quot;</td><td>&quot;0.0642&quot;</td><td>&quot;[0.038-0.091] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;14q32.12&quot;</td><td>&quot;14&quot;</td><td>91422581</td><td>&quot;NR&quot;</td><td>&quot;CCDC88C - PPP4…</td><td>&quot;ENSG0000001513…</td><td>&quot;ENSG0000010079…</td><td>null</td><td>4737</td><td>34927</td><td>&quot;rs150658557-TA…</td><td>&quot;rs150658557&quot;</td><td>0</td><td>&quot;150658557&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.2029&quot;</td><td>2.0000e-13</td><td>12.69897</td><td>&quot;(EA)&quot;</td><td>&quot;0.0605&quot;</td><td>&quot;[0.044-0.077] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;15q22.33&quot;</td><td>&quot;15&quot;</td><td>67115961</td><td>&quot;NR&quot;</td><td>&quot;SMAD3&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000016694…</td><td>null</td><td>null</td><td>&quot;rs151090251-GG…</td><td>&quot;rs151090251&quot;</td><td>0</td><td>&quot;151090251&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.0524&quot;</td><td>0.000002</td><td>5.69897</td><td>&quot;(EA)&quot;</td><td>&quot;0.0719&quot;</td><td>&quot;[0.042-0.101] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;4q28.3&quot;</td><td>&quot;4&quot;</td><td>132819226</td><td>&quot;NR&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs151170584-T&quot;</td><td>&quot;rs151170584&quot;</td><td>0</td><td>&quot;151170584&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.0362&quot;</td><td>0.000009</td><td>5.045757</td><td>&quot;(EA)&quot;</td><td>&quot;0.0989&quot;</td><td>&quot;[0.055-0.143] …</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2017-11-30&quot;</td><td>&quot;29059683&quot;</td><td>&quot;Michailidou K&quot;</td><td>&quot;2017-10-23&quot;</td><td>&quot;Nature&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;Association an…</td><td>&quot;Breast cancer&quot;</td><td>&quot;76,192 Europea…</td><td>&quot;46,785 Europea…</td><td>&quot;17q21.31&quot;</td><td>&quot;17&quot;</td><td>42807276</td><td>&quot;NR&quot;</td><td>&quot;CNTD1&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000017656…</td><td>null</td><td>null</td><td>&quot;rs151329939-A&quot;</td><td>&quot;rs151329939&quot;</td><td>0</td><td>&quot;151329939&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.01&quot;</td><td>5.0000e-9</td><td>8.30103</td><td>&quot;(EA)&quot;</td><td>&quot;0.1879&quot;</td><td>&quot;[0.12-0.25] un…</td><td>&quot;Illumina [~ 11…</td><td>&quot;N&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>2.0000e-12</td><td>11.69897</td><td>&quot;(362 ms)&quot;</td><td>&quot;0.0395349&quot;</td><td>&quot;[0.028-0.051] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>2.0000e-11</td><td>10.69897</td><td>&quot;(364 ms)&quot;</td><td>&quot;0.0380775&quot;</td><td>&quot;[0.027-0.049] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>8.0000e-10</td><td>9.09691</td><td>&quot;(366 ms)&quot;</td><td>&quot;0.0346457&quot;</td><td>&quot;[0.024-0.046] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>1.0000e-8</td><td>8.0</td><td>&quot;(368 ms)&quot;</td><td>&quot;0.0322342&quot;</td><td>&quot;[0.021-0.043] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>8.0000e-9</td><td>8.09691</td><td>&quot;(370 ms)&quot;</td><td>&quot;0.0325839&quot;</td><td>&quot;[0.022-0.044] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>2.0000e-8</td><td>7.69897</td><td>&quot;(372 ms)&quot;</td><td>&quot;0.0315134&quot;</td><td>&quot;[0.02-0.043] u…</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;6q22.31&quot;</td><td>&quot;6&quot;</td><td>118346359</td><td>&quot;SLC35F1&quot;</td><td>&quot;SLC35F1 - CEP8…</td><td>&quot;ENSG0000019637…</td><td>&quot;ENSG0000011186…</td><td>null</td><td>28683</td><td>114413</td><td>&quot;rs11153730-T&quot;</td><td>&quot;rs11153730&quot;</td><td>0</td><td>&quot;11153730&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.520014&quot;</td><td>4.0000e-54</td><td>53.39794</td><td>&quot;(-16 ms)&quot;</td><td>&quot;0.0849022&quot;</td><td>&quot;[0.074-0.096] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;5q14.1&quot;</td><td>&quot;5&quot;</td><td>77842993</td><td>&quot;TBCA&quot;</td><td>&quot;TBCA&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000017153…</td><td>null</td><td>null</td><td>&quot;rs111778721-C&quot;</td><td>&quot;rs111778721&quot;</td><td>0</td><td>&quot;111778721&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.963504&quot;</td><td>3.0000e-8</td><td>7.522879</td><td>&quot;(6 ms)&quot;</td><td>&quot;0.0830913&quot;</td><td>&quot;[0.053-0.113] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;5q14.1&quot;</td><td>&quot;5&quot;</td><td>77842993</td><td>&quot;TBCA&quot;</td><td>&quot;TBCA&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000017153…</td><td>null</td><td>null</td><td>&quot;rs111778721-C&quot;</td><td>&quot;rs111778721&quot;</td><td>0</td><td>&quot;111778721&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.963504&quot;</td><td>5.0000e-8</td><td>7.30103</td><td>&quot;(10 ms)&quot;</td><td>&quot;0.0819697&quot;</td><td>&quot;[0.052-0.112] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;5q14.1&quot;</td><td>&quot;5&quot;</td><td>77842993</td><td>&quot;TBCA&quot;</td><td>&quot;TBCA&quot;</td><td>null</td><td>null</td><td>&quot;ENSG0000017153…</td><td>null</td><td>null</td><td>&quot;rs111778721-C&quot;</td><td>&quot;rs111778721&quot;</td><td>0</td><td>&quot;111778721&quot;</td><td>&quot;intron_variant…</td><td>0</td><td>&quot;0.963504&quot;</td><td>1.0000e-8</td><td>8.0</td><td>&quot;(8 ms)&quot;</td><td>&quot;0.0853889&quot;</td><td>&quot;[0.055-0.116] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;5p12&quot;</td><td>&quot;5&quot;</td><td>45014943</td><td>&quot;MRPS30&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs111869282-C&quot;</td><td>&quot;rs111869282&quot;</td><td>0</td><td>&quot;111869282&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.791595&quot;</td><td>7.0000e-9</td><td>8.154902</td><td>&quot;(322 ms)&quot;</td><td>&quot;0.0384843&quot;</td><td>&quot;[0.025-0.052] …</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr><tr><td>&quot;2020-11-30&quot;</td><td>&quot;32916098&quot;</td><td>&quot;Verweij N&quot;</td><td>&quot;2020-09-02&quot;</td><td>&quot;Cell Syst&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>&quot;The Genetic Ma…</td><td>&quot;Electrocardiog…</td><td>&quot;63,706 Europea…</td><td>&quot;NA&quot;</td><td>&quot;5p12&quot;</td><td>&quot;5&quot;</td><td>45014943</td><td>&quot;MRPS30&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;rs111869282-C&quot;</td><td>&quot;rs111869282&quot;</td><td>0</td><td>&quot;111869282&quot;</td><td>&quot;intergenic_var…</td><td>1</td><td>&quot;0.791595&quot;</td><td>3.0000e-8</td><td>7.522879</td><td>&quot;(288 ms)&quot;</td><td>&quot;0.0367307&quot;</td><td>&quot;[0.023-0.05] u…</td><td>&quot;Affymetrix [10…</td><td>&quot;N&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (471_113, 34)\n",
       "┌────────────┬──────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬─────┐\n",
       "│ DATE ADDED ┆ PUBMEDID ┆ FIRST      ┆ DATE       ┆ … ┆ OR or BETA ┆ 95% CI     ┆ PLATFORM   ┆ CNV │\n",
       "│ TO CATALOG ┆ ---      ┆ AUTHOR     ┆ ---        ┆   ┆ ---        ┆ (TEXT)     ┆ [SNPS      ┆ --- │\n",
       "│ ---        ┆ str      ┆ ---        ┆ str        ┆   ┆ str        ┆ ---        ┆ PASSING    ┆ str │\n",
       "│ str        ┆          ┆ str        ┆            ┆   ┆            ┆ str        ┆ QC]        ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ ---        ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ str        ┆     │\n",
       "╞════════════╪══════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪═════╡\n",
       "│ 2017-11-30 ┆ 29059683 ┆ Michailido ┆ 2017-10-23 ┆ … ┆ 0.052      ┆ [0.029-0.0 ┆ Illumina   ┆ N   │\n",
       "│            ┆          ┆ u K        ┆            ┆   ┆            ┆ 75] unit   ┆ [~         ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ decrease   ┆ 11800000]  ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ (imputed)  ┆     │\n",
       "│ 2017-11-30 ┆ 29059683 ┆ Michailido ┆ 2017-10-23 ┆ … ┆ 0.0875     ┆ [0.053-0.1 ┆ Illumina   ┆ N   │\n",
       "│            ┆          ┆ u K        ┆            ┆   ┆            ┆ 22] unit   ┆ [~         ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ decrease   ┆ 11800000]  ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ (imputed)  ┆     │\n",
       "│ 2017-11-30 ┆ 29059683 ┆ Michailido ┆ 2017-10-23 ┆ … ┆ 0.0532     ┆ [0.039-0.0 ┆ Illumina   ┆ N   │\n",
       "│            ┆          ┆ u K        ┆            ┆   ┆            ┆ 68] unit   ┆ [~         ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ decrease   ┆ 11800000]  ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ (imputed)  ┆     │\n",
       "│ 2017-11-30 ┆ 29059683 ┆ Michailido ┆ 2017-10-23 ┆ … ┆ 0.0483     ┆ [0.028-0.0 ┆ Illumina   ┆ N   │\n",
       "│            ┆          ┆ u K        ┆            ┆   ┆            ┆ 69] unit   ┆ [~         ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ decrease   ┆ 11800000]  ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆            ┆ (imputed)  ┆     │\n",
       "│ …          ┆ …        ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …   │\n",
       "│ 2020-11-30 ┆ 32916098 ┆ Verweij N  ┆ 2020-09-02 ┆ … ┆ 0.0819697  ┆ [0.052-0.1 ┆ Affymetrix ┆ N   │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ 12] unit   ┆ [10000000] ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ increase   ┆ (imputed)  ┆     │\n",
       "│ 2020-11-30 ┆ 32916098 ┆ Verweij N  ┆ 2020-09-02 ┆ … ┆ 0.0853889  ┆ [0.055-0.1 ┆ Affymetrix ┆ N   │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ 16] unit   ┆ [10000000] ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ increase   ┆ (imputed)  ┆     │\n",
       "│ 2020-11-30 ┆ 32916098 ┆ Verweij N  ┆ 2020-09-02 ┆ … ┆ 0.0384843  ┆ [0.025-0.0 ┆ Affymetrix ┆ N   │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ 52] unit   ┆ [10000000] ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ increase   ┆ (imputed)  ┆     │\n",
       "│ 2020-11-30 ┆ 32916098 ┆ Verweij N  ┆ 2020-09-02 ┆ … ┆ 0.0367307  ┆ [0.023-0.0 ┆ Affymetrix ┆ N   │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ 5] unit    ┆ [10000000] ┆     │\n",
       "│            ┆          ┆            ┆            ┆   ┆            ┆ increase   ┆ (imputed)  ┆     │\n",
       "└────────────┴──────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴─────┘"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "DF_GWAS_FULL = pl.read_csv(\"data/gwas_catalog_v1.0-associations_e108.tsv\",\n",
    "                           separator = \"\\t\",\n",
    "                           quote_char = '',\n",
    "                           dtypes = {\"PUBMEDID\": pl.Utf8, \n",
    "                                     \"LINK\": pl.Utf8,\n",
    "                                     \"P-VALUE\": pl.Float64},\n",
    "                           infer_schema_length = 1_000_000)\n",
    "\n",
    "# Remove quotes from all string columns\n",
    "# DF_GWAS_FULL = DF_GWAS_FULL.with_columns([pl.col(column).str.replace(r'^\"|\"$', \"\")\n",
    "#                                           for column in DF_GWAS_FULL.columns])\n",
    "\n",
    "DF_GWAS_FULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (449_339, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SNP_Position</th><th>SNP_Name_GWAS</th><th>Risk_Allele</th><th>Risk_Allele_Frequency</th><th>Phenotype</th><th>Study</th><th>Pubmed_ID</th><th>Link</th><th>P-Value</th></tr><tr><td>str</td><td>i32</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr2&quot;</td><td>160772472</td><td>&quot;rs146071273&quot;</td><td>&quot;A&quot;</td><td>&quot;0.0867&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000007</td></tr><tr><td>&quot;chr3&quot;</td><td>47364189</td><td>&quot;rs146331761&quot;</td><td>&quot;A&quot;</td><td>&quot;0.0367&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.0000005</td></tr><tr><td>&quot;chr5&quot;</td><td>82217129</td><td>&quot;rs146817970&quot;</td><td>&quot;T&quot;</td><td>&quot;0.2459&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>4.0000e-13</td></tr><tr><td>&quot;chr14&quot;</td><td>38968140</td><td>&quot;rs147032157&quot;</td><td>&quot;A&quot;</td><td>&quot;0.1018&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000004</td></tr><tr><td>&quot;chr13&quot;</td><td>107031171</td><td>&quot;rs148731571&quot;</td><td>&quot;GCTTAGATATTAC&quot;</td><td>&quot;0.0639&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000006</td></tr><tr><td>&quot;chr1&quot;</td><td>121695219</td><td>&quot;rs149253773&quot;</td><td>&quot;T&quot;</td><td>&quot;0.5359&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>6.0000e-40</td></tr><tr><td>&quot;chr22&quot;</td><td>28532870</td><td>&quot;rs149936356&quot;</td><td>&quot;C&quot;</td><td>&quot;0.0246&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-18</td></tr><tr><td>&quot;chr17&quot;</td><td>41094871</td><td>&quot;rs150537328&quot;</td><td>&quot;C&quot;</td><td>&quot;0.0647&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000002</td></tr><tr><td>&quot;chr14&quot;</td><td>91422581</td><td>&quot;rs150658557&quot;</td><td>&quot;TA&quot;</td><td>&quot;0.2029&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-13</td></tr><tr><td>&quot;chr15&quot;</td><td>67115961</td><td>&quot;rs151090251&quot;</td><td>&quot;GGACTTCTGTAGAG…</td><td>&quot;0.0524&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000002</td></tr><tr><td>&quot;chr4&quot;</td><td>132819226</td><td>&quot;rs151170584&quot;</td><td>&quot;T&quot;</td><td>&quot;0.0362&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.000009</td></tr><tr><td>&quot;chr17&quot;</td><td>42807276</td><td>&quot;rs151329939&quot;</td><td>&quot;A&quot;</td><td>&quot;0.01&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>5.0000e-9</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-12</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-11</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>8.0000e-10</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-8</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>8.0000e-9</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-8</td></tr><tr><td>&quot;chr6&quot;</td><td>118346359</td><td>&quot;rs11153730&quot;</td><td>&quot;T&quot;</td><td>&quot;0.520014&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>4.0000e-54</td></tr><tr><td>&quot;chr5&quot;</td><td>77842993</td><td>&quot;rs111778721&quot;</td><td>&quot;C&quot;</td><td>&quot;0.963504&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-8</td></tr><tr><td>&quot;chr5&quot;</td><td>77842993</td><td>&quot;rs111778721&quot;</td><td>&quot;C&quot;</td><td>&quot;0.963504&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>5.0000e-8</td></tr><tr><td>&quot;chr5&quot;</td><td>77842993</td><td>&quot;rs111778721&quot;</td><td>&quot;C&quot;</td><td>&quot;0.963504&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-8</td></tr><tr><td>&quot;chr5&quot;</td><td>45014943</td><td>&quot;rs111869282&quot;</td><td>&quot;C&quot;</td><td>&quot;0.791595&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>7.0000e-9</td></tr><tr><td>&quot;chr5&quot;</td><td>45014943</td><td>&quot;rs111869282&quot;</td><td>&quot;C&quot;</td><td>&quot;0.791595&quot;</td><td>&quot;Electrocardiog…</td><td>&quot;The Genetic Ma…</td><td>&quot;32916098&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-8</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (449_339, 10)\n",
       "┌──────────┬────────────┬────────────┬───────────┬───┬───────────┬─────────┬───────────┬───────────┐\n",
       "│ Chromoso ┆ SNP_Positi ┆ SNP_Name_G ┆ Risk_Alle ┆ … ┆ Study     ┆ Pubmed_ ┆ Link      ┆ P-Value   │\n",
       "│ me       ┆ on         ┆ WAS        ┆ le        ┆   ┆ ---       ┆ ID      ┆ ---       ┆ ---       │\n",
       "│ ---      ┆ ---        ┆ ---        ┆ ---       ┆   ┆ str       ┆ ---     ┆ str       ┆ f64       │\n",
       "│ str      ┆ i32        ┆ str        ┆ str       ┆   ┆           ┆ str     ┆           ┆           │\n",
       "╞══════════╪════════════╪════════════╪═══════════╪═══╪═══════════╪═════════╪═══════════╪═══════════╡\n",
       "│ chr2     ┆ 160772472  ┆ rs14607127 ┆ A         ┆ … ┆ Associati ┆ 2905968 ┆ www.ncbi. ┆ 0.000007  │\n",
       "│          ┆            ┆ 3          ┆           ┆   ┆ on        ┆ 3       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ analysis  ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ identifie ┆         ┆ /2905…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ s …       ┆         ┆           ┆           │\n",
       "│ chr3     ┆ 47364189   ┆ rs14633176 ┆ A         ┆ … ┆ Associati ┆ 2905968 ┆ www.ncbi. ┆ 0.0000005 │\n",
       "│          ┆            ┆ 1          ┆           ┆   ┆ on        ┆ 3       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ analysis  ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ identifie ┆         ┆ /2905…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ s …       ┆         ┆           ┆           │\n",
       "│ chr5     ┆ 82217129   ┆ rs14681797 ┆ T         ┆ … ┆ Associati ┆ 2905968 ┆ www.ncbi. ┆ 4.0000e-1 │\n",
       "│          ┆            ┆ 0          ┆           ┆   ┆ on        ┆ 3       ┆ nlm.nih.g ┆ 3         │\n",
       "│          ┆            ┆            ┆           ┆   ┆ analysis  ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ identifie ┆         ┆ /2905…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ s …       ┆         ┆           ┆           │\n",
       "│ chr14    ┆ 38968140   ┆ rs14703215 ┆ A         ┆ … ┆ Associati ┆ 2905968 ┆ www.ncbi. ┆ 0.000004  │\n",
       "│          ┆            ┆ 7          ┆           ┆   ┆ on        ┆ 3       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ analysis  ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ identifie ┆         ┆ /2905…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ s …       ┆         ┆           ┆           │\n",
       "│ …        ┆ …          ┆ …          ┆ …         ┆ … ┆ …         ┆ …       ┆ …         ┆ …         │\n",
       "│ chr5     ┆ 77842993   ┆ rs11177872 ┆ C         ┆ … ┆ The       ┆ 3291609 ┆ www.ncbi. ┆ 5.0000e-8 │\n",
       "│          ┆            ┆ 1          ┆           ┆   ┆ Genetic   ┆ 8       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Makeup of ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ the       ┆         ┆ /3291…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Electr…   ┆         ┆           ┆           │\n",
       "│ chr5     ┆ 77842993   ┆ rs11177872 ┆ C         ┆ … ┆ The       ┆ 3291609 ┆ www.ncbi. ┆ 1.0000e-8 │\n",
       "│          ┆            ┆ 1          ┆           ┆   ┆ Genetic   ┆ 8       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Makeup of ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ the       ┆         ┆ /3291…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Electr…   ┆         ┆           ┆           │\n",
       "│ chr5     ┆ 45014943   ┆ rs11186928 ┆ C         ┆ … ┆ The       ┆ 3291609 ┆ www.ncbi. ┆ 7.0000e-9 │\n",
       "│          ┆            ┆ 2          ┆           ┆   ┆ Genetic   ┆ 8       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Makeup of ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ the       ┆         ┆ /3291…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Electr…   ┆         ┆           ┆           │\n",
       "│ chr5     ┆ 45014943   ┆ rs11186928 ┆ C         ┆ … ┆ The       ┆ 3291609 ┆ www.ncbi. ┆ 3.0000e-8 │\n",
       "│          ┆            ┆ 2          ┆           ┆   ┆ Genetic   ┆ 8       ┆ nlm.nih.g ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Makeup of ┆         ┆ ov/pubmed ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ the       ┆         ┆ /3291…    ┆           │\n",
       "│          ┆            ┆            ┆           ┆   ┆ Electr…   ┆         ┆           ┆           │\n",
       "└──────────┴────────────┴────────────┴───────────┴───┴───────────┴─────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas = DF_GWAS_FULL.select([\n",
    "    pl.col(\"CHR_ID\"),\n",
    "    pl.col(\"CHR_POS\").cast(pl.Int32),\n",
    "    pl.col(\"SNPS\"),\n",
    "    pl.col(\"STRONGEST SNP-RISK ALLELE\"),\n",
    "    pl.col(\"RISK ALLELE FREQUENCY\"),\n",
    "    pl.col(\"DISEASE/TRAIT\"),\n",
    "    pl.col(\"STUDY\"),\n",
    "    pl.col(\"PUBMEDID\"),\n",
    "    pl.col(\"LINK\"),\n",
    "    pl.col(\"P-VALUE\")\n",
    "]).drop_nulls([\"CHR_ID\"]).with_columns([\n",
    "    (pl.lit(\"chr\") + pl.col(\"CHR_ID\")).alias(\"CHR_ID\"),\n",
    "    pl.col(\"SNPS\").str.split(\";\").arr.first(),\n",
    "    pl.col(\"STRONGEST SNP-RISK ALLELE\").str.split(\"-\").arr.last()\n",
    "]).filter(pl.col(\"P-VALUE\") <= 1)\n",
    "\n",
    "# Rename columns in the output\n",
    "df_gwas = df_gwas.rename({\n",
    "    \"CHR_ID\": \"Chromosome\",\n",
    "    \"CHR_POS\" : \"SNP_Position\",\n",
    "    \"SNPS\": \"SNP_Name_GWAS\",\n",
    "    \"STRONGEST SNP-RISK ALLELE\": \"Risk_Allele\",\n",
    "    \"RISK ALLELE FREQUENCY\": \"Risk_Allele_Frequency\",\n",
    "    \"DISEASE/TRAIT\": \"Phenotype\",\n",
    "    \"STUDY\": \"Study\",\n",
    "    \"PUBMEDID\": \"Pubmed_ID\",\n",
    "    \"LINK\": \"Link\",\n",
    "    \"P-VALUE\": \"P-Value\"\n",
    "})\n",
    "\n",
    "df_gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (87_183, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SV_Start</th><th>SV_End</th><th>SV_Name</th><th>SV_Type</th><th>SV_Length</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>i64</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>11025</td><td>11068</td><td>&quot;P0000000001&quot;</td><td>&quot;del&quot;</td><td>44</td></tr><tr><td>&quot;chr1&quot;</td><td>31973</td><td>32073</td><td>&quot;P0000000005&quot;</td><td>&quot;ins&quot;</td><td>101</td></tr><tr><td>&quot;chr1&quot;</td><td>50937</td><td>51053</td><td>&quot;P0000000006&quot;</td><td>&quot;del&quot;</td><td>117</td></tr><tr><td>&quot;chr1&quot;</td><td>54656</td><td>55064</td><td>&quot;P0000000007&quot;</td><td>&quot;ins&quot;</td><td>409</td></tr><tr><td>&quot;chr1&quot;</td><td>66162</td><td>66530</td><td>&quot;P0000000010&quot;</td><td>&quot;ins&quot;</td><td>369</td></tr><tr><td>&quot;chr1&quot;</td><td>66232</td><td>66275</td><td>&quot;P0000000011&quot;</td><td>&quot;del&quot;</td><td>44</td></tr><tr><td>&quot;chr1&quot;</td><td>67897</td><td>68333</td><td>&quot;P0000000014&quot;</td><td>&quot;del&quot;</td><td>437</td></tr><tr><td>&quot;chr1&quot;</td><td>90051</td><td>90287</td><td>&quot;P0000000020&quot;</td><td>&quot;del&quot;</td><td>237</td></tr><tr><td>&quot;chr1&quot;</td><td>90050</td><td>90442</td><td>&quot;P0000000021&quot;</td><td>&quot;ins&quot;</td><td>393</td></tr><tr><td>&quot;chr1&quot;</td><td>90129</td><td>90308</td><td>&quot;P0000000022&quot;</td><td>&quot;ins&quot;</td><td>180</td></tr><tr><td>&quot;chr1&quot;</td><td>95027</td><td>95127</td><td>&quot;P0000000023&quot;</td><td>&quot;ins&quot;</td><td>101</td></tr><tr><td>&quot;chr1&quot;</td><td>120985</td><td>121159</td><td>&quot;P0000000025&quot;</td><td>&quot;ins&quot;</td><td>175</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>155985485</td><td>156030781</td><td>&quot;P0000225345&quot;</td><td>&quot;complex&quot;</td><td>45297</td></tr><tr><td>&quot;chrX&quot;</td><td>156008862</td><td>156009184</td><td>&quot;P0000225349&quot;</td><td>&quot;dup&quot;</td><td>323</td></tr><tr><td>&quot;chrX&quot;</td><td>156020145</td><td>156030801</td><td>&quot;P0000225355&quot;</td><td>&quot;complex&quot;</td><td>10657</td></tr><tr><td>&quot;chrX&quot;</td><td>156021578</td><td>156022829</td><td>&quot;P0000225357&quot;</td><td>&quot;del&quot;</td><td>1252</td></tr><tr><td>&quot;chrX&quot;</td><td>156029013</td><td>156029427</td><td>&quot;P0000225359&quot;</td><td>&quot;ins&quot;</td><td>415</td></tr><tr><td>&quot;chrX&quot;</td><td>156029095</td><td>156030799</td><td>&quot;P0000225358&quot;</td><td>&quot;complex&quot;</td><td>1705</td></tr><tr><td>&quot;chrX&quot;</td><td>156029110</td><td>156029150</td><td>&quot;P0000225360&quot;</td><td>&quot;del&quot;</td><td>41</td></tr><tr><td>&quot;chrX&quot;</td><td>156029060</td><td>156029422</td><td>&quot;P0000225361&quot;</td><td>&quot;ins&quot;</td><td>363</td></tr><tr><td>&quot;chrX&quot;</td><td>156029104</td><td>156029422</td><td>&quot;P0000225363&quot;</td><td>&quot;ins&quot;</td><td>319</td></tr><tr><td>&quot;chrX&quot;</td><td>156029152</td><td>156029303</td><td>&quot;P0000225364&quot;</td><td>&quot;ins&quot;</td><td>152</td></tr><tr><td>&quot;chrX&quot;</td><td>156029257</td><td>156029314</td><td>&quot;P0000225365&quot;</td><td>&quot;del&quot;</td><td>58</td></tr><tr><td>&quot;chrX&quot;</td><td>156029320</td><td>156029379</td><td>&quot;P0000225366&quot;</td><td>&quot;del&quot;</td><td>60</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (87_183, 6)\n",
       "┌────────────┬───────────┬───────────┬─────────────┬─────────┬───────────┐\n",
       "│ Chromosome ┆ SV_Start  ┆ SV_End    ┆ SV_Name     ┆ SV_Type ┆ SV_Length │\n",
       "│ ---        ┆ ---       ┆ ---       ┆ ---         ┆ ---     ┆ ---       │\n",
       "│ str        ┆ i64       ┆ i64       ┆ str         ┆ str     ┆ i64       │\n",
       "╞════════════╪═══════════╪═══════════╪═════════════╪═════════╪═══════════╡\n",
       "│ chr1       ┆ 11025     ┆ 11068     ┆ P0000000001 ┆ del     ┆ 44        │\n",
       "│ chr1       ┆ 31973     ┆ 32073     ┆ P0000000005 ┆ ins     ┆ 101       │\n",
       "│ chr1       ┆ 50937     ┆ 51053     ┆ P0000000006 ┆ del     ┆ 117       │\n",
       "│ chr1       ┆ 54656     ┆ 55064     ┆ P0000000007 ┆ ins     ┆ 409       │\n",
       "│ …          ┆ …         ┆ …         ┆ …           ┆ …       ┆ …         │\n",
       "│ chrX       ┆ 156029104 ┆ 156029422 ┆ P0000225363 ┆ ins     ┆ 319       │\n",
       "│ chrX       ┆ 156029152 ┆ 156029303 ┆ P0000225364 ┆ ins     ┆ 152       │\n",
       "│ chrX       ┆ 156029257 ┆ 156029314 ┆ P0000225365 ┆ del     ┆ 58        │\n",
       "│ chrX       ┆ 156029320 ┆ 156029379 ┆ P0000225366 ┆ del     ┆ 60        │\n",
       "└────────────┴───────────┴───────────┴─────────────┴─────────┴───────────┘"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## SV Annotation data - created by Dr. Zhuozhi Wang on March 13, 2023\n",
    "DF_ANNO_FULL = pl.read_csv(\"data/sv_annotations.tsv\",\n",
    "                           separator = \"\\t\",\n",
    "                           quote_char = '',\n",
    "                           infer_schema_length = 1_000_000)\n",
    "\n",
    "# Select and rename specific columns\n",
    "df_anno = DF_ANNO_FULL.select([\n",
    "    pl.col(\"Chromosome\"),\n",
    "    pl.col(\"Start\").alias(\"SV_Start\"),\n",
    "    pl.col(\"End\").cast(pl.Int64).alias(\"SV_End\"),\n",
    "    pl.col(\"ID\").alias(\"SV_Name\"),\n",
    "    pl.col(\"Type\").alias(\"SV_Type\"),\n",
    "    pl.col(\"Length\").alias(\"SV_Length\")\n",
    "])\n",
    "\n",
    "df_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_100_088, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>SNP_Name</th><th>SNP_Position</th><th>SV_Name</th><th>SV_Position</th><th>R2</th><th>D&#x27;</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>i64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr8_220692_T_…</td><td>220692</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.000516</td><td>0.0734076</td></tr><tr><td>&quot;chr8_482881_G_…</td><td>482881</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.007804</td><td>0.100612</td></tr><tr><td>&quot;chr8_627101_A_…</td><td>627101</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.0292134</td><td>0.438748</td></tr><tr><td>&quot;chr8_685274_A_…</td><td>685274</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.0120243</td><td>0.180577</td></tr><tr><td>&quot;chr8_685275_C_…</td><td>685275</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.0107826</td><td>0.172748</td></tr><tr><td>&quot;chr8_698048_G_…</td><td>698048</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.009875</td><td>0.163646</td></tr><tr><td>&quot;chr8_756008_C_…</td><td>756008</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.00169</td><td>0.0636381</td></tr><tr><td>&quot;chr8_766894_C_…</td><td>766894</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.000098</td><td>0.0152936</td></tr><tr><td>&quot;chr8_813540_A_…</td><td>813540</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.000887</td><td>0.0725187</td></tr><tr><td>&quot;chr8_997830_A_…</td><td>997830</td><td>&quot;P0000096843&quot;</td><td>81301</td><td>0.000772</td><td>0.0844073</td></tr><tr><td>&quot;chr8_220692_T_…</td><td>220692</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>0.0253343</td><td>0.773603</td></tr><tr><td>&quot;chr8_482881_G_…</td><td>482881</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>0.0313019</td><td>0.233668</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX_155225806…</td><td>155225806</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>0.0511303</td><td>0.293499</td></tr><tr><td>&quot;chrX_155491696…</td><td>155491696</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>0.0152581</td><td>0.235849</td></tr><tr><td>&quot;chrX_155757485…</td><td>155757485</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>0.004401</td><td>0.104517</td></tr><tr><td>&quot;chrX_155954788…</td><td>155954788</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>0.140102</td><td>0.415345</td></tr><tr><td>&quot;chrX_156000962…</td><td>156000962</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>0.228066</td><td>0.786437</td></tr><tr><td>&quot;chrX_155050522…</td><td>155050522</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.008633</td><td>0.112857</td></tr><tr><td>&quot;chrX_155104381…</td><td>155104381</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.0162475</td><td>0.158743</td></tr><tr><td>&quot;chrX_155225806…</td><td>155225806</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.0184194</td><td>0.157063</td></tr><tr><td>&quot;chrX_155491696…</td><td>155491696</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.0330469</td><td>0.30947</td></tr><tr><td>&quot;chrX_155757485…</td><td>155757485</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.0320938</td><td>0.316576</td></tr><tr><td>&quot;chrX_155954788…</td><td>155954788</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.0664414</td><td>0.260534</td></tr><tr><td>&quot;chrX_156000962…</td><td>156000962</td><td>&quot;P0000225359&quot;</td><td>156029063</td><td>0.160441</td><td>0.588111</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_100_088, 6)\n",
       "┌────────────────────┬──────────────┬─────────────┬─────────────┬───────────┬───────────┐\n",
       "│ SNP_Name           ┆ SNP_Position ┆ SV_Name     ┆ SV_Position ┆ R2        ┆ D'        │\n",
       "│ ---                ┆ ---          ┆ ---         ┆ ---         ┆ ---       ┆ ---       │\n",
       "│ str                ┆ i64          ┆ str         ┆ i64         ┆ f64       ┆ f64       │\n",
       "╞════════════════════╪══════════════╪═════════════╪═════════════╪═══════════╪═══════════╡\n",
       "│ chr8_220692_T_C    ┆ 220692       ┆ P0000096843 ┆ 81301       ┆ 0.000516  ┆ 0.0734076 │\n",
       "│ chr8_482881_G_A    ┆ 482881       ┆ P0000096843 ┆ 81301       ┆ 0.007804  ┆ 0.100612  │\n",
       "│ chr8_627101_A_G    ┆ 627101       ┆ P0000096843 ┆ 81301       ┆ 0.0292134 ┆ 0.438748  │\n",
       "│ chr8_685274_A_G    ┆ 685274       ┆ P0000096843 ┆ 81301       ┆ 0.0120243 ┆ 0.180577  │\n",
       "│ …                  ┆ …            ┆ …           ┆ …           ┆ …         ┆ …         │\n",
       "│ chrX_155491696_G_C ┆ 155491696    ┆ P0000225359 ┆ 156029063   ┆ 0.0330469 ┆ 0.30947   │\n",
       "│ chrX_155757485_T_C ┆ 155757485    ┆ P0000225359 ┆ 156029063   ┆ 0.0320938 ┆ 0.316576  │\n",
       "│ chrX_155954788_C_T ┆ 155954788    ┆ P0000225359 ┆ 156029063   ┆ 0.0664414 ┆ 0.260534  │\n",
       "│ chrX_156000962_A_G ┆ 156000962    ┆ P0000225359 ┆ 156029063   ┆ 0.160441  ┆ 0.588111  │\n",
       "└────────────────────┴──────────────┴─────────────┴─────────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LD_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                       \"*_ld_stats.txt\"))\n",
    "LD_tables = [pl.read_csv(file_path,\n",
    "                         separator = \"\\t\",\n",
    "                         has_header = True,\n",
    "                         new_columns = [\"SNP_Name\", \n",
    "                                        \"SNP_Position\",\n",
    "                                        \"SV_Name\",\n",
    "                                        \"SV_Position\",\n",
    "                                        \"R2\",\n",
    "                                        \"D'\"])\n",
    "             for file_path in LD_file_paths]\n",
    "\n",
    "DF_LD = pl.concat(LD_tables)\n",
    "DF_LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8_814_835, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SNP_Position</th><th>SV_SNP_Name</th><th>Reference_Allele</th><th>Alternate_Allele</th><th>Sample_AF</th><th>gnomAD_nfe_AF</th><th>dbSNP</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr13&quot;</td><td>17767059</td><td>&quot;P0000147722&quot;</td><td>&quot;N&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.128713</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171311</td><td>&quot;chr13_18171311…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.589109</td><td>0.0974076</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171393</td><td>&quot;chr13_18171393…</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.29703</td><td>0.473681</td><td>&quot;rs76065716&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171453</td><td>&quot;chr13_18171453…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.623762</td><td>0.422176</td><td>&quot;rs9740229&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171494</td><td>&quot;chr13_18171494…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.559406</td><td>0.427025</td><td>&quot;rs78165403&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171497</td><td>&quot;chr13_18171497…</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.430693</td><td>0.440541</td><td>&quot;rs75406701&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171585</td><td>&quot;chr13_18171585…</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.405941</td><td>0.469126</td><td>&quot;rs1831219&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171823</td><td>&quot;chr13_18171823…</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.371287</td><td>0.085863</td><td>&quot;rs7337304&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171866</td><td>&quot;chr13_18171866…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.450495</td><td>0.27309</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171869</td><td>&quot;chr13_18171869…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.514852</td><td>0.282088</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171886</td><td>&quot;chr13_18171886…</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.504951</td><td>0.261488</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171889</td><td>&quot;chr13_18171889…</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>0.371287</td><td>0.251085</td><td>&quot;NA&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr16&quot;</td><td>90224639</td><td>&quot;chr16_90224639…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.158416</td><td>0.00003</td><td>&quot;rs371896311&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90224655</td><td>&quot;chr16_90224655…</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>0.549505</td><td>0.366169</td><td>&quot;rs10163278&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90224669</td><td>&quot;chr16_90224669…</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.29703</td><td>0.025566</td><td>&quot;rs200995380&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90225460</td><td>&quot;chr16_90225460…</td><td>&quot;CG&quot;</td><td>&quot;C&quot;</td><td>0.282178</td><td>0.000031</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90225585</td><td>&quot;chr16_90225585…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.227723</td><td>0.0</td><td>&quot;rs71225296&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90226403</td><td>&quot;chr16_90226403…</td><td>&quot;T&quot;</td><td>&quot;A&quot;</td><td>0.643564</td><td>0.31166</td><td>&quot;rs9746261&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90226650</td><td>&quot;chr16_90226650…</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.188119</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90227085</td><td>&quot;chr16_90227085…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.341584</td><td>0.001972</td><td>&quot;rs200019921&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90227369</td><td>&quot;chr16_90227369…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.470297</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90227485</td><td>&quot;chr16_90227485…</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.10396</td><td>0.000293</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90227639</td><td>&quot;chr16_90227639…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.410891</td><td>0.000586</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90228078</td><td>&quot;chr16_90228078…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>0.193069</td><td>0.0</td><td>&quot;rs367934211&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8_814_835, 8)\n",
       "┌──────────┬────────────┬───────────┬────────────┬────────────┬───────────┬────────────┬───────────┐\n",
       "│ Chromoso ┆ SNP_Positi ┆ SV_SNP_Na ┆ Reference_ ┆ Alternate_ ┆ Sample_AF ┆ gnomAD_nfe ┆ dbSNP     │\n",
       "│ me       ┆ on         ┆ me        ┆ Allele     ┆ Allele     ┆ ---       ┆ _AF        ┆ ---       │\n",
       "│ ---      ┆ ---        ┆ ---       ┆ ---        ┆ ---        ┆ f64       ┆ ---        ┆ str       │\n",
       "│ str      ┆ i64        ┆ str       ┆ str        ┆ str        ┆           ┆ f64        ┆           │\n",
       "╞══════════╪════════════╪═══════════╪════════════╪════════════╪═══════════╪════════════╪═══════════╡\n",
       "│ chr13    ┆ 17767059   ┆ P00001477 ┆ N          ┆ <INS>      ┆ 0.128713  ┆ 0.0        ┆ NA        │\n",
       "│          ┆            ┆ 22        ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr13    ┆ 18171311   ┆ chr13_181 ┆ A          ┆ G          ┆ 0.589109  ┆ 0.0974076  ┆ NA        │\n",
       "│          ┆            ┆ 71311_A_G ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr13    ┆ 18171393   ┆ chr13_181 ┆ T          ┆ C          ┆ 0.29703   ┆ 0.473681   ┆ rs7606571 │\n",
       "│          ┆            ┆ 71393_T_C ┆            ┆            ┆           ┆            ┆ 6         │\n",
       "│ chr13    ┆ 18171453   ┆ chr13_181 ┆ G          ┆ A          ┆ 0.623762  ┆ 0.422176   ┆ rs9740229 │\n",
       "│          ┆            ┆ 71453_G_A ┆            ┆            ┆           ┆            ┆           │\n",
       "│ …        ┆ …          ┆ …         ┆ …          ┆ …          ┆ …         ┆ …          ┆ …         │\n",
       "│ chr16    ┆ 90227369   ┆ chr16_902 ┆ C          ┆ T          ┆ 0.470297  ┆ 0.0        ┆ NA        │\n",
       "│          ┆            ┆ 27369_C_T ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90227485   ┆ chr16_902 ┆ A          ┆ T          ┆ 0.10396   ┆ 0.000293   ┆ NA        │\n",
       "│          ┆            ┆ 27485_A_T ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90227639   ┆ chr16_902 ┆ C          ┆ T          ┆ 0.410891  ┆ 0.000586   ┆ NA        │\n",
       "│          ┆            ┆ 27639_C_T ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90228078   ┆ chr16_902 ┆ C          ┆ A          ┆ 0.193069  ┆ 0.0        ┆ rs3679342 │\n",
       "│          ┆            ┆ 28078_C_A ┆            ┆            ┆           ┆            ┆ 11        │\n",
       "└──────────┴────────────┴───────────┴────────────┴────────────┴───────────┴────────────┴───────────┘"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## SV/SNP allele annotations done by Thomas Nalpathamkalam on April 10, 2023\n",
    "allele_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                            \"*_allele_freq.txt\"))\n",
    "allele_tables = [pl.read_csv(file_path,\n",
    "                             separator = \"\\t\",\n",
    "                             has_header = True,\n",
    "                             new_columns = [\"Chromosome\",\n",
    "                                            \"SNP_Position\",\n",
    "                                            \"SV_SNP_Name\",\n",
    "                                            \"Reference_Allele\",\n",
    "                                            \"Alternate_Allele\",\n",
    "                                            \"Sample_AF\",\n",
    "                                            \"gnomAD_nfe_AF\",\n",
    "                                            \"dbSNP\"])\n",
    "                 for file_path in allele_file_paths]\n",
    "\n",
    "DF_ALLELE = pl.concat(allele_tables)\n",
    "DF_ALLELE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'transcript_id', 'transcript_type', 'transcript_name', 'tag', 'protein_id', 'db_xref', 'exon_number', 'exon_id']\n"
     ]
    }
   ],
   "source": [
    "## Gene/exon data from MANE.GRCh38.v1.0.ensembl_genomic.gtf extracted January 5, 2023\n",
    "\n",
    "DF_GTF = read_gtf(\"data/MANE.GRCh38.v1.0.ensembl_genomic.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (220_152, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chr</th><th>start</th><th>end</th><th>strand</th><th>feature</th><th>gene</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>1471765</td><td>1497848</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1471765</td><td>1472089</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1477274</td><td>1477350</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1478644</td><td>1478745</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1479049</td><td>1479108</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1480867</td><td>1480936</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1482138</td><td>1482303</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1482545</td><td>1482614</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1485016</td><td>1485171</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1485782</td><td>1485838</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1486110</td><td>1486235</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>1486544</td><td>1486668</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;ATAD3B&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrY&quot;</td><td>22415029</td><td>22415139</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;RBMY1J&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>22415574</td><td>22415684</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;RBMY1J&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>22416608</td><td>22416718</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;RBMY1J&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>22416805</td><td>22416893</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;RBMY1J&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>22417305</td><td>22417881</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;RBMY1J&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9487267</td><td>9490081</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9487267</td><td>9487816</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9488424</td><td>9488501</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9488630</td><td>9488741</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9488843</td><td>9488988</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9489095</td><td>9489176</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr><tr><td>&quot;chrY&quot;</td><td>9489871</td><td>9490081</td><td>&quot;+&quot;</td><td>&quot;exon&quot;</td><td>&quot;TSPY9&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (220_152, 6)\n",
       "┌──────┬─────────┬─────────┬────────┬────────────┬────────┐\n",
       "│ chr  ┆ start   ┆ end     ┆ strand ┆ feature    ┆ gene   │\n",
       "│ ---  ┆ ---     ┆ ---     ┆ ---    ┆ ---        ┆ ---    │\n",
       "│ str  ┆ i64     ┆ i64     ┆ str    ┆ str        ┆ str    │\n",
       "╞══════╪═════════╪═════════╪════════╪════════════╪════════╡\n",
       "│ chr1 ┆ 1471765 ┆ 1497848 ┆ +      ┆ transcript ┆ ATAD3B │\n",
       "│ chr1 ┆ 1471765 ┆ 1472089 ┆ +      ┆ exon       ┆ ATAD3B │\n",
       "│ chr1 ┆ 1477274 ┆ 1477350 ┆ +      ┆ exon       ┆ ATAD3B │\n",
       "│ chr1 ┆ 1478644 ┆ 1478745 ┆ +      ┆ exon       ┆ ATAD3B │\n",
       "│ …    ┆ …       ┆ …       ┆ …      ┆ …          ┆ …      │\n",
       "│ chrY ┆ 9488630 ┆ 9488741 ┆ +      ┆ exon       ┆ TSPY9  │\n",
       "│ chrY ┆ 9488843 ┆ 9488988 ┆ +      ┆ exon       ┆ TSPY9  │\n",
       "│ chrY ┆ 9489095 ┆ 9489176 ┆ +      ┆ exon       ┆ TSPY9  │\n",
       "│ chrY ┆ 9489871 ┆ 9490081 ┆ +      ┆ exon       ┆ TSPY9  │\n",
       "└──────┴─────────┴─────────┴────────┴────────────┴────────┘"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene = pl.DataFrame(DF_GTF).filter(\n",
    "    (pl.col(\"feature\") == \"transcript\") | (pl.col(\"feature\") == \"exon\")\n",
    ").filter(pl.col(\"tag\").str.contains(\"MANE_Select\")).select([\n",
    "    pl.col(\"seqname\").alias(\"chr\"),\n",
    "    pl.col(\"start\").cast(pl.Int64),\n",
    "    pl.col(\"end\").cast(pl.Int64),\n",
    "    pl.col(\"strand\"),\n",
    "    pl.col(\"feature\"),\n",
    "    pl.col(\"gene_name\").alias(\"gene\")\n",
    "])\n",
    "\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (36_295, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SV_Name</th><th>Alternate_Allele</th><th>Sample_AF</th></tr><tr><td>str</td><td>str</td><td>str</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147727&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.846535</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147729&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.49505</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147765&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.841584</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147871&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.79703</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147936&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.386139</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147943&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.178218</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147976&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.212871</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000147992&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.158416</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000148034&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.20297</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000148077&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.207921</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000148098&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.10396</td></tr><tr><td>&quot;chr13&quot;</td><td>&quot;P0000148102&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.39604</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175235&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.128713</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175253&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.207921</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175278&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.10396</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175302&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.732673</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175332&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.534653</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175342&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.460396</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175392&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.193069</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175420&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.158416</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175461&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.19802</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175464&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.242574</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175509&quot;</td><td>&quot;&lt;DEL&gt;&quot;</td><td>0.163366</td></tr><tr><td>&quot;chr16&quot;</td><td>&quot;P0000175551&quot;</td><td>&quot;&lt;INS&gt;&quot;</td><td>0.534653</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (36_295, 4)\n",
       "┌────────────┬─────────────┬──────────────────┬───────────┐\n",
       "│ Chromosome ┆ SV_Name     ┆ Alternate_Allele ┆ Sample_AF │\n",
       "│ ---        ┆ ---         ┆ ---              ┆ ---       │\n",
       "│ str        ┆ str         ┆ str              ┆ f64       │\n",
       "╞════════════╪═════════════╪══════════════════╪═══════════╡\n",
       "│ chr13      ┆ P0000147727 ┆ <DEL>            ┆ 0.846535  │\n",
       "│ chr13      ┆ P0000147729 ┆ <INS>            ┆ 0.49505   │\n",
       "│ chr13      ┆ P0000147765 ┆ <INS>            ┆ 0.841584  │\n",
       "│ chr13      ┆ P0000147871 ┆ <DEL>            ┆ 0.79703   │\n",
       "│ …          ┆ …           ┆ …                ┆ …         │\n",
       "│ chr16      ┆ P0000175461 ┆ <INS>            ┆ 0.19802   │\n",
       "│ chr16      ┆ P0000175464 ┆ <INS>            ┆ 0.242574  │\n",
       "│ chr16      ┆ P0000175509 ┆ <DEL>            ┆ 0.163366  │\n",
       "│ chr16      ┆ P0000175551 ┆ <INS>            ┆ 0.534653  │\n",
       "└────────────┴─────────────┴──────────────────┴───────────┘"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### DATA WRANGLING ####\n",
    "## Subset allele data for only SV information\n",
    "df_sv_allele = (DF_ALLELE\n",
    "                .filter(pl.col(\"SV_SNP_Name\").str.starts_with(\"P\"))\n",
    "                .rename({\"SV_SNP_Name\": \"SV_Name\"})\n",
    "                .drop([\"Reference_Allele\", \"SNP_Position\", \"gnomAD_nfe_AF\", \"dbSNP\"])\n",
    "                .unique())\n",
    "df_sv_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8_518_382, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SNP_Position</th><th>SNP_Name</th><th>Reference_Allele</th><th>Alternate_Allele</th><th>Sample_AF</th><th>gnomAD_nfe_AF</th><th>dbSNP</th></tr><tr><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr13&quot;</td><td>18171869</td><td>&quot;chr13_18171869…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.514852</td><td>0.282088</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18171889</td><td>&quot;chr13_18171889…</td><td>&quot;T&quot;</td><td>&quot;G&quot;</td><td>0.371287</td><td>0.251085</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172188</td><td>&quot;chr13_18172188…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.465347</td><td>0.498936</td><td>&quot;rs3966844&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172252</td><td>&quot;chr13_18172252…</td><td>&quot;T&quot;</td><td>&quot;TGTGCC&quot;</td><td>0.306931</td><td>0.324367</td><td>&quot;rs71206482&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172261</td><td>&quot;chr13_18172261…</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.475248</td><td>0.503913</td><td>&quot;rs80043307&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172297</td><td>&quot;chr13_18172297…</td><td>&quot;A&quot;</td><td>&quot;T&quot;</td><td>0.504951</td><td>0.503739</td><td>&quot;rs78685402&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172566</td><td>&quot;chr13_18172566…</td><td>&quot;T&quot;</td><td>&quot;A&quot;</td><td>0.509901</td><td>0.021309</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18172868</td><td>&quot;chr13_18172868…</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.465347</td><td>0.0652766</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18173728</td><td>&quot;chr13_18173728…</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.435644</td><td>0.0494675</td><td>&quot;rs76277713&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18173863</td><td>&quot;chr13_18173863…</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>0.509901</td><td>0.493995</td><td>&quot;rs7491599&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18173875</td><td>&quot;chr13_18173875…</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.504951</td><td>0.492239</td><td>&quot;rs7489673&quot;</td></tr><tr><td>&quot;chr13&quot;</td><td>18173887</td><td>&quot;chr13_18173887…</td><td>&quot;AT&quot;</td><td>&quot;A&quot;</td><td>0.173267</td><td>0.186011</td><td>&quot;rs386832888&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chr16&quot;</td><td>90203981</td><td>&quot;chr16_90203981…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.173267</td><td>0.002038</td><td>&quot;rs369340635&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90206948</td><td>&quot;chr16_90206948…</td><td>&quot;G&quot;</td><td>&quot;T&quot;</td><td>0.247525</td><td>0.003526</td><td>&quot;rs879632557&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90208536</td><td>&quot;chr16_90208536…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.450495</td><td>0.344604</td><td>&quot;rs374101430&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90212798</td><td>&quot;chr16_90212798…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.108911</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90213176</td><td>&quot;chr16_90213176…</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>0.594059</td><td>0.000388</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90214038</td><td>&quot;chr16_90214038…</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>0.10396</td><td>0.000015</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90214130</td><td>&quot;chr16_90214130…</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.158416</td><td>0.000075</td><td>&quot;rs12232469&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90214183</td><td>&quot;chr16_90214183…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.376238</td><td>0.00003</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90214782</td><td>&quot;chr16_90214782…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.405941</td><td>0.000567</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90219453</td><td>&quot;chr16_90219453…</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.128713</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90221569</td><td>&quot;chr16_90221569…</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.10396</td><td>0.0</td><td>&quot;NA&quot;</td></tr><tr><td>&quot;chr16&quot;</td><td>90224626</td><td>&quot;chr16_90224626…</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.113861</td><td>0.000907</td><td>&quot;rs368080977&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8_518_382, 8)\n",
       "┌──────────┬────────────┬───────────┬────────────┬────────────┬───────────┬────────────┬───────────┐\n",
       "│ Chromoso ┆ SNP_Positi ┆ SNP_Name  ┆ Reference_ ┆ Alternate_ ┆ Sample_AF ┆ gnomAD_nfe ┆ dbSNP     │\n",
       "│ me       ┆ on         ┆ ---       ┆ Allele     ┆ Allele     ┆ ---       ┆ _AF        ┆ ---       │\n",
       "│ ---      ┆ ---        ┆ str       ┆ ---        ┆ ---        ┆ f64       ┆ ---        ┆ str       │\n",
       "│ str      ┆ i64        ┆           ┆ str        ┆ str        ┆           ┆ f64        ┆           │\n",
       "╞══════════╪════════════╪═══════════╪════════════╪════════════╪═══════════╪════════════╪═══════════╡\n",
       "│ chr13    ┆ 18171869   ┆ chr13_181 ┆ C          ┆ T          ┆ 0.514852  ┆ 0.282088   ┆ NA        │\n",
       "│          ┆            ┆ 71869_C_T ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr13    ┆ 18171889   ┆ chr13_181 ┆ T          ┆ G          ┆ 0.371287  ┆ 0.251085   ┆ NA        │\n",
       "│          ┆            ┆ 71889_T_G ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr13    ┆ 18172188   ┆ chr13_181 ┆ A          ┆ G          ┆ 0.465347  ┆ 0.498936   ┆ rs3966844 │\n",
       "│          ┆            ┆ 72188_A_G ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr13    ┆ 18172252   ┆ chr13_181 ┆ T          ┆ TGTGCC     ┆ 0.306931  ┆ 0.324367   ┆ rs7120648 │\n",
       "│          ┆            ┆ 72252_T_T ┆            ┆            ┆           ┆            ┆ 2         │\n",
       "│          ┆            ┆ GTGCC     ┆            ┆            ┆           ┆            ┆           │\n",
       "│ …        ┆ …          ┆ …         ┆ …          ┆ …          ┆ …         ┆ …          ┆ …         │\n",
       "│ chr16    ┆ 90214782   ┆ chr16_902 ┆ A          ┆ G          ┆ 0.405941  ┆ 0.000567   ┆ NA        │\n",
       "│          ┆            ┆ 14782_A_G ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90219453   ┆ chr16_902 ┆ T          ┆ C          ┆ 0.128713  ┆ 0.0        ┆ NA        │\n",
       "│          ┆            ┆ 19453_T_C ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90221569   ┆ chr16_902 ┆ A          ┆ G          ┆ 0.10396   ┆ 0.0        ┆ NA        │\n",
       "│          ┆            ┆ 21569_A_G ┆            ┆            ┆           ┆            ┆           │\n",
       "│ chr16    ┆ 90224626   ┆ chr16_902 ┆ G          ┆ A          ┆ 0.113861  ┆ 0.000907   ┆ rs3680809 │\n",
       "│          ┆            ┆ 24626_G_A ┆            ┆            ┆           ┆            ┆ 77        │\n",
       "└──────────┴────────────┴───────────┴────────────┴────────────┴───────────┴────────────┴───────────┘"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Subset allele data for only SNP information\n",
    "df_snp_allele = (DF_ALLELE.filter(~pl.col(\"SV_SNP_Name\").str.starts_with(\"P\"))\n",
    "                 .rename({\"SV_SNP_Name\": \"SNP_Name\"})\n",
    "                 .unique())\n",
    "df_snp_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (36_295, 38)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>SV Name</th><th>Chromosome</th><th>Start</th><th>End</th><th>Type</th><th>Length</th><th>SV Sample AF</th><th>GC (%)</th><th>Cytoband</th><th>Gene Count</th><th>Gene Name(s)</th><th>Gene at Start</th><th>Gene at End</th><th>Exon Name</th><th>CDS Name</th><th>Dark Genes % Overlap</th><th>ClinGen Haploinsufficient</th><th>ClinGen Triplosensitive</th><th>gnomAD O/E LoF Upper</th><th>gnomAD O/E Mis Upper</th><th>gnomAD pLI</th><th>gnomAD pRec</th><th>Repeat % Overlap</th><th>Dirty Region % Overlap</th><th>Chromosome Region</th><th>CGD</th><th>OMIM Pheno</th><th>OMIM Inh</th><th>ClinGen Region</th><th>Decipher Region</th><th>ClinVar VarID</th><th>gnomAD AF Max 90% RO</th><th>gnomAD Population AF Max 90% RO</th><th>gnomAD Hom/Ref Frequency 90% RO</th><th>gnomAD Het Frequency 90% RO</th><th>gnomAD Hom/Alt Frequency 90% RO</th><th>DGV % Overlap</th><th>DGV 50% RO</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>i64</td><td>f64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td></tr></thead><tbody><tr><td>&quot;P0000000010&quot;</td><td>&quot;chr1&quot;</td><td>66162</td><td>66530.0</td><td>&quot;ins&quot;</td><td>369</td><td>0.133663</td><td>&quot;1&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;OR4F5&quot;</td><td>&quot;OR4F5&quot;</td><td>&quot;OR4F5&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;OR4F5:1.828&quot;</td><td>&quot;OR4F5:1.33&quot;</td><td>&quot;OR4F5:0.34887&quot;</td><td>&quot;OR4F5:0.53121&quot;</td><td>100.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000011&quot;</td><td>&quot;chr1&quot;</td><td>66232</td><td>66275.0</td><td>&quot;del&quot;</td><td>44</td><td>0.168317</td><td>&quot;0&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;OR4F5&quot;</td><td>&quot;OR4F5&quot;</td><td>&quot;OR4F5&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;OR4F5:1.828&quot;</td><td>&quot;OR4F5:1.33&quot;</td><td>&quot;OR4F5:0.34887&quot;</td><td>&quot;OR4F5:0.53121&quot;</td><td>100.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;100&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000020&quot;</td><td>&quot;chr1&quot;</td><td>90051</td><td>90287.0</td><td>&quot;del&quot;</td><td>237</td><td>0.123762</td><td>&quot;56&quot;</td><td>&quot;1p36.33&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>2.53</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;100&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000021&quot;</td><td>&quot;chr1&quot;</td><td>90050</td><td>90442.0</td><td>&quot;ins&quot;</td><td>393</td><td>0.460396</td><td>&quot;56&quot;</td><td>&quot;1p36.33&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>1.78</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000029&quot;</td><td>&quot;chr1&quot;</td><td>136302</td><td>137043.0</td><td>&quot;ins&quot;</td><td>742</td><td>0.475248</td><td>&quot;70&quot;</td><td>&quot;1p36.33&quot;</td><td>2</td><td>&quot;LOC124900384|L…</td><td>&quot;LOC124900384|L…</td><td>&quot;LOC124900384|L…</td><td>&quot;LOC729737&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000034&quot;</td><td>&quot;chr1&quot;</td><td>157298</td><td>157398.0</td><td>&quot;ins&quot;</td><td>101</td><td>0.173267</td><td>&quot;32&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC124900384&quot;</td><td>&quot;LOC124900384&quot;</td><td>&quot;LOC124900384&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000053&quot;</td><td>&quot;chr1&quot;</td><td>368870</td><td>369948.0</td><td>&quot;ins&quot;</td><td>1079</td><td>0.178218</td><td>&quot;61&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000051&quot;</td><td>&quot;chr1&quot;</td><td>368950</td><td>369341.0</td><td>&quot;del&quot;</td><td>392</td><td>0.450495</td><td>&quot;60&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;100&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000065&quot;</td><td>&quot;chr1&quot;</td><td>372573</td><td>372758.0</td><td>&quot;del&quot;</td><td>186</td><td>0.425743</td><td>&quot;34&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>83.33</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;100&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000072&quot;</td><td>&quot;chr1&quot;</td><td>373923</td><td>374190.0</td><td>&quot;ins&quot;</td><td>268</td><td>0.450495</td><td>&quot;59&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000077&quot;</td><td>&quot;chr1&quot;</td><td>381104</td><td>381747.0</td><td>&quot;ins&quot;</td><td>644</td><td>0.589109</td><td>&quot;74&quot;</td><td>&quot;1p36.33&quot;</td><td>1</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>&quot;LOC112268260&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000000082&quot;</td><td>&quot;chr1&quot;</td><td>382439</td><td>382610.0</td><td>&quot;del&quot;</td><td>172</td><td>0.539604</td><td>&quot;51&quot;</td><td>&quot;1p36.33&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>100.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;1p36 microdele…</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;100&quot;</td><td>0.0</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;P0000225309&quot;</td><td>&quot;chrX&quot;</td><td>155560527</td><td>1.55573602e8</td><td>&quot;del&quot;</td><td>13076</td><td>0.247525</td><td>&quot;42&quot;</td><td>&quot;Xq28&quot;</td><td>1</td><td>&quot;TMLHE&quot;</td><td>&quot;TMLHE&quot;</td><td>&quot;TMLHE&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;TMLHE:1&quot;</td><td>&quot;TMLHE:0&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>77.65</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;TMLHE:Epsilon-…</td><td>&quot;TMLHE:{Autism,…</td><td>&quot;XR,XL&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0.89893&quot;</td><td>&quot;AF_fin&quot;</td><td>&quot;0.136916&quot;</td><td>&quot;0.38686&quot;</td><td>&quot;0.476224&quot;</td><td>&quot;100&quot;</td><td>100.0</td></tr><tr><td>&quot;P0000225312&quot;</td><td>&quot;chrX&quot;</td><td>155638294</td><td>1.55638338e8</td><td>&quot;del&quot;</td><td>45</td><td>0.79703</td><td>&quot;0&quot;</td><td>&quot;Xq28&quot;</td><td>1</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225314&quot;</td><td>&quot;chrX&quot;</td><td>155641367</td><td>1.559582e8</td><td>&quot;dup&quot;</td><td>316834</td><td>0.173267</td><td>&quot;NA&quot;</td><td>&quot;Xq28&quot;</td><td>2</td><td>&quot;SPRY3|VAMP7&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;NA&quot;</td><td>&quot;SPRY3|VAMP7&quot;</td><td>&quot;SPRY3|VAMP7&quot;</td><td>0.0</td><td>&quot;VAMP7:0&quot;</td><td>&quot;VAMP7:0&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>75.3</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225317&quot;</td><td>&quot;chrX&quot;</td><td>155688799</td><td>1.55691958e8</td><td>&quot;del&quot;</td><td>3160</td><td>0.173267</td><td>&quot;34&quot;</td><td>&quot;Xq28&quot;</td><td>1</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>98.64</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0.513415&quot;</td><td>&quot;AF_afr&quot;</td><td>&quot;0.555834&quot;</td><td>&quot;0.397691&quot;</td><td>&quot;0.0464749&quot;</td><td>&quot;100&quot;</td><td>100.0</td></tr><tr><td>&quot;P0000225318&quot;</td><td>&quot;chrX&quot;</td><td>155700563</td><td>1.55700683e8</td><td>&quot;ins&quot;</td><td>121</td><td>0.237624</td><td>&quot;28&quot;</td><td>&quot;Xq28&quot;</td><td>1</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;SPRY3&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>97.52</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225330&quot;</td><td>&quot;chrX&quot;</td><td>155803458</td><td>1.55803911e8</td><td>&quot;ins&quot;</td><td>454</td><td>0.504951</td><td>&quot;45&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>98.68</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225332&quot;</td><td>&quot;chrX&quot;</td><td>155803592</td><td>1.55803811e8</td><td>&quot;dup&quot;</td><td>220</td><td>0.49505</td><td>&quot;54&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225335&quot;</td><td>&quot;chrX&quot;</td><td>155803823</td><td>1.55984417e8</td><td>&quot;del&quot;</td><td>180595</td><td>0.519802</td><td>&quot;NA&quot;</td><td>&quot;Xq28&quot;</td><td>1</td><td>&quot;VAMP7&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;VAMP7&quot;</td><td>&quot;VAMP7&quot;</td><td>0.0</td><td>&quot;VAMP7:0&quot;</td><td>&quot;VAMP7:0&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>76.58</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;2.3&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225342&quot;</td><td>&quot;chrX&quot;</td><td>155873308</td><td>1.55873715e8</td><td>&quot;ins&quot;</td><td>408</td><td>0.311881</td><td>&quot;41&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>28.68</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225347&quot;</td><td>&quot;chrX&quot;</td><td>155985105</td><td>1.55985196e8</td><td>&quot;del&quot;</td><td>92</td><td>0.504951</td><td>&quot;42&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>93.48</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225357&quot;</td><td>&quot;chrX&quot;</td><td>156021578</td><td>1.56022829e8</td><td>&quot;del&quot;</td><td>1252</td><td>0.584158</td><td>&quot;62&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;0&quot;</td><td>0.0</td></tr><tr><td>&quot;P0000225359&quot;</td><td>&quot;chrX&quot;</td><td>156029013</td><td>1.56029427e8</td><td>&quot;ins&quot;</td><td>415</td><td>0.361386</td><td>&quot;80&quot;</td><td>&quot;Xq28&quot;</td><td>0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>100.0</td><td>0.0</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>&quot;NA&quot;</td><td>0.0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (36_295, 38)\n",
       "┌───────────┬──────────┬──────────┬──────────┬───┬────────────┬────────────┬────────────┬──────────┐\n",
       "│ SV Name   ┆ Chromoso ┆ Start    ┆ End      ┆ … ┆ gnomAD Het ┆ gnomAD     ┆ DGV %      ┆ DGV 50%  │\n",
       "│ ---       ┆ me       ┆ ---      ┆ ---      ┆   ┆ Frequency  ┆ Hom/Alt    ┆ Overlap    ┆ RO       │\n",
       "│ str       ┆ ---      ┆ i64      ┆ f64      ┆   ┆ 90% RO     ┆ Frequency  ┆ ---        ┆ ---      │\n",
       "│           ┆ str      ┆          ┆          ┆   ┆ ---        ┆ 90% RO     ┆ str        ┆ f64      │\n",
       "│           ┆          ┆          ┆          ┆   ┆ str        ┆ ---        ┆            ┆          │\n",
       "│           ┆          ┆          ┆          ┆   ┆            ┆ str        ┆            ┆          │\n",
       "╞═══════════╪══════════╪══════════╪══════════╪═══╪════════════╪════════════╪════════════╪══════════╡\n",
       "│ P00000000 ┆ chr1     ┆ 66162    ┆ 66530.0  ┆ … ┆ NA         ┆ NA         ┆ NA         ┆ 0.0      │\n",
       "│ 10        ┆          ┆          ┆          ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00000000 ┆ chr1     ┆ 66232    ┆ 66275.0  ┆ … ┆ NA         ┆ NA         ┆ 100        ┆ 0.0      │\n",
       "│ 11        ┆          ┆          ┆          ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00000000 ┆ chr1     ┆ 90051    ┆ 90287.0  ┆ … ┆ NA         ┆ NA         ┆ 100        ┆ 0.0      │\n",
       "│ 20        ┆          ┆          ┆          ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00000000 ┆ chr1     ┆ 90050    ┆ 90442.0  ┆ … ┆ NA         ┆ NA         ┆ NA         ┆ 0.0      │\n",
       "│ 21        ┆          ┆          ┆          ┆   ┆            ┆            ┆            ┆          │\n",
       "│ …         ┆ …        ┆ …        ┆ …        ┆ … ┆ …          ┆ …          ┆ …          ┆ …        │\n",
       "│ P00002253 ┆ chrX     ┆ 15587330 ┆ 1.558737 ┆ … ┆ NA         ┆ NA         ┆ NA         ┆ 0.0      │\n",
       "│ 42        ┆          ┆ 8        ┆ 15e8     ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00002253 ┆ chrX     ┆ 15598510 ┆ 1.559851 ┆ … ┆ NA         ┆ NA         ┆ 0          ┆ 0.0      │\n",
       "│ 47        ┆          ┆ 5        ┆ 96e8     ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00002253 ┆ chrX     ┆ 15602157 ┆ 1.560228 ┆ … ┆ NA         ┆ NA         ┆ 0          ┆ 0.0      │\n",
       "│ 57        ┆          ┆ 8        ┆ 29e8     ┆   ┆            ┆            ┆            ┆          │\n",
       "│ P00002253 ┆ chrX     ┆ 15602901 ┆ 1.560294 ┆ … ┆ NA         ┆ NA         ┆ NA         ┆ 0.0      │\n",
       "│ 59        ┆          ┆ 3        ┆ 27e8     ┆   ┆            ┆            ┆            ┆          │\n",
       "└───────────┴──────────┴──────────┴──────────┴───┴────────────┴────────────┴────────────┴──────────┘"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Subset DF_ANNO_FULL with columns that can be displayed to the public. Add SV AF to this table as well.\n",
    "df_sv_anno = (DF_ANNO_FULL\n",
    "              .join(\n",
    "                  df_sv_allele.select([\"SV_Name\",\n",
    "                                       \"Alternate_Allele\", \n",
    "                                       \"Sample_AF\"]),\n",
    "                                       left_on = \"ID\",\n",
    "                                       right_on = \"SV_Name\",\n",
    "                                       how = \"inner\")\n",
    "              .drop(\"Alternate_Allele\")\n",
    "              .select([\"ID\",\n",
    "                       *DF_ANNO_FULL.columns[1:6],\n",
    "                       \"Sample_AF\",\n",
    "                       *DF_ANNO_FULL.columns[6:]])\n",
    "              .rename({\"ID\": \"SV Name\",\n",
    "                       \"Sample_AF\": \"SV Sample AF\"}))\n",
    "df_sv_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_100_088, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SV_Name</th><th>SV_Start</th><th>SV_End</th><th>SV_Type</th><th>SV_AF</th><th>SNP_Name</th><th>SNP_Position</th><th>R2</th><th>D&#x27;</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>f64</td><td>str</td><td>i64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_220692_T_…</td><td>220692</td><td>0.000516</td><td>0.0734076</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_482881_G_…</td><td>482881</td><td>0.007804</td><td>0.100612</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_627101_A_…</td><td>627101</td><td>0.0292134</td><td>0.438748</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_685274_A_…</td><td>685274</td><td>0.0120243</td><td>0.180577</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_685275_C_…</td><td>685275</td><td>0.0107826</td><td>0.172748</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_698048_G_…</td><td>698048</td><td>0.009875</td><td>0.163646</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_756008_C_…</td><td>756008</td><td>0.00169</td><td>0.0636381</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_766894_C_…</td><td>766894</td><td>0.000098</td><td>0.0152936</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_813540_A_…</td><td>813540</td><td>0.000887</td><td>0.0725187</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;chr8_997830_A_…</td><td>997830</td><td>0.000772</td><td>0.0844073</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>87350</td><td>&quot;Deletion&quot;</td><td>0.10396</td><td>&quot;chr8_220692_T_…</td><td>220692</td><td>0.0253343</td><td>0.773603</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>87350</td><td>&quot;Deletion&quot;</td><td>0.10396</td><td>&quot;chr8_482881_G_…</td><td>482881</td><td>0.0313019</td><td>0.233668</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;chrX_155225806…</td><td>155225806</td><td>0.0511303</td><td>0.293499</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;chrX_155491696…</td><td>155491696</td><td>0.0152581</td><td>0.235849</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;chrX_155757485…</td><td>155757485</td><td>0.004401</td><td>0.104517</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;chrX_155954788…</td><td>155954788</td><td>0.140102</td><td>0.415345</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;chrX_156000962…</td><td>156000962</td><td>0.228066</td><td>0.786437</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155050522…</td><td>155050522</td><td>0.008633</td><td>0.112857</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155104381…</td><td>155104381</td><td>0.0162475</td><td>0.158743</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155225806…</td><td>155225806</td><td>0.0184194</td><td>0.157063</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155491696…</td><td>155491696</td><td>0.0330469</td><td>0.30947</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155757485…</td><td>155757485</td><td>0.0320938</td><td>0.316576</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_155954788…</td><td>155954788</td><td>0.0664414</td><td>0.260534</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;chrX_156000962…</td><td>156000962</td><td>0.160441</td><td>0.588111</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_100_088, 10)\n",
       "┌──────────┬───────────┬───────────┬───────────┬───┬───────────┬────────────┬───────────┬──────────┐\n",
       "│ Chromoso ┆ SV_Name   ┆ SV_Start  ┆ SV_End    ┆ … ┆ SNP_Name  ┆ SNP_Positi ┆ R2        ┆ D'       │\n",
       "│ me       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ on         ┆ ---       ┆ ---      │\n",
       "│ ---      ┆ str       ┆ i64       ┆ i64       ┆   ┆ str       ┆ ---        ┆ f64       ┆ f64      │\n",
       "│ str      ┆           ┆           ┆           ┆   ┆           ┆ i64        ┆           ┆          │\n",
       "╞══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪════════════╪═══════════╪══════════╡\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ chr8_2206 ┆ 220692     ┆ 0.000516  ┆ 0.073407 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆ 92_T_C    ┆            ┆           ┆ 6        │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ chr8_4828 ┆ 482881     ┆ 0.007804  ┆ 0.100612 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆ 81_G_A    ┆            ┆           ┆          │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ chr8_6271 ┆ 627101     ┆ 0.0292134 ┆ 0.438748 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆ 01_A_G    ┆            ┆           ┆          │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ chr8_6852 ┆ 685274     ┆ 0.0120243 ┆ 0.180577 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆ 74_A_G    ┆            ┆           ┆          │\n",
       "│ …        ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …          ┆ …         ┆ …        │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ chrX_1554 ┆ 155491696  ┆ 0.0330469 ┆ 0.30947  │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆ 91696_G_C ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ chrX_1557 ┆ 155757485  ┆ 0.0320938 ┆ 0.316576 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆ 57485_T_C ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ chrX_1559 ┆ 155954788  ┆ 0.0664414 ┆ 0.260534 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆ 54788_C_T ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ chrX_1560 ┆ 156000962  ┆ 0.160441  ┆ 0.588111 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆ 00962_A_G ┆            ┆           ┆          │\n",
       "└──────────┴───────────┴───────────┴───────────┴───┴───────────┴────────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Creation of SV table for easier access of data required - join the dataframes based on SV names\n",
    "df_sv_join = (DF_LD\n",
    "              .join(df_anno, \n",
    "                    on = \"SV_Name\", \n",
    "                    how = \"left\")\n",
    "              .select([\"Chromosome\", \n",
    "                       \"SV_Name\", \n",
    "                       \"SV_Start\", \n",
    "                       \"SV_End\",\n",
    "                       \"SNP_Name\", \n",
    "                       \"SNP_Position\", \n",
    "                       \"R2\", \n",
    "                       \"D'\"])\n",
    "              .join(df_sv_allele, \n",
    "                    on = [\"Chromosome\", \n",
    "                          \"SV_Name\"], \n",
    "                    how=\"left\")\n",
    "              .rename({\"Alternate_Allele\": \"SV_Type\",\n",
    "                       \"Sample_AF\": \"SV_AF\"})\n",
    "              .with_columns([pl.when(pl.col(\"SV_Type\") == \"<DEL>\").then(\"Deletion\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<INS>\").then(\"Insertion\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<DUP>\").then(\"Duplication\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<INV>\").then(\"Inversion\")\n",
    "                               .otherwise(pl.col(\"SV_Type\"))\n",
    "                               .alias(\"SV_Type\")])\n",
    "              .select([\"Chromosome\", \n",
    "                       \"SV_Name\", \n",
    "                       \"SV_Start\", \n",
    "                       \"SV_End\",\n",
    "                       \"SV_Type\", \n",
    "                       \"SV_AF\", \n",
    "                       \"SNP_Name\", \n",
    "                       \"SNP_Position\", \n",
    "                       \"R2\", \n",
    "                       \"D'\"]))\n",
    "df_sv_join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3_100_088, 14)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SV_Name</th><th>SV_Start</th><th>SV_End</th><th>SV_Type</th><th>SV_AF</th><th>SNP_Name_dbSNP</th><th>SNP_Position</th><th>Reference_Allele</th><th>Alternate_Allele</th><th>Sample_AF</th><th>gnomAD_nfe_AF</th><th>R2</th><th>D&#x27;</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>f64</td><td>str</td><td>i32</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.000516</td><td>0.0734076</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs74331971&quot;</td><td>482881</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.168317</td><td>0.0103772</td><td>0.007804</td><td>0.100612</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs336433&quot;</td><td>627101</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.633663</td><td>0.556564</td><td>0.0292134</td><td>0.438748</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11137025&quot;</td><td>685274</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.415842</td><td>0.379853</td><td>0.0120243</td><td>0.180577</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11137026&quot;</td><td>685275</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.420792</td><td>0.37983</td><td>0.0107826</td><td>0.172748</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs58004422&quot;</td><td>698048</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.415842</td><td>0.270646</td><td>0.009875</td><td>0.163646</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs10100132&quot;</td><td>756008</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.386139</td><td>0.201356</td><td>0.00169</td><td>0.0636381</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11784243&quot;</td><td>766894</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>0.386139</td><td>0.204618</td><td>0.000098</td><td>0.0152936</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs12676895&quot;</td><td>813540</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.391089</td><td>0.246815</td><td>0.000887</td><td>0.0725187</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs9650538&quot;</td><td>997830</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>0.292079</td><td>0.134747</td><td>0.000772</td><td>0.0844073</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>87350</td><td>&quot;Deletion&quot;</td><td>0.10396</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.0253343</td><td>0.773603</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096867&quot;</td><td>87278</td><td>87350</td><td>&quot;Deletion&quot;</td><td>0.10396</td><td>&quot;rs74331971&quot;</td><td>482881</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.168317</td><td>0.0103772</td><td>0.0313019</td><td>0.233668</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;rs35296872&quot;</td><td>155225806</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.29703</td><td>0.244635</td><td>0.0511303</td><td>0.293499</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;rs55695925&quot;</td><td>155491696</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.836634</td><td>0.872589</td><td>0.0152581</td><td>0.235849</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.004401</td><td>0.104517</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;rs75303099&quot;</td><td>155954788</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.366337</td><td>0.163391</td><td>0.140102</td><td>0.415345</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225357&quot;</td><td>156021578</td><td>156022829</td><td>&quot;Deletion&quot;</td><td>0.584158</td><td>&quot;rs3093479&quot;</td><td>156000962</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.207921</td><td>0.0795638</td><td>0.228066</td><td>0.786437</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs114209171&quot;</td><td>155050522</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.277228</td><td>0.228304</td><td>0.008633</td><td>0.112857</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs7051718&quot;</td><td>155104381</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.230898</td><td>0.0162475</td><td>0.158743</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs35296872&quot;</td><td>155225806</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.29703</td><td>0.244635</td><td>0.0184194</td><td>0.157063</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs55695925&quot;</td><td>155491696</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.836634</td><td>0.872589</td><td>0.0330469</td><td>0.30947</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs75303099&quot;</td><td>155954788</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.366337</td><td>0.163391</td><td>0.0664414</td><td>0.260534</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs3093479&quot;</td><td>156000962</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.207921</td><td>0.0795638</td><td>0.160441</td><td>0.588111</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3_100_088, 14)\n",
       "┌──────────┬───────────┬───────────┬───────────┬───┬───────────┬────────────┬───────────┬──────────┐\n",
       "│ Chromoso ┆ SV_Name   ┆ SV_Start  ┆ SV_End    ┆ … ┆ Sample_AF ┆ gnomAD_nfe ┆ R2        ┆ D'       │\n",
       "│ me       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ _AF        ┆ ---       ┆ ---      │\n",
       "│ ---      ┆ str       ┆ i64       ┆ i64       ┆   ┆ f64       ┆ ---        ┆ f64       ┆ f64      │\n",
       "│ str      ┆           ┆           ┆           ┆   ┆           ┆ f64        ┆           ┆          │\n",
       "╞══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪════════════╪═══════════╪══════════╡\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 0.267327  ┆ 0.215066   ┆ 0.000516  ┆ 0.073407 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆            ┆           ┆ 6        │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 0.168317  ┆ 0.0103772  ┆ 0.007804  ┆ 0.100612 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 0.633663  ┆ 0.556564   ┆ 0.0292134 ┆ 0.438748 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 0.415842  ┆ 0.379853   ┆ 0.0120243 ┆ 0.180577 │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ …        ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …          ┆ …         ┆ …        │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 0.836634  ┆ 0.872589   ┆ 0.0330469 ┆ 0.30947  │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 0.638614  ┆ 0.250862   ┆ 0.0320938 ┆ 0.316576 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 0.366337  ┆ 0.163391   ┆ 0.0664414 ┆ 0.260534 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 0.207921  ┆ 0.0795638  ┆ 0.160441  ┆ 0.588111 │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆            ┆           ┆          │\n",
       "└──────────┴───────────┴───────────┴───────────┴───┴───────────┴────────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Merging SNP allele information with the LD dataset\n",
    "df_sv_snp_join = (\n",
    "    df_sv_join\n",
    "    .join(\n",
    "        df_snp_allele, \n",
    "        on=[\"Chromosome\", \"SNP_Name\", \"SNP_Position\"], \n",
    "        how=\"left\"\n",
    "    )\n",
    "    .drop(\"SNP_Name\")\n",
    "    .with_columns(\n",
    "        pl.col(\"SNP_Position\").cast(pl.Int32))\n",
    "    .select([\n",
    "        \"Chromosome\", \"SV_Name\", \"SV_Start\", \"SV_End\", \"SV_Type\", \"SV_AF\",\n",
    "        \"dbSNP\", \"SNP_Position\", \n",
    "        \"Reference_Allele\", \"Alternate_Allele\", \"Sample_AF\", \"gnomAD_nfe_AF\", \"R2\", \"D'\"\n",
    "    ])  # Reorder columns\n",
    "    .rename({\"dbSNP\": \"SNP_Name_dbSNP\"})  # Rename column\n",
    ")\n",
    "df_sv_snp_join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (7_714_074, 23)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SV_Name</th><th>SV_Start</th><th>SV_End</th><th>SV_Type</th><th>SV_AF</th><th>SNP_Name_dbSNP</th><th>SNP_Position</th><th>Reference_Allele</th><th>Alternate_Allele</th><th>Sample_AF</th><th>gnomAD_nfe_AF</th><th>R2</th><th>D&#x27;</th><th>SNP_Name_GWAS</th><th>Risk_Allele</th><th>Risk_Allele_Frequency</th><th>Phenotype</th><th>Study</th><th>Pubmed_ID</th><th>Link</th><th>P-Value</th><th>P-Value_log10</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>f64</td><td>str</td><td>i32</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.000516</td><td>0.0734076</td><td>&quot;rs66823261&quot;</td><td>&quot;C&quot;</td><td>&quot;0.2289&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Association an…</td><td>&quot;29059683&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>0.0000003</td><td>6.522879</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.000516</td><td>0.0734076</td><td>&quot;rs66823261&quot;</td><td>&quot;C&quot;</td><td>&quot;0.23&quot;</td><td>&quot;Breast cancer …</td><td>&quot;Identification…</td><td>&quot;29058716&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>6.0000e-9</td><td>8.221849</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.000516</td><td>0.0734076</td><td>&quot;rs66823261&quot;</td><td>&quot;C&quot;</td><td>&quot;0.23&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Identification…</td><td>&quot;29058716&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-8</td><td>7.522879</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs66823261&quot;</td><td>220692</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.267327</td><td>0.215066</td><td>0.000516</td><td>0.0734076</td><td>&quot;rs66823261&quot;</td><td>&quot;T&quot;</td><td>&quot;NR&quot;</td><td>&quot;Breast cancer&quot;</td><td>&quot;Identification…</td><td>&quot;32139696&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-8</td><td>8.0</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs74331971&quot;</td><td>482881</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.168317</td><td>0.0103772</td><td>0.007804</td><td>0.100612</td><td>&quot;rs74331971&quot;</td><td>&quot;A&quot;</td><td>&quot;0.03&quot;</td><td>&quot;Neonatal cytok…</td><td>&quot;Cross-genetic …</td><td>&quot;30134952&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-9</td><td>8.69897</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs336433&quot;</td><td>627101</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.633663</td><td>0.556564</td><td>0.0292134</td><td>0.438748</td><td>&quot;rs336433&quot;</td><td>&quot;A&quot;</td><td>&quot;0.4398&quot;</td><td>&quot;Educational at…</td><td>&quot;Polygenic pred…</td><td>&quot;35361970&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-13</td><td>12.69897</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11137025&quot;</td><td>685274</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.415842</td><td>0.379853</td><td>0.0120243</td><td>0.180577</td><td>&quot;rs11137025&quot;</td><td>&quot;G&quot;</td><td>&quot;0.387&quot;</td><td>&quot;Smoking initia…</td><td>&quot;Genetic divers…</td><td>&quot;36477530&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-13</td><td>12.522879</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11137025&quot;</td><td>685274</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.415842</td><td>0.379853</td><td>0.0120243</td><td>0.180577</td><td>&quot;rs11137025&quot;</td><td>&quot;G&quot;</td><td>&quot;0.391&quot;</td><td>&quot;Smoking initia…</td><td>&quot;Genetic divers…</td><td>&quot;36477530&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-13</td><td>12.522879</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11137026&quot;</td><td>685275</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.420792</td><td>0.37983</td><td>0.0107826</td><td>0.172748</td><td>&quot;rs11137026&quot;</td><td>&quot;T&quot;</td><td>&quot;0.391&quot;</td><td>&quot;Smoking initia…</td><td>&quot;Genetic divers…</td><td>&quot;36477530&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>7.0000e-13</td><td>12.154902</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs58004422&quot;</td><td>698048</td><td>&quot;G&quot;</td><td>&quot;A&quot;</td><td>0.415842</td><td>0.270646</td><td>0.009875</td><td>0.163646</td><td>&quot;rs58004422&quot;</td><td>&quot;A&quot;</td><td>&quot;NR&quot;</td><td>&quot;Hemoglobin&quot;</td><td>&quot;A cross-popula…</td><td>&quot;34594039&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-8</td><td>8.0</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs10100132&quot;</td><td>756008</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.386139</td><td>0.201356</td><td>0.00169</td><td>0.0636381</td><td>&quot;rs10100132&quot;</td><td>&quot;T&quot;</td><td>&quot;0.202&quot;</td><td>&quot;Drinks per wee…</td><td>&quot;Genetic divers…</td><td>&quot;36477530&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-9</td><td>8.69897</td></tr><tr><td>&quot;chr8&quot;</td><td>&quot;P0000096843&quot;</td><td>81251</td><td>81609</td><td>&quot;Insertion&quot;</td><td>0.207921</td><td>&quot;rs11784243&quot;</td><td>766894</td><td>&quot;C&quot;</td><td>&quot;G&quot;</td><td>0.386139</td><td>0.204618</td><td>0.000098</td><td>0.0152936</td><td>&quot;rs11784243&quot;</td><td>&quot;G&quot;</td><td>&quot;0.194&quot;</td><td>&quot;Drinks per wee…</td><td>&quot;Genetic divers…</td><td>&quot;36477530&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-10</td><td>9.522879</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs55695925&quot;</td><td>155491696</td><td>&quot;G&quot;</td><td>&quot;C&quot;</td><td>0.836634</td><td>0.872589</td><td>0.0330469</td><td>0.30947</td><td>&quot;rs150926226&quot;</td><td>&quot;C&quot;</td><td>&quot;0.62&quot;</td><td>&quot;Factor VIII le…</td><td>&quot;Genome-Wide As…</td><td>&quot;30586737&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-9</td><td>8.522879</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;0.754&quot;</td><td>&quot;Low density li…</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>9.0000e-15</td><td>14.045757</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;NR&quot;</td><td>&quot;Low density li…</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>4.0000e-19</td><td>18.39794</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;0.754&quot;</td><td>&quot;Total choleste…</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>7.0000e-19</td><td>18.154902</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;NR&quot;</td><td>&quot;Total choleste…</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-22</td><td>22.0</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;0.754&quot;</td><td>&quot;Triglycerides&quot;</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>8.0000e-12</td><td>11.09691</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;T&quot;</td><td>&quot;NR&quot;</td><td>&quot;Triglycerides&quot;</td><td>&quot;A large electr…</td><td>&quot;29507422&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>2.0000e-12</td><td>11.69897</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs306890&quot;</td><td>155757485</td><td>&quot;T&quot;</td><td>&quot;C&quot;</td><td>0.638614</td><td>0.250862</td><td>0.0320938</td><td>0.316576</td><td>&quot;rs306890&quot;</td><td>&quot;C&quot;</td><td>&quot;NR&quot;</td><td>&quot;Body mass inde…</td><td>&quot;A Large Multi-…</td><td>&quot;30108127&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>6.0000e-20</td><td>19.221849</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs75303099&quot;</td><td>155954788</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.366337</td><td>0.163391</td><td>0.0664414</td><td>0.260534</td><td>&quot;rs75303099&quot;</td><td>&quot;T&quot;</td><td>&quot;0.165941&quot;</td><td>&quot;Eosinophil cou…</td><td>&quot;The Polygenic …</td><td>&quot;32888494&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>5.0000e-13</td><td>12.30103</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs75303099&quot;</td><td>155954788</td><td>&quot;C&quot;</td><td>&quot;T&quot;</td><td>0.366337</td><td>0.163391</td><td>0.0664414</td><td>0.260534</td><td>&quot;rs75303099&quot;</td><td>&quot;T&quot;</td><td>&quot;0.165937&quot;</td><td>&quot;Eosinophil per…</td><td>&quot;The Polygenic …</td><td>&quot;32888494&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-14</td><td>14.0</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs3093479&quot;</td><td>156000962</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.207921</td><td>0.0795638</td><td>0.160441</td><td>0.588111</td><td>&quot;rs3093479&quot;</td><td>&quot;G&quot;</td><td>&quot;0.062421&quot;</td><td>&quot;Eosinophil cou…</td><td>&quot;The Polygenic …</td><td>&quot;32888494&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>3.0000e-11</td><td>10.522879</td></tr><tr><td>&quot;chrX&quot;</td><td>&quot;P0000225359&quot;</td><td>156029013</td><td>156029427</td><td>&quot;Insertion&quot;</td><td>0.361386</td><td>&quot;rs3093479&quot;</td><td>156000962</td><td>&quot;A&quot;</td><td>&quot;G&quot;</td><td>0.207921</td><td>0.0795638</td><td>0.160441</td><td>0.588111</td><td>&quot;rs3093479&quot;</td><td>&quot;G&quot;</td><td>&quot;0.062412&quot;</td><td>&quot;Eosinophil per…</td><td>&quot;The Polygenic …</td><td>&quot;32888494&quot;</td><td>&quot;www.ncbi.nlm.n…</td><td>1.0000e-11</td><td>11.0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (7_714_074, 23)\n",
       "┌──────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬──────────┬────────────┐\n",
       "│ Chromoso ┆ SV_Name   ┆ SV_Start  ┆ SV_End    ┆ … ┆ Pubmed_ID ┆ Link      ┆ P-Value  ┆ P-Value_lo │\n",
       "│ me       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ ---       ┆ ---      ┆ g10        │\n",
       "│ ---      ┆ str       ┆ i64       ┆ i64       ┆   ┆ str       ┆ str       ┆ f64      ┆ ---        │\n",
       "│ str      ┆           ┆           ┆           ┆   ┆           ┆           ┆          ┆ f64        │\n",
       "╞══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪══════════╪════════════╡\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 29059683  ┆ www.ncbi. ┆ 0.000000 ┆ 6.522879   │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 3        ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /2905…    ┆          ┆            │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 29058716  ┆ www.ncbi. ┆ 6.0000e- ┆ 8.221849   │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 9        ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /2905…    ┆          ┆            │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 29058716  ┆ www.ncbi. ┆ 3.0000e- ┆ 7.522879   │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 8        ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /2905…    ┆          ┆            │\n",
       "│ chr8     ┆ P00000968 ┆ 81251     ┆ 81609     ┆ … ┆ 32139696  ┆ www.ncbi. ┆ 1.0000e- ┆ 8.0        │\n",
       "│          ┆ 43        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 8        ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /3213…    ┆          ┆            │\n",
       "│ …        ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …        ┆ …          │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 32888494  ┆ www.ncbi. ┆ 5.0000e- ┆ 12.30103   │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 13       ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /3288…    ┆          ┆            │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 32888494  ┆ www.ncbi. ┆ 1.0000e- ┆ 14.0       │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 14       ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /3288…    ┆          ┆            │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 32888494  ┆ www.ncbi. ┆ 3.0000e- ┆ 10.522879  │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 11       ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /3288…    ┆          ┆            │\n",
       "│ chrX     ┆ P00002253 ┆ 156029013 ┆ 156029427 ┆ … ┆ 32888494  ┆ www.ncbi. ┆ 1.0000e- ┆ 11.0       │\n",
       "│          ┆ 59        ┆           ┆           ┆   ┆           ┆ nlm.nih.g ┆ 11       ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ ov/pubmed ┆          ┆            │\n",
       "│          ┆           ┆           ┆           ┆   ┆           ┆ /3288…    ┆          ┆            │\n",
       "└──────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴──────────┴────────────┘"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Creation of a full dataframe to make data extraction much easier\n",
    "df_full_join = (df_sv_snp_join\n",
    "                .join(df_gwas, \n",
    "                      on = [\"Chromosome\", \n",
    "                            \"SNP_Position\"], \n",
    "                      how = \"left\")\n",
    "                .filter(pl.col(\"SV_AF\").is_not_null())\n",
    "                .with_columns(pl.when(pl.col(\"P-Value\") < 1e-50)\n",
    "                                .then(1e-50)\n",
    "                                .otherwise(pl.col(\"P-Value\"))\n",
    "                                .alias(\"P-Value\"))\n",
    "                .with_columns((-pl.col(\"P-Value\")\n",
    "                                  .log10())\n",
    "                                  .alias(\"P-Value_log10\")))\n",
    "df_full_join"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset df_anno for the SVs in df_full_foin\n",
    "sv_names = df_full_join.select(\"SV_Name\").unique()\n",
    "df_anno_subset = df_anno.filter(pl.col(\"SV_Name\")\n",
    "                                  .is_in(sv_names[\"SV_Name\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of phenotype dictionary to encompass SV for each gene - computationally faster when creating SV table\n",
    "df_phenotype = (\n",
    "    df_full_join\n",
    "    .groupby(\"Phenotype\")\n",
    "    .agg(pl.col(\"SV_Name\").unique())  # Ensure SV_Name is unique within each group\n",
    ")\n",
    "\n",
    "dict_pheno = {\n",
    "    row[0]: set(row[1]) for row in df_phenotype.iter_rows()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of y-coordinates to prevent overlap in genes when plotted\n",
    "def assign_y_coordinates(df: pl.DataFrame,\n",
    "                         y_start: float = -0.3,\n",
    "                         y_increment: float = 0.5,\n",
    "                         col_name: str = \"y_coord\") -> pl.DataFrame:\n",
    "    \n",
    "    # Priority queue to keep track of the end positions of genes for each y-coordinate\n",
    "    df_filtered = (df.filter(pl.col(\"feature\") == \"transcript\")\n",
    "                     .sort([\"chr\", \"start\"]))\n",
    "    \n",
    "    rows = df_filtered.iter_rows()\n",
    "    \n",
    "     # Dictionary to store priority queues for each chromosome\n",
    "    chromosome_heaps = {}\n",
    "\n",
    "    # List to store the y-coordinate for each gene annotation\n",
    "    y_coordinates = []\n",
    "\n",
    "    for row in rows:\n",
    "        chrom, start, end = row[0], row[1], row[2]\n",
    "\n",
    "        # Initialize heap for new chromosomes\n",
    "        if chrom not in chromosome_heaps:\n",
    "            chromosome_heaps[chrom] = []\n",
    "\n",
    "        heap = chromosome_heaps[chrom]\n",
    "\n",
    "        # Check for available y-coordinates (no overlap)\n",
    "        if heap and heap[0][0] <= start:\n",
    "            _, y_coord = heapq.heappop(heap)\n",
    "            heapq.heappush(heap, (end, y_coord))\n",
    "        else:\n",
    "            # Create a new y-coordinate if none available\n",
    "            y_coord = heap[0][1] - y_increment if heap else y_start\n",
    "            heapq.heappush(heap, (end, y_coord))\n",
    "\n",
    "        y_coordinates.append(y_coord)\n",
    "\n",
    "    # Add the y-coordinates as a new column\n",
    "    return df_filtered.with_columns(pl.Series(col_name, y_coordinates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (19_062, 8)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chr</th><th>start</th><th>end</th><th>strand</th><th>feature</th><th>gene</th><th>y_coord_nopheno</th><th>y_coord_pheno</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>65419</td><td>71585</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;OR4F5&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>450740</td><td>451678</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;OR4F29&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>685716</td><td>686654</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;OR4F16&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>923923</td><td>944574</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;SAMD11&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>944203</td><td>959256</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;NOC2L&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chr1&quot;</td><td>960584</td><td>965719</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;KLHL17&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>966482</td><td>975865</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;PLEKHN1&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chr1&quot;</td><td>975198</td><td>982093</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;PERM1&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>998964</td><td>1000097</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;HES4&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chr1&quot;</td><td>1013497</td><td>1014540</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;ISG15&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chr1&quot;</td><td>1020120</td><td>1056116</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;AGRN&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chr1&quot;</td><td>1070967</td><td>1074306</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;RNF223&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrY&quot;</td><td>22168542</td><td>22182923</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;RBMY1F&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>22403410</td><td>22417881</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;RBMY1J&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chrY&quot;</td><td>22490397</td><td>22514637</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;PRY&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>22984263</td><td>23005465</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;BPY2&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chrY&quot;</td><td>23129355</td><td>23199008</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;DAZ1&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>23219533</td><td>23291356</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;DAZ2&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chrY&quot;</td><td>24045793</td><td>24048019</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;CDY1B&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>24618004</td><td>24639207</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;BPY2B&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chrY&quot;</td><td>24763069</td><td>24813393</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;DAZ3&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>24833919</td><td>24907040</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;DAZ4&quot;</td><td>-0.3</td><td>-6.25</td></tr><tr><td>&quot;chrY&quot;</td><td>25030901</td><td>25052104</td><td>&quot;-&quot;</td><td>&quot;transcript&quot;</td><td>&quot;BPY2C&quot;</td><td>-0.45</td><td>-10.25</td></tr><tr><td>&quot;chrY&quot;</td><td>25622117</td><td>25624902</td><td>&quot;+&quot;</td><td>&quot;transcript&quot;</td><td>&quot;CDY1&quot;</td><td>-0.3</td><td>-6.25</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (19_062, 8)\n",
       "┌──────┬──────────┬──────────┬────────┬────────────┬────────┬─────────────────┬───────────────┐\n",
       "│ chr  ┆ start    ┆ end      ┆ strand ┆ feature    ┆ gene   ┆ y_coord_nopheno ┆ y_coord_pheno │\n",
       "│ ---  ┆ ---      ┆ ---      ┆ ---    ┆ ---        ┆ ---    ┆ ---             ┆ ---           │\n",
       "│ str  ┆ i64      ┆ i64      ┆ str    ┆ str        ┆ str    ┆ f64             ┆ f64           │\n",
       "╞══════╪══════════╪══════════╪════════╪════════════╪════════╪═════════════════╪═══════════════╡\n",
       "│ chr1 ┆ 65419    ┆ 71585    ┆ +      ┆ transcript ┆ OR4F5  ┆ -0.3            ┆ -6.25         │\n",
       "│ chr1 ┆ 450740   ┆ 451678   ┆ -      ┆ transcript ┆ OR4F29 ┆ -0.3            ┆ -6.25         │\n",
       "│ chr1 ┆ 685716   ┆ 686654   ┆ -      ┆ transcript ┆ OR4F16 ┆ -0.3            ┆ -6.25         │\n",
       "│ chr1 ┆ 923923   ┆ 944574   ┆ +      ┆ transcript ┆ SAMD11 ┆ -0.3            ┆ -6.25         │\n",
       "│ …    ┆ …        ┆ …        ┆ …      ┆ …          ┆ …      ┆ …               ┆ …             │\n",
       "│ chrY ┆ 24763069 ┆ 24813393 ┆ -      ┆ transcript ┆ DAZ3   ┆ -0.45           ┆ -10.25        │\n",
       "│ chrY ┆ 24833919 ┆ 24907040 ┆ +      ┆ transcript ┆ DAZ4   ┆ -0.3            ┆ -6.25         │\n",
       "│ chrY ┆ 25030901 ┆ 25052104 ┆ -      ┆ transcript ┆ BPY2C  ┆ -0.45           ┆ -10.25        │\n",
       "│ chrY ┆ 25622117 ┆ 25624902 ┆ +      ┆ transcript ┆ CDY1   ┆ -0.3            ┆ -6.25         │\n",
       "└──────┴──────────┴──────────┴────────┴────────────┴────────┴─────────────────┴───────────────┘"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_transcript = assign_y_coordinates(df = df_gene,\n",
    "                                     y_start = -0.3,\n",
    "                                     y_increment = 0.15,\n",
    "                                     col_name = \"y_coord_nopheno\")\n",
    "\n",
    "df_transcript = assign_y_coordinates(df = df_transcript,\n",
    "                                     y_start = -6.25,\n",
    "                                     y_increment = 4,\n",
    "                                     col_name = \"y_coord_pheno\")\n",
    "df_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add icon for help popup\n",
    "question_icon = html.I(className = 'bi bi-question-diamond',\n",
    "                       style = {\"display\" : \"inline-block\",\n",
    "                                \"color\" : \"lightsalmon\",\n",
    "                                \"font-size\" : \"70%\"})\n",
    "                    #    id = 'question-icon')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### FILTERS ####\n",
    "\n",
    "# Make dropdown box for chromosomes and textbox for genomic range\n",
    "REGION_LABEL = html.P(\"Genomic Region\")\n",
    "\n",
    "chromosomes = [\"Any\", *range(1, 23), \"X\", \"Y\"]\n",
    "\n",
    "CHR_DD = dcc.Dropdown(id = 'chromosome-dropdown',\n",
    "                      placeholder = \"Select a chromosome\",\n",
    "                      options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in chromosomes])\n",
    "\n",
    "\n",
    "EMPTY_SPACE_REGION = html.Div(id = 'empty-space-region')\n",
    "\n",
    "RANGE_START_TB = dcc.Input(id = 'range-start-textbox',\n",
    "                           type = \"number\",\n",
    "                           placeholder = \"Start bp\",\n",
    "                           min = 1,\n",
    "                           debounce = True,\n",
    "                           className = 'class-range')\n",
    "\n",
    "RANGE_END_TB = dcc.Input(id = 'range-end-textbox',\n",
    "                         type = \"number\",\n",
    "                         placeholder = \"End bp\",\n",
    "                         min = 2,\n",
    "                         debounce = True,\n",
    "                         className = 'class-range')\n",
    "\n",
    "\n",
    "REGION_DIV = html.Div(id = 'chromosome-dropdown-div',\n",
    "                      children = [REGION_LABEL,\n",
    "                                  CHR_DD,\n",
    "                                  EMPTY_SPACE_REGION,\n",
    "                                  RANGE_START_TB,\n",
    "                                  RANGE_END_TB],\n",
    "                      className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for phenotypes\n",
    "# PHENO_LINE = html.Div(id = 'pheno-line-div')\n",
    "\n",
    "PHENO_LABEL = html.Div([html.P([\"Phenotype\",\n",
    "                                question_icon],\n",
    "                               id = 'pheno-label'),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Select phenotype of interest availible in GWAS Catalog.\\n\\nSVs are selected if they have an LD calculation with an associated SNP for the phenotype.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'pheno-label',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'pheno-label-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes = sorted(df_full_join.select(\"Phenotype\")\n",
    "                                .unique()\n",
    "                                .to_series()\n",
    "                                .to_list())\n",
    "phenotypes.insert(0, \"Any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "PHENO_DD = dcc.Dropdown(id = 'phenotype-dropdown',\n",
    "                        placeholder = \"Select a phenotype of interest\",\n",
    "                        options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in phenotypes],\n",
    "                        optionHeight = 75)\n",
    "\n",
    "PHENO_DD_DIV = html.Div(id = 'phenotype-dropdown-div',\n",
    "                        children = [# PHENO_LINE,\n",
    "                                    PHENO_LABEL,\n",
    "                                    PHENO_DD],\n",
    "                        className = 'class-section-filter-div')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for genes\n",
    "GENE_LABEL = html.P(\"Gene\")\n",
    "\n",
    "genes = sorted(list(set(df_gene[\"gene\"])))\n",
    "\n",
    "GENE_DD = dcc.Dropdown(id = 'gene-dropdown',\n",
    "                       placeholder = \"Select a gene of interest\",\n",
    "                       options = [{\"label\" : str(i),\n",
    "                                   \"value\" : str(i)} for i in genes])\n",
    "\n",
    "GENE_DD_DIV = html.Div(id = 'gene-dropdown-div',\n",
    "                       children = [GENE_LABEL,\n",
    "                                   GENE_DD],\n",
    "                       className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create a tab section for filter between gene and genomic region\n",
    "TABS_FILTER_DIV = html.Div([dbc.Tabs([dbc.Tab(children = [GENE_DD_DIV],\n",
    "                                         label = \"by Gene\",\n",
    "                                         tab_id = \"tab-gene\"),\n",
    "                                     dbc.Tab(children = [REGION_DIV],\n",
    "                                             label = \"by Region\",\n",
    "                                             tab_id = \"tab-region\")],\n",
    "                                     id = \"tab-filters\",\n",
    "                                     active_tab = \"tab-gene\")],\n",
    "                           className = 'class-section-filter-div')\n",
    "\n",
    "\n",
    "FILTER_DIV = html.Div(id = 'filter-div',\n",
    "                      children = [TABS_FILTER_DIV,\n",
    "                                  PHENO_DD_DIV],\n",
    "                      className = 'class-filter-div')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### TABLE ####\n",
    "\n",
    "# Make DataTable\n",
    "def make_table(chrom = chromosomes,\n",
    "               range_start = 0,\n",
    "               range_end = 250000000,\n",
    "               gene = genes,\n",
    "               pheno = phenotypes):\n",
    "\n",
    "    global tab_show\n",
    "\n",
    "    # Filter by chromosome\n",
    "    if isinstance(chrom, list):\n",
    "        sv_list_chr = df_anno_subset[\"SV_Name\"].to_list()\n",
    "    else:\n",
    "        chrom = f\"chr{chrom}\"\n",
    "        sv_list_chr = df_anno_subset.filter(pl.col(\"Chromosome\") == chrom)[\"SV_Name\"].to_list()\n",
    "\n",
    "    # Filter by range\n",
    "    sv_list_range = df_anno_subset.filter((pl.col(\"SV_Start\") >= range_start) & \\\n",
    "                                          (pl.col(\"SV_End\") <= range_end))[\"SV_Name\"].to_list()\n",
    "\n",
    "    # Filter by phenotype\n",
    "    if isinstance(pheno, list):\n",
    "        sv_list_pheno = df_anno_subset[\"SV_Name\"].to_list()\n",
    "    else:\n",
    "        sv_list_pheno = dict_pheno[pheno]\n",
    "\n",
    "    # Filter by gene\n",
    "    if isinstance(gene, list):\n",
    "        sv_list_gene = list(df_anno_subset[\"SV_Name\"])\n",
    "    else:\n",
    "        df_1gene = df_gene.filter((pl.col(\"feature\") == \"transcript\") & \\\n",
    "                                  (pl.col(\"gene\") == gene))\n",
    "        start = df_1gene[\"start\"].min()\n",
    "        end = df_1gene[\"end\"].max()\n",
    "        chrom = df_1gene.select(\"chr\").to_series()[0]\n",
    "        \n",
    "        sv_list_gene = df_sv_anno.filter((pl.col(\"Chromosome\") == chrom) & \\\n",
    "                                         (pl.col(\"Start\") >= (start - 100000)) & \\\n",
    "                                         (pl.col(\"End\") <= (end + 100000)))[\"SV Name\"].to_list()\n",
    "\n",
    "    # Find overlaps sv's from all lists\n",
    "    sv_list_complete = set(sv_list_pheno).intersection(sv_list_chr, sv_list_gene, sv_list_range)\n",
    "\n",
    "    tab_show = df_sv_anno.filter(pl.col(\"SV Name\").is_in(sv_list_complete)).select([\n",
    "        pl.col(\"SV Name\").alias(\"ID\"),\n",
    "        pl.col(\"Chromosome\").alias(\"Chrom\"),\n",
    "        pl.col(\"Start\"),\n",
    "        pl.col(\"End\"),\n",
    "        pl.col(\"Type\"),\n",
    "        pl.col(\"Length\").alias(\"Size (bp)\"),\n",
    "        pl.col(\"SV Sample AF\").alias(\"AF\")])\n",
    "\n",
    "    return dash_table.DataTable(id ='strvar-table',\n",
    "                                data = tab_show.to_dicts(),\n",
    "                                columns = [{'id': c, 'name': c} for c in [\"Chrom\", \"Start\", \"End\", \"Type\", \"Size (bp)\", \"AF\"]],\n",
    "                                page_size = 100,\n",
    "                                fixed_rows = {'headers': True},\n",
    "                                style_table = {'height': '400px',\n",
    "                                               'overflowY': 'auto'},\n",
    "                                style_cell = {'textAlign': 'center',\n",
    "                                              'font-family': 'Nunito Sans',\n",
    "                                              'backgroundColor': 'whitesmoke',\n",
    "                                              'minWidth' : '100px'},\n",
    "                                style_header = {'backgroundColor': 'turquoise',\n",
    "                                                'fontWeight': 'bold'},\n",
    "                                row_selectable = 'single',\n",
    "                                selected_rows = [])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SV annotation table\n",
    "def make_annotation_table(sv = \"None\"):\n",
    "    if sv == \"None\":\n",
    "        df_anno_subset_sv = pl.DataFrame()\n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = [],\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_anno_subset_sv = df_sv_anno.filter(pl.col(\"SV Name\") == sv)\n",
    "        df_anno_subset_sv = (df_anno_subset_sv.melt()\n",
    "                                              .rename({\"variable\": \"Header\",\n",
    "                                                       \"value\": \"Information\"}))\n",
    "        \n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = df_anno_subset_sv.to_dicts(),\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowX': 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell = {'textAlign': 'left',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': \"Header\"},\n",
    "                                                              'backgroundColor': 'turquoise',\n",
    "                                                              'fontWeight': 'bold',\n",
    "                                                              'maxWidth' : '200px',\n",
    "                                                              'whiteSpace' : 'normal',\n",
    "                                                              'height' : 'auto'}],\n",
    "                                    style_header = {'display' : 'none',\n",
    "                                                    'height' : '0px'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "DTABLE = html.Div(id = 'strvar-table-div',\n",
    "                  children = [make_table()],\n",
    "                  className = 'sv-table')\n",
    "\n",
    "# Create a header for the SV table\n",
    "DTABLE_HEADER = html.Div([html.H4([\"SV Information\",\n",
    "                                   question_icon],\n",
    "                                   id = 'dtable-header',\n",
    "                                   style = {'textAlign' : 'center',\n",
    "                                            'color' : '#2F2E2D',\n",
    "                                            'height' : '30 px',\n",
    "                                            'text-transform': 'none'}),\n",
    "                          dbc.Tooltip(dcc.Markdown(\"List of SVs subsetted by the filters applied.\\n\\nSelect an SV to visualize LD with GWAS-associated SNPs.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                      target = 'dtable-header',\n",
    "                                      placement = \"auto\",\n",
    "                                      delay = {\"show\" : 500,\n",
    "                                               \"hide\" : 50},\n",
    "                                      id = 'dtable-header-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "DTABLE_TEXTBOX = html.Div(id = 'dtable-textbox')\n",
    "\n",
    "\n",
    "DTABLE_DIV = html.Div(id = 'dtable-div',\n",
    "                      children = [DTABLE_HEADER,\n",
    "                                  DTABLE,\n",
    "                                  DTABLE_TEXTBOX],\n",
    "                      className = 'class-filter-div')\n",
    "\n",
    "\n",
    "ANNOTABLE = html.Div(id = 'anno-table-div',\n",
    "                     children = [make_annotation_table()],\n",
    "                     className = 'sv-anno-table')\n",
    "\n",
    "\n",
    "# Create the SV details collapse and table\n",
    "SV_TABLE_BUTTON = dbc.Button(\"SV Annotations\",\n",
    "                             id = \"sv-button-collapse\",\n",
    "                             color = \"secondary\",\n",
    "                             outline = True,\n",
    "                             size = \"sm\",\n",
    "                             n_clicks = 0)\n",
    "\n",
    "\n",
    "# Create the filter division\n",
    "ALL_FILTER_DIV = html.Div(id = 'all-filter-div',\n",
    "                          children = [FILTER_DIV,\n",
    "                                      DTABLE_DIV],\n",
    "                                      style = {'display' : 'flex',\n",
    "                                               'flexWrap' : \"wrap\",\n",
    "                                               'justifyContent' : \"space-evenly\",\n",
    "                                               'alignContent' : 'space-around'},\n",
    "                                               className = 'sections')\n",
    "\n",
    "SV_TABLE_BUTTON_DIV = html.Div(id = \"sv-button-div\",\n",
    "                               children = [SV_TABLE_BUTTON])\n",
    "\n",
    "\n",
    "SV_TABLE_COLLAPSE = dbc.Collapse(children = [ANNOTABLE],\n",
    "                                 id = \"sv-table-collapse\",\n",
    "                                 is_open = False)\n",
    "\n",
    "SV_TABLE = html.Div(id = 'sv-table-collapse-div',\n",
    "                    children = [SV_TABLE_COLLAPSE],\n",
    "                    style = {'display' : 'flex',\n",
    "                             'flexWrap' : \"wrap\",\n",
    "                             'justifyContent' : \"space-evenly\",\n",
    "                             'alignContent' : 'space-around'})\n",
    "\n",
    "\n",
    "SV_TABLE_DIV = html.Div(id = 'sv-table-div',\n",
    "                        children = [SV_TABLE,\n",
    "                                    SV_TABLE_BUTTON_DIV],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### PLOT ####\n",
    "\n",
    "# Make scatter plot to compare r2 and D' values as seen in DataTable\n",
    "def make_plot(sv = \"None\",\n",
    "              pheno = \"None\",\n",
    "              toggle = False,\n",
    "              toggle_pheno = False):\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage_value = \"D'\"\n",
    "        size_value = \"R2\"\n",
    "        legend_title = \"D'\"\n",
    "    else:\n",
    "        linkage_value = \"R2\"\n",
    "        size_value = \"D'\"\n",
    "        legend_title = \"r<sup>2</sup>\"\n",
    "\n",
    "    global tab, sv_start_plot, sv_end_plot, extra_points, extra_points_pheno\n",
    "\n",
    "    if sv == \"None\":\n",
    "        tab = pl.DataFrame({\"SNP_Position\": [],\n",
    "                            \"-log10(P-Value)\": []})\n",
    "        \n",
    "        sc_plot = px.scatter(data_frame = tab.to_pandas(),\n",
    "                             x = \"SNP_Position\",\n",
    "                             y = \"-log10(P-Value)\",\n",
    "                             labels = {\"-log10(P-Value)\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                       \"SNP_Position\" : \"Position in hg38 (bp)\"},\n",
    "                             width = 850,\n",
    "                             height = 650,\n",
    "                            #  marginal_x = \"rug\",\n",
    "                             template = \"ggplot2\")\n",
    "\n",
    "        sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                              autosize = False,\n",
    "                              font_family = \"Nunito Sans\",\n",
    "                              font_size = 20)\n",
    "    else:\n",
    "        tab = df_full_join.filter(pl.col(\"SV_Name\") == sv)\n",
    "\n",
    "        # Display +/- 1KB from SV position\n",
    "        sv_start = df_anno_subset.filter(pl.col(\"SV_Name\") == sv).select(\"SV_Start\").to_numpy()[0, 0]\n",
    "        sv_end = df_anno_subset.filter(pl.col(\"SV_Name\") == sv).select(\"SV_End\").to_numpy()[0, 0]\n",
    "\n",
    "        sv_start_plot = sv_start - 1000\n",
    "        sv_end_plot = sv_end + 1000\n",
    "\n",
    "\n",
    "        if pheno == \"None\":\n",
    "            tab_lab = tab.to_dicts()\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"],\n",
    "                        row[\"SV_Name\"],\n",
    "                        row[\"SV_Start\"],\n",
    "                        row[\"SV_End\"],\n",
    "                        row[\"SV_Type\"]]\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab.to_pandas(),\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = linkage_value,\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"R2\" : \"r<sup>2</sup>\"},\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                #  marginal_x = \"rug\",\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(marker = dict(color = \"rgba(57,201,187,0.5)\",\n",
    "                                                line = dict(color = \"rgba(57,201,187,1)\"),\n",
    "                                                size = 10),\n",
    "                                  hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                  '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                  '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                  '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                  '<br>P-Value: %{customdata[4]}' + \n",
    "                                                  '<br>r2: %{customdata[5]}' + \n",
    "                                                  \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab])\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                  autosize = False,\n",
    "                                  font_family = \"Nunito Sans\",\n",
    "                                  font_size = 20)\n",
    "                                #   xaxis_range = [sv_start_plot, sv_end_plot],\n",
    "                                #   yaxis_range = [-0.6, 1])\n",
    "\n",
    "             # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = tab.select(\"SNP_Position\").min().item() - 100000\n",
    "            max_SNP = tab.select(\"SNP_Position\").max().item() + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = tab.select(\"SNP_Position\").min()[0, 0] - 100000\n",
    "                max_SNP = tab.select(\"SNP_Position\").max()[0, 0] + 100000\n",
    "\n",
    "\n",
    "            chromo = tab.select(\"Chromosome\").unique().item()\n",
    "            svname = tab.select(\"SV_Name\").unique().item()\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -0.1,\n",
    "                              y1 = -0.15,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                           width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.01,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-0.07]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + \n",
    "                                                         '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                         '<br>Start: %{customdata[8]}' + \n",
    "                                                         '<br>End: %{customdata[9]}' + \n",
    "                                                         '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Add extra points from df_sv with R2 and D' values\n",
    "            extra_points = df_sv_snp_join.filter(pl.col(\"SV_Name\") == sv)\n",
    "\n",
    "            extra_points = extra_points.filter(~pl.col(\"SNP_Name_dbSNP\").is_in(tab.select(\"SNP_Name_dbSNP\").to_series()))\n",
    "\n",
    "             # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dicts()\n",
    "\n",
    "            def extra_label_maker(row):\n",
    "\n",
    "                return [row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"]]\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                           y = extra_points.select(linkage_value).to_series().to_list(),\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>SNP Name: %{customdata[0]}</b>' +\n",
    "                                                           '<br>Chromosome: %{customdata[1]}' + \n",
    "                                                           '<br>SNP Position: %{customdata[2]}' + \n",
    "                                                           '<br>r2: %{customdata[3]}' + \n",
    "                                                           \"<br>D': %{customdata[4]}\",\n",
    "                                           customdata = [extra_label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"Not in \\nGWAS Catalog\"))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene.filter((pl.col(\"chr\") == chromo) &\n",
    "                                   (pl.col(\"start\") >= min_SNP) &\n",
    "                                   (pl.col(\"end\") <= max_SNP))\n",
    "\n",
    "            exons_genes = (exons.groupby([\"chr\", \"gene\", \"strand\"])\n",
    "                                .agg([pl.col(\"start\").min().alias(\"start\"),\n",
    "                                      pl.col(\"end\").max().alias(\"end\")]))\n",
    "            \n",
    "            # Compute `gene_line_position` for each row in `exons_genes`\n",
    "            exons_genes = exons_genes.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene).select(\"y_coord_nopheno\").item()))\n",
    "                                     .alias(\"gene_line_position\"))\n",
    "\n",
    "\n",
    "            for row in exons_genes.iter_rows(named = True):\n",
    "\n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = row[\"gene_line_position\"],\n",
    "                                  y1 = row[\"gene_line_position\"],\n",
    "                                  line_width = 3,\n",
    "                                  line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons.filter(pl.col(\"feature\") == \"exon\")\n",
    "\n",
    "            # Draw exons in the plot\n",
    "            exons_exons.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene)\n",
    "                                                                                      .select(\"y_coord_nopheno\").item()))\n",
    "                                     .alias(\"gene_line_position\")).apply(lambda row: sc_plot.add_shape(type = \"rect\",\n",
    "                                                                                                      x0 = row[\"start\"],\n",
    "                                                                                                      x1 = row[\"end\"],\n",
    "                                                                                                      y0 = row[\"gene_line_position\"] + 0.020,\n",
    "                                                                                                      y1 = row[\"gene_line_position\"] - 0.020,\n",
    "                                                                                                      line = dict(color = \"mediumseagreen\", \n",
    "                                                                                                                  width=2),\n",
    "                                                                                                      fillcolor = \"mediumseagreen\"))\n",
    " \n",
    "\n",
    "        else:\n",
    "            tab = tab.filter(pl.col(\"Phenotype\") == pheno)\n",
    "\n",
    "            tab_lab = tab.to_dicts()\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"],\n",
    "                        row[\"SV_Name\"],\n",
    "                        row[\"SV_Start\"],\n",
    "                        row[\"SV_End\"],\n",
    "                        row[\"SV_Type\"]]\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab,\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = \"P-Value_log10\",\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"P-Value_log10\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                           linkage_value : legend_title},\n",
    "                                 color = linkage_value,\n",
    "                                 color_continuous_scale = [(0, \"#4643CE\"),\n",
    "                                                           (0.2, \"#4643CE\"),\n",
    "                                                           (0.2, \"#90CAEE\"),\n",
    "                                                           (0.4, \"#90CAEE\"),\n",
    "                                                           (0.4, \"#46E17C\"),\n",
    "                                                           (0.6, \"#46E17C\"),\n",
    "                                                           (0.6, \"#EEC06B\"),\n",
    "                                                           (0.8, \"#EEC06B\"),\n",
    "                                                           (0.8, \"#CE5C5C\"),\n",
    "                                                           (1, \"#CE5C5C\")],\n",
    "                                 range_color = [0, 1],\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                  '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                  '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                  '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                  '<br>P-Value: %{customdata[4]}' + \n",
    "                                                  '<br>r2: %{customdata[5]}' + \n",
    "                                                  \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab],\n",
    "                                  marker = dict(size = 10))\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                 autosize = True,\n",
    "                                 font_family = \"Nunito Sans\",\n",
    "                                 font_size = 20,\n",
    "                                 coloraxis_colorbar = dict(lenmode = \"pixels\",\n",
    "                                                           len = 150,\n",
    "                                                           yanchor = \"top\",\n",
    "                                                           y = 1,\n",
    "                                                           dtick = 0.2))\n",
    "\n",
    "            # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = tab.select(\"SNP_Position\").min().item() - 100000\n",
    "            max_SNP = tab.select(\"SNP_Position\").max().item() + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = tab.select(\"SNP_Position\").min()[0, 0] - 100000\n",
    "                max_SNP = tab.select(\"SNP_Position\").max()[0, 0] + 100000\n",
    "\n",
    "\n",
    "            chromo = tab.select(\"Chromosome\").unique().item()\n",
    "            svname = tab.select(\"SV_Name\").unique().item()\n",
    "\n",
    "            # Add points from gwas for pheno, without r2 and D' values\n",
    "            extra_points = (df_gwas.filter((pl.col(\"Chromosome\") == chromo) &\n",
    "                                           (pl.col(\"SNP_Position\") >= min_SNP) &\n",
    "                                           (pl.col(\"SNP_Position\") <= max_SNP) &\n",
    "                                           (pl.col(\"Phenotype\") == pheno))\n",
    "                                    .filter(~pl.col(\"SNP_Name_GWAS\").is_in(tab.select(\"SNP_Name_GWAS\").to_series())))\n",
    "\n",
    "            extra_points = (extra_points.with_columns(pl.when(pl.col(\"P-Value\") < 1e-50)\n",
    "                                                        .then(1e-50)\n",
    "                                                        .otherwise(pl.col(\"P-Value\"))\n",
    "                                                        .alias(\"P-Value\"),\n",
    "                                                        (-pl.col(\"P-Value\").log10()).alias(\"P-Value_log10\")))\n",
    "\n",
    "            # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dicts()\n",
    "\n",
    "            def label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_GWAS\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"]]\n",
    "\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                           y = extra_points.select(\"P-Value_log10\").to_series().to_list(),\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                           '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                           '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                           '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                           '<br>P-Value: %{customdata[4]}',\n",
    "                                           customdata = [label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"No LD Data\"))\n",
    "\n",
    "            # Add points from other phenos in LD with SV\n",
    "            if toggle_pheno == True:\n",
    "                extra_points_pheno = (df_full_join.filter((pl.col(\"Chromosome\") == chromo) &\n",
    "                                                          (pl.col(\"SNP_Position\") >= min_SNP) &\n",
    "                                                          (pl.col(\"SNP_Position\") <= max_SNP) &\n",
    "                                                          (pl.col(\"SV_Name\") == sv))\n",
    "                                                  .filter(~pl.col(\"SNP_Name_GWAS\").is_in(tab.select(\"SNP_Name_GWAS\").to_series())))\n",
    "\n",
    "\n",
    "                # Create a list of labels for hovertext\n",
    "                df_lab_2 = extra_points_pheno.to_dicts()\n",
    "\n",
    "                def label_maker(row):\n",
    "\n",
    "                    return [row[\"Phenotype\"],\n",
    "                            row[\"SNP_Name_GWAS\"],\n",
    "                            row[\"Chromosome\"],\n",
    "                            row[\"SNP_Position\"],\n",
    "                            row[\"P-Value\"],\n",
    "                            row[\"R2\"],\n",
    "                            row[\"D'\"]]\n",
    "\n",
    "\n",
    "                sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                               y = extra_points.select(\"P-Value_log10\").to_series().to_list(),\n",
    "                                               mode = \"markers\",\n",
    "                                               marker = dict(size = 10,\n",
    "                                                             color = \"rgba(168,168,168,0.5)\",\n",
    "                                                             line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                               hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                               '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                               '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                               '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                               '<br>P-Value: %{customdata[4]}' + \n",
    "                                                               '<br>r2: %{customdata[5]}' + \n",
    "                                                               \"<br>D': %{customdata[6]}\",\n",
    "                                               customdata = [label_maker(row) for row in df_lab_2],\n",
    "                                               showlegend = False,\n",
    "                                               name = \"Other Phenotypes\"))\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -2,\n",
    "                              y1 = -3,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                          width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.5,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-1.5]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + \n",
    "                                                         '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                         '<br>Start: %{customdata[8]}' + \n",
    "                                                         '<br>End: %{customdata[9]}' + \n",
    "                                                         '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene.filter((pl.col(\"chr\") == chromo) &\n",
    "                                   (pl.col(\"start\") >= min_SNP) &\n",
    "                                   (pl.col(\"end\") <= max_SNP))\n",
    "\n",
    "            exons_genes = (exons.groupby([\"chr\", \"gene\", \"strand\"])\n",
    "                                .agg([pl.col(\"start\").min().alias(\"start\"),\n",
    "                                      pl.col(\"end\").max().alias(\"end\")]))\n",
    "            \n",
    "            # Compute `gene_line_position` for each row in `exons_genes`\n",
    "            exons_genes = exons_genes.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene).select(\"y_coord_pheno\").item()))\n",
    "                                     .alias(\"gene_line_position\"))\n",
    "\n",
    "\n",
    "            for row in exons_genes.iter_rows(named = True):\n",
    "\n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = row[\"gene_line_position\"],\n",
    "                                  y1 = row[\"gene_line_position\"],\n",
    "                                  line_width = 3,\n",
    "                                  line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] +  1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons.filter(pl.col(\"feature\") == \"exon\")\n",
    "\n",
    "            # Draw exons in the plot\n",
    "            exons_exons.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene)\n",
    "                                                                                      .select(\"y_coord_pheno\").item()))\n",
    "                                     .alias(\"gene_line_position\")).apply(lambda row: sc_plot.add_shape(type = \"rect\",\n",
    "                                                                                                      x0 = row[\"start\"],\n",
    "                                                                                                      x1 = row[\"end\"],\n",
    "                                                                                                      y0 = row[\"gene_line_position\"] + 0.75,\n",
    "                                                                                                      y1 = row[\"gene_line_position\"] - 0.75,\n",
    "                                                                                                      line = dict(color = \"mediumseagreen\", \n",
    "                                                                                                                  width=2),\n",
    "                                                                                                      fillcolor = \"mediumseagreen\"))\n",
    "\n",
    "\n",
    "    return sc_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SCATPLOT = dcc.Loading(id = \"plot-loading\",\n",
    "                       children = [dcc.Graph(id = \"scatter-plot\",\n",
    "                                             figure = make_plot(),\n",
    "                                             config = {\"modeBarButtonsToRemove\": [\"select2d\",\n",
    "                                                                                  \"lasso2d\",\n",
    "                                                                                  \"zoomIn2d\",\n",
    "                                                                                  \"zoomOut2d\"],\n",
    "                                                        \"toImageButtonOptions\": {\"format\": \"svg\",\n",
    "                                                                                 \"filename\": \"gwas_svatalog_plot\",\n",
    "                                                                                 \"height\": 500,\n",
    "                                                                                 \"width\": 700,\n",
    "                                                                                 \"scale\": 1}})],\n",
    "                       type = \"dot\",\n",
    "                       color = \"turquoise\",\n",
    "                       parent_className = \"loading_wrapper\")\n",
    "\n",
    "\n",
    "TOGGLE_SWITCH = html.Div([dbc.Row([dbc.Col([\"D'\"],\n",
    "                                           id = 'toggle-d'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'toggle-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"r\", html.Sup(2,\n",
    "                                                          style = {'font-size' : '15px'})],\n",
    "                                            id = 'toggle-r')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region'),\n",
    "                          dbc.Row([html.P([\"Show Other Phenotypes\",\n",
    "                                           question_icon],\n",
    "                                          id = 'toggle-pheno-header'),\n",
    "                                   dbc.Tooltip(dcc.Markdown(\"Show GWAS-associated SNPs from other phenotypes in respect to selected SV.\",\n",
    "                                                            style = {'white-space':'pre-wrap'}),\n",
    "                                               target = 'toggle-pheno-header',\n",
    "                                               placement = \"auto\",\n",
    "                                               delay = {\"show\" : 500,\n",
    "                                                        \"hide\" : 50},\n",
    "                                               id = 'toggle-pheno-header-tooltip')],\n",
    "                                  id = 'toggle-pheno-header-row'),\n",
    "                          dbc.Row([dbc.Col([\"OFF\"],\n",
    "                                           id = 'toggle-off'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'other-pheno-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"ON\"],\n",
    "                                            id = 'toggle-on')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region-pheno'),\n",
    "                          dbc.Row([dbc.Button(\"Export SNP Data to CSV\",\n",
    "                                               id = 'download-button',\n",
    "                                               color = \"secondary\",\n",
    "                                               outline = True,\n",
    "                                               size = \"sm\",),\n",
    "                                   dcc.Download(id = 'download-csv')],\n",
    "                                   id = 'download-region')])\n",
    "\n",
    "\n",
    "\n",
    "SCATPLOT_DIV = html.Div(id = 'scatter-plot-div',\n",
    "                        children = [SCATPLOT,\n",
    "                                    TOGGLE_SWITCH],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SNP table from GWAS data based on clickData from plot\n",
    "def make_SNP_table(snp = \"None\"):\n",
    "    if snp == \"None\":\n",
    "        df_snp = pl.DataFrame()\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = [],\n",
    "                                    columns = [],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '500px', 'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_snp = (df_full_join.filter((pl.col(\"SNP_Name_dbSNP\") == snp) | (pl.col(\"SNP_Name_GWAS\") == snp))\n",
    "                              .with_columns([\n",
    "                pl.format(\"[{}](https://{})\", pl.col(\"Pubmed_ID\"), pl.col(\"Link\")).alias(\"Link\"),\n",
    "                pl.format(\"[{}](https://www.ncbi.nlm.nih.gov/snp/{})\", pl.col(\"SNP_Name_dbSNP\"), pl.col(\"SNP_Name_dbSNP\")).alias(\"SNP_Name_dbSNP\"),\n",
    "                pl.format(\"[{}](https://www.ebi.ac.uk/gwas/variants/{})\", pl.col(\"SNP_Name_GWAS\"), pl.col(\"SNP_Name_GWAS\")).alias(\"SNP_Name_GWAS\")])\n",
    "                              .select([\n",
    "                pl.col(\"Chromosome\").alias(\"Chrom\"),\n",
    "                pl.col(\"SNP_Position\").alias(\"SNP Position\"),\n",
    "                pl.col(\"SNP_Name_dbSNP\").alias(\"SNP Name: dbSNP\"),\n",
    "                pl.col(\"SNP_Name_GWAS\").alias(\"SNP Name: GWAS\"),\n",
    "                pl.col(\"Reference_Allele\").alias(\"Reference Allele\"),\n",
    "                pl.col(\"Alternate_Allele\").alias(\"Alternate Allele\"),\n",
    "                pl.col(\"Risk_Allele\").alias(\"Risk Allele\"),\n",
    "                pl.col(\"Risk_Allele_Frequency\").alias(\"Risk AF\"),\n",
    "                pl.col(\"Sample_AF\").alias(\"Sample AF\"),\n",
    "                pl.col(\"gnomAD_nfe_AF\").alias(\"gnomAD NFE AF\"),\n",
    "                pl.col(\"Phenotype\"),\n",
    "                pl.col(\"P-Value\"),\n",
    "                pl.col(\"Study\"),\n",
    "                pl.col(\"Link\").alias(\"Pubmed Link\")])\n",
    "                               .unique())\n",
    "\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = df_snp.to_dicts(),\n",
    "                                    columns = [{'id': c, 'name': c, 'presentation': 'markdown'} \n",
    "                                               if (c == \"Pubmed Link\" or c == \"SNP Name: dbSNP\" or c == \"SNP Name: GWAS\") \n",
    "                                               else ({'id': c, 'name': c, 'type':'numeric', 'format': {'specifier': '.3f'}} \n",
    "                                                     if (c == \"Risk AF\" or c == \"Sample AF\" or c == \"gnomAD NFE AF\") \n",
    "                                                     else {'id': c, 'name': c}) for c in df_snp.columns],\n",
    "                                    markdown_options = {\"html\" : True,\n",
    "                                                        \"link_target\": \"_blank\"},\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '500px',\n",
    "                                                   'minWidth': '100%',\n",
    "                                                   'width' : '100%',\n",
    "                                                   'maxWidth' : '100%',\n",
    "                                                   'overflowX' : 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': [\"Chrom\",\n",
    "                                                                                    \"SNP Position\",\n",
    "                                                                                    \"SNP Name: dbSNP\",\n",
    "                                                                                    \"SNP Name: GWAS\",\n",
    "                                                                                    \"Reference Allele\",\n",
    "                                                                                    \"Alternate Allele\",\n",
    "                                                                                    \"Risk Allele\",\n",
    "                                                                                    \"Risk AF\",\n",
    "                                                                                    \"Sample AF\",\n",
    "                                                                                    \"gnomAD NFE AF\",\n",
    "                                                                                    \"P-Value\",\n",
    "                                                                                    \"Pubmed Link\"]},\n",
    "                                                               'minWidth' : '100px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke'},\n",
    "                                                               {'if': {'column_id': [\"Study\",\n",
    "                                                                                     \"Phenotype\"]},\n",
    "                                                               'maxWidth' : '200px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke',\n",
    "                                                               'whiteSpace' : 'normal',\n",
    "                                                               'height' : 'auto'}],\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold',\n",
    "                                                    'textAlign' : 'left',\n",
    "                                                    'whiteSpace' : 'normal',\n",
    "                                                    'height' : 'auto'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SNP_TABLE = html.Div([html.Div(id = 'snp-table-div',\n",
    "                               children = [make_SNP_table()],\n",
    "                               className = 'sections'),\n",
    "                      dbc.Modal([dbc.ModalHeader(),\n",
    "                                 dbc.ModalBody(\"SNP table populated below.\")],\n",
    "                                 id = 'modal-snp-table',\n",
    "                                 is_open = False)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### HEADERS ####\n",
    "\n",
    "## Navigation Bar to match LocusFocus\n",
    "NAVBAR = dbc.NavbarSimple(id = 'navbar-lf',\n",
    "                          brand = \"LocusFocus\",\n",
    "                          brand_href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                          brand_external_link = \"_blank\",\n",
    "                          color = \"#40e0d0\",\n",
    "                          links_left = True,\n",
    "                          expand = \"lg\",\n",
    "                          children = [dbc.NavItem(dbc.NavLink(\"Colocalization\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Set-Based Test\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/setbasedtest\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"GWAS SVatalog\",\n",
    "                                                              href = \"https://svatalog.research.sickkids.ca/\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Documentation\",\n",
    "                                                              href = \"https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Contact Us\",\n",
    "                                                              id = 'button-modal-contact-us',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Contact Us\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         <h5 children=\"GWAS SVatalog\"  />\n",
    "                                                                                         Shalvi Chirmade <span style=\"color: #2ba089\" children=\"shalvi.chirmade@sickkids.ca\" />\n",
    "\n",
    "                                                                                         &nbsp;\n",
    "\n",
    "                                                                                         <h5 children=\"LocusFocus\" />\n",
    "                                                                                         Mackenzie Frew <span style=\"color: #2ba089\" children=\"mackenzie.frew@sickkids.ca\" />\n",
    "                                                                                        ''',\n",
    "                                                                                        dangerously_allow_html = True,\n",
    "                                                                                        id = 'contact-us-mrkdwn'))],\n",
    "                                                              id = 'modal-contact-us',\n",
    "                                                              is_open = False)]),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Subscribe\",\n",
    "                                                              href = \"https://mailchi.mp/752ab1c4d516/locusfocus\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Citation\",\n",
    "                                                              id = 'button-modal-citation',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Citation\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         *to be published*\n",
    "                                                                                        '''))],\n",
    "                                                              id = 'modal-citation',\n",
    "                                                              is_open = False)])])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Logo for the top of the webpage, disclaimer and info\n",
    "GSV_LOGO = html.Img(src = \"assets/gwas-svatalog-name.png\",\n",
    "                    id = 'gsv-logo-image')\n",
    "\n",
    "TCAG_LOGO = html.Img(src = \"assets/tcaglogo.png\",\n",
    "                     id = 'tcag-logo-image')\n",
    "\n",
    "SK_LOGO = html.Img(src = \"assets/SKlogo.svg\",\n",
    "                   id = 'sk-logo-image')\n",
    "\n",
    "DISCLAIMER_BUTTON = dbc.Button(\"Disclaimer\",\n",
    "                               color = \"secondary\",\n",
    "                               outline = True,\n",
    "                               size = \"sm\",\n",
    "                               id = 'disclaimer-button')\n",
    "\n",
    "DISCLAIMER_COLLAPSE = dbc.Collapse(dbc.Card(dbc.CardBody(\"Database constructed from predominantly European population of 101 individuals with Cystic Fibrosis (CF). The alleles affected by CF aside, the remainder of the genome is comparable to a healthy population of European descent (citation TBD). Genomic location is referenced against GRCh38.\")),\n",
    "                                   id = 'disclaimer-collapse',\n",
    "                                   is_open = False)\n",
    "\n",
    "DISCLAIMER_DIV = html.Div(id = 'disclaimer-div',\n",
    "                          children = [DISCLAIMER_BUTTON,\n",
    "                                      DISCLAIMER_COLLAPSE])\n",
    "\n",
    "# DOCUMENTATION = dcc.Markdown('''\n",
    "# Documentation for GWAS SVatalog can be found [here](https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html).\n",
    "# Colocalization testing of GWAS loci across various datasets can be conducted at [LocusFocus] (https://locusfocus.research.sickkids.ca/).\n",
    "# ''',\n",
    "#                              id = 'doc-markdown')\n",
    "\n",
    "# DISC_DOC_DIV = html.Div(id = 'disc-doc-div',\n",
    "#                         children = [DOCUMENTATION,\n",
    "#                                     DISCLAIMER_DIV])\n",
    "\n",
    "GSV_LOGO_DIV = html.Div(id = 'gsv-logo-image-div',\n",
    "                        children = [GSV_LOGO])\n",
    "\n",
    "COMP_LOGO_DIV = html.Div(id = 'comp-logo-image-div',\n",
    "                         children = [TCAG_LOGO,\n",
    "                                     SK_LOGO,\n",
    "                                     DISCLAIMER_DIV])\n",
    "\n",
    "LOGO_DIV = html.Div(id = 'logo-image-div',\n",
    "                    children = [GSV_LOGO_DIV,\n",
    "                                COMP_LOGO_DIV])\n",
    "\n",
    "\n",
    "HEADER_DIV = html.Div(id = 'header-line-div')\n",
    "\n",
    "\n",
    "# Header for filters\n",
    "FILTER_HEADER = html.H3(\"Search for Structural Variants\",\n",
    "                        id = 'filter-header',\n",
    "                        style = {'textAlign' : 'center',\n",
    "                                 'color' : '#2F2E2D',\n",
    "                                 'height' : '70 px'})\n",
    "\n",
    "FILTER_HEADER_DIV = html.Div(id = 'filter-header-div',\n",
    "                             children = [FILTER_HEADER],\n",
    "                             className = 'section-headers')\n",
    "\n",
    "\n",
    "\n",
    "# Header for plot\n",
    "PLOT_HEADER = html.Div([html.H3([\"SVs and GWAS Hits Linkage Disequilibirum Plot\",\n",
    "                                 question_icon],\n",
    "                                 id = 'plot-header',\n",
    "                                 style = {'textAlign' : 'center',\n",
    "                                          'color' : '#2F2E2D',\n",
    "                                          'height' : '70 px'}),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Interactive plot visualizing linkage disequilibrium (LD) between the selected SV and GWAS-associated SNPs from GWAS Catalog.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'plot-header',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'plot-header-tooltip')])\n",
    "\n",
    "PLOT_HEADER_DIV = html.Div(id = 'plot-header-div',\n",
    "                           children = [PLOT_HEADER],\n",
    "                           className = 'section-headers')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### CALLBACKS ####\n",
    "\n",
    "# Incorporate together to create interactive webpage\n",
    "app.layout = html.Div([NAVBAR,\n",
    "                       LOGO_DIV,\n",
    "                       HEADER_DIV,\n",
    "                       FILTER_HEADER_DIV,\n",
    "                       ALL_FILTER_DIV,\n",
    "                       SV_TABLE_DIV,\n",
    "                       PLOT_HEADER_DIV,\n",
    "                       SCATPLOT_DIV,\n",
    "                       SNP_TABLE])\n",
    "\n",
    "\n",
    "# Contact Us modal\n",
    "@app.callback(\n",
    "    Output('modal-contact-us', 'is_open'),\n",
    "    [Input('button-modal-contact-us', 'n_clicks')],\n",
    "    [State('modal-contact-us', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_contact, is_open):\n",
    "    if modal_contact:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Citation modal\n",
    "@app.callback(\n",
    "    Output('modal-citation', 'is_open'),\n",
    "    [Input('button-modal-citation', 'n_clicks')],\n",
    "    [State('modal-citation', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_citation, is_open):\n",
    "    if modal_citation:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Disclaimer cutton opens collapsed text\n",
    "@app.callback(\n",
    "    Output('disclaimer-collapse', 'is_open'),\n",
    "    [Input('disclaimer-button', 'n_clicks')],\n",
    "    [State('disclaimer-collapse', 'is_open')]\n",
    ")\n",
    "def dislcaimer_collapse(clicks, is_open):\n",
    "    if clicks:\n",
    "        return not is_open\n",
    "    return is_open\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Chromosome dropdown changes the genes and phenotypes displayed\n",
    "@app.callback(\n",
    "    Output('gene-dropdown', 'options'),\n",
    "    Output('phenotype-dropdown', 'options'),\n",
    "    Input('chromosome-dropdown', 'value')\n",
    ")\n",
    "def update_gene_dropdown(chrom):\n",
    "\n",
    "    if chrom is None:\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    elif chrom == \"Any\":\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    else:\n",
    "        chromosome = f\"chr{chrom}\"\n",
    "        new_genes = df_gene.filter(pl.col(\"chr\") == chromosome).select(\"gene\").unique().to_series().to_list()\n",
    "        new_genes = sorted(new_genes)\n",
    "        new_pheno = df_full_join.filter(pl.col(\"Chromosome\") == chromosome).select(\"Phenotype\").unique().to_series().to_list()\n",
    "        new_pheno = sorted(map(str, new_pheno))\n",
    "\n",
    "    return [{'label': str(i), 'value': str(i)} for i in new_genes], [{'label': str(i), 'value': str(i)} for i in new_pheno]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Filters alter the table displayed\n",
    "@app.callback(\n",
    "    Output('strvar-table-div', 'children'),\n",
    "    Input('chromosome-dropdown', 'value'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('gene-dropdown', 'value'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_table(chrom, pheno, gene, range_start, range_end):\n",
    "\n",
    "    chrom = chromosomes if chrom is None else chrom\n",
    "    pheno = phenotypes if pheno is None else pheno\n",
    "    gene = genes if gene is None else gene\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if chrom == \"Any\":\n",
    "        chrom = chromosomes\n",
    "    else:\n",
    "        chrom = chrom\n",
    "    \n",
    "    if pheno == \"Any\":\n",
    "        pheno = phenotypes\n",
    "    else:\n",
    "        pheno = pheno\n",
    "\n",
    "    new_df = make_table(chrom = chrom,\n",
    "                        range_start = range_start,\n",
    "                        range_end = range_end,\n",
    "                        pheno = pheno,\n",
    "                        gene = gene)\n",
    "\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('dtable-textbox', 'children'),\n",
    "    Input('strvar-table', 'derived_virtual_data'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_sv_textbox(data, selected_rows, range_start, range_end):\n",
    "\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if range_end <= range_start:\n",
    "        return html.P(\"Fix range values.\")\n",
    "\n",
    "    if data == []:\n",
    "        return html.P(\"No SVs in selected filters.\")\n",
    "    \n",
    "    elif selected_rows:\n",
    "        return html.P(\"Plot generated below.\")\n",
    "    \n",
    "    else:\n",
    "        return html.P(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('anno-table-div', 'children'),\n",
    "    Input('strvar-table', 'selected_rows')\n",
    ")\n",
    "def update_anno_table(selected_rows):\n",
    "\n",
    "    if selected_rows:\n",
    "        sv_selected = tab_show[selected_rows[0], 0]\n",
    "        new_anno_table = make_annotation_table(sv = sv_selected)\n",
    "\n",
    "    else:\n",
    "        new_anno_table = make_annotation_table(sv = \"None\")\n",
    "\n",
    "    return new_anno_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Collapse SV annotation table\n",
    "@app.callback(\n",
    "    Output(\"sv-table-collapse\", \"is_open\"),\n",
    "    [Input(\"sv-button-collapse\", \"n_clicks\")],\n",
    "    [State(\"sv-table-collapse\", \"is_open\")],\n",
    ")\n",
    "def toggle_collapse(n, is_open):\n",
    "    if n:\n",
    "        return not is_open\n",
    "    return is_open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Only row selection on table changes plot, SV div textbox and button enabling\n",
    "@app.callback(\n",
    "    Output('scatter-plot', 'figure'),\n",
    "    Output('scatter-plot', 'clickData'),\n",
    "    Output('download-button', 'disabled'),\n",
    "    Output('other-pheno-switch', 'disabled'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value')\n",
    ")\n",
    "def update_plot(selected_rows, pheno, toggle, toggle_pheno):\n",
    "\n",
    "    pheno = \"None\" if pheno is None else pheno\n",
    "    toggle = False if toggle is None else toggle\n",
    "    toggle_pheno = False if toggle_pheno is None else toggle_pheno\n",
    "\n",
    "    if selected_rows:\n",
    "        sv_name = tab_show[selected_rows[0], 0]\n",
    "        button = False\n",
    "\n",
    "    else:\n",
    "        sv_name = \"None\"\n",
    "        button = True\n",
    "\n",
    "    if pheno == \"None\":\n",
    "        pheno_toggle_disable = True\n",
    "\n",
    "    else:\n",
    "        pheno_toggle_disable = False\n",
    "\n",
    "    new_plot = make_plot(sv = sv_name,\n",
    "                         pheno = pheno,\n",
    "                         toggle = toggle,\n",
    "                         toggle_pheno = toggle_pheno)\n",
    "\n",
    "    return new_plot, {}, button, pheno_toggle_disable\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Update SNP table based on data point clicked on plot\n",
    "@app.callback(\n",
    "        Output('snp-table-div', 'children'),\n",
    "        Output('modal-snp-table', 'is_open'),\n",
    "        Input('scatter-plot', 'clickData'),\n",
    "        Input('strvar-table', 'selected_rows'),\n",
    "        State('modal-snp-table', 'is_open')\n",
    ")\n",
    "def update_SNP_table(clickData, selected_rows, is_open):\n",
    "\n",
    "    if clickData:\n",
    "        if clickData == {}:\n",
    "            rsID = \"None\"\n",
    "            is_open = False\n",
    "        else:\n",
    "            rsID = clickData[\"points\"][0][\"customdata\"][1]\n",
    "            is_open = True\n",
    "\n",
    "    else:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    if not selected_rows:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    new_snp_table = make_SNP_table(snp = rsID)\n",
    "\n",
    "    return new_snp_table, is_open\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:4321/\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:dash.dash:Dash is running on http://127.0.0.1:4321/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://127.0.0.1:4321/\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Download SNP data seen in plot\n",
    "@app.callback(\n",
    "    Output('download-csv', 'data'),\n",
    "    Input('download-button', 'n_clicks'),\n",
    "    Input('scatter-plot', 'relayoutData'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value'),\n",
    "    prevent_initial_call = True\n",
    ")\n",
    "def download_plot_data(n_clicks, relayoutData, toggle, toggle_pheno):\n",
    "    # Example from https://towardsdatascience.com/building-a-dashboard-in-plotly-dash-c748588e2920\n",
    "    global df_download\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage = \"D'\"\n",
    "    elif toggle == True:\n",
    "        linkage = \"R2\"\n",
    "    else:\n",
    "        linkage = \"D'\"\n",
    "\n",
    "\n",
    "    # If button was triggered\n",
    "    if ctx.triggered[0]['prop_id'] == 'download-button.n_clicks':\n",
    "        # Default range (no zoom)\n",
    "        if relayoutData is None or \"xaxis.autorange\" in relayoutData:\n",
    "            df_download = tab\n",
    "            df_extra = extra_points\n",
    "\n",
    "            if toggle_pheno:\n",
    "                df_download = pl.concat([df_download, df_extra, extra_points_pheno])\n",
    "            else:\n",
    "                df_download = pl.concat([df_download, df_extra])\n",
    "\n",
    "        # Filter by x-axis and y-axis ranges\n",
    "        elif \"xaxis.range[0]\" in relayoutData:\n",
    "            x_min, x_max = relayoutData[\"xaxis.range[0]\"], relayoutData[\"xaxis.range[1]\"]\n",
    "\n",
    "            if \"yaxis.range[0]\" in relayoutData:\n",
    "                y_min, y_max = relayoutData[\"yaxis.range[0]\"], relayoutData[\"yaxis.range[1]\"]\n",
    "\n",
    "                if \"P-Value\" in extra_points.columns:\n",
    "                    df_download = tab.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                        (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                        (pl.col(\"P-Value_log10\") <= y_max)\n",
    "                    )\n",
    "                    df_extra = extra_points.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                        (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                        (pl.col(\"P-Value_log10\") <= y_max)\n",
    "                    )\n",
    "                    df_extra_pheno = extra_points_pheno.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                        (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                        (pl.col(\"P-Value_log10\") <= y_max)\n",
    "                    )\n",
    "                else:\n",
    "                    df_download = tab.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                        (pl.col(linkage) >= y_min) &\n",
    "                        (pl.col(linkage) <= y_max)\n",
    "                    )\n",
    "                    df_extra = extra_points.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                        (pl.col(linkage) >= y_min) &\n",
    "                        (pl.col(linkage) <= y_max)\n",
    "                    )\n",
    "            else:  # Only x-axis range\n",
    "                df_download = tab.filter(\n",
    "                    (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                    (pl.col(\"SNP_Position\") <= x_max)\n",
    "                )\n",
    "                df_extra = extra_points.filter(\n",
    "                    (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                    (pl.col(\"SNP_Position\") <= x_max)\n",
    "                )\n",
    "                if toggle_pheno:\n",
    "                    df_extra_pheno = extra_points_pheno.filter(\n",
    "                        (pl.col(\"SNP_Position\") >= x_min) &\n",
    "                        (pl.col(\"SNP_Position\") <= x_max)\n",
    "                    )\n",
    "                    df_download = pl.concat([df_download, df_extra, df_extra_pheno])\n",
    "                else:\n",
    "                    df_download = pl.concat([df_download, df_extra])\n",
    "        else:  # Only y-axis range\n",
    "            y_min, y_max = relayoutData[\"yaxis.range[0]\"], relayoutData[\"yaxis.range[1]\"]\n",
    "            df_download = tab.filter(\n",
    "                (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                (pl.col(\"P-Value_log10\") <= y_max)\n",
    "            )\n",
    "            df_extra = extra_points.filter(\n",
    "                (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                (pl.col(\"P-Value_log10\") <= y_max)\n",
    "            )\n",
    "            if toggle_pheno:\n",
    "                df_extra_pheno = extra_points_pheno.filter(\n",
    "                    (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                    (pl.col(\"P-Value_log10\") <= y_max)\n",
    "                )\n",
    "                df_download = pl.concat([df_download, df_extra, df_extra_pheno])\n",
    "            else:\n",
    "                df_download = pl.concat([df_download, df_extra])\n",
    "            #raise PreventUpdate\n",
    "\n",
    "        return dcc.send_data_frame(df_download.to_pandas().to_csv, \"gwas_svatalog_snp_data.csv\")\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug = True,\n",
    "                   host = \"127.0.0.1\",\n",
    "                   port = 4321)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extra_points = df_sv_snp_join.filter(pl.col(\"SV_Name\") == \"P0000085185\")\n",
    "\n",
    "extra_points = extra_points.filter(pl.col(\"SNP_Name_dbSNP\").is_in(df_full_join.select(\"SNP_Name_dbSNP\").to_series()))\n",
    "\n",
    "    # Create a list of labels for hovertext\n",
    "df_lab = extra_points.to_dicts()\n",
    "\n",
    "def extra_label_maker(row):\n",
    "\n",
    "    return [row[\"SNP_Name_dbSNP\"],\n",
    "            row[\"Chromosome\"],\n",
    "            row[\"SNP_Position\"],\n",
    "            row[\"R2\"],\n",
    "            row[\"D'\"]]\n",
    "\n",
    "# sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "#                                 y = extra_points.select(linkage_value).to_series().to_list(),\n",
    "#                                 mode = \"markers\",\n",
    "#                                 marker = dict(size = 10,\n",
    "#                                                 color = \"rgba(168,168,168,0.5)\",\n",
    "#                                                 line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "#                                 hovertemplate = '<br><b>SNP Name: %{customdata[0]}</b>' +\n",
    "#                                                 '<br>Chromosome: %{customdata[1]}' + \n",
    "#                                                 '<br>SNP Position: %{customdata[2]}' + \n",
    "#                                                 '<br>r2: %{customdata[3]}' + \n",
    "#                                                 \"<br>D': %{customdata[4]}\",\n",
    "#                                 customdata = [extra_label_maker(row) for row in df_lab],\n",
    "#                                 showlegend = False,\n",
    "#                                 name = \"Not in \\nGWAS Catalog\"))\n",
    "extra_points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "60"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sv_snp_join.filter(pl.col(\"SV_Name\") == \"P0000085185\").select(\"SNP_Name_dbSNP\").n_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr > th,\n",
       ".dataframe > tbody > tr > td {\n",
       "  text-align: right;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (0, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Chromosome</th><th>SNP_Position</th><th>SNP_Name_GWAS</th><th>Risk_Allele</th><th>Risk_Allele_Frequency</th><th>Phenotype</th><th>Study</th><th>Pubmed_ID</th><th>Link</th><th>P-Value</th><th>P-Value_log10</th></tr><tr><td>str</td><td>i32</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td></tr></thead><tbody></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (0, 11)\n",
       "┌────────────┬────────────┬────────────┬───────────┬───┬───────────┬──────┬─────────┬──────────────┐\n",
       "│ Chromosome ┆ SNP_Positi ┆ SNP_Name_G ┆ Risk_Alle ┆ … ┆ Pubmed_ID ┆ Link ┆ P-Value ┆ P-Value_log1 │\n",
       "│ ---        ┆ on         ┆ WAS        ┆ le        ┆   ┆ ---       ┆ ---  ┆ ---     ┆ 0            │\n",
       "│ str        ┆ ---        ┆ ---        ┆ ---       ┆   ┆ str       ┆ str  ┆ f64     ┆ ---          │\n",
       "│            ┆ i32        ┆ str        ┆ str       ┆   ┆           ┆      ┆         ┆ f64          │\n",
       "╞════════════╪════════════╪════════════╪═══════════╪═══╪═══════════╪══════╪═════════╪══════════════╡\n",
       "└────────────┴────────────┴────────────┴───────────┴───┴───────────┴──────┴─────────┴──────────────┘"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = df_full_join.filter((pl.col(\"SV_Name\") == \"P0000085185\") & (pl.col(\"Phenotype\") == \"Depression\"))\n",
    "\n",
    "# Add points from gwas for pheno, without r2 and D' values\n",
    "extra_points = (df_gwas.filter((pl.col(\"Chromosome\") == \"chr7\") &\n",
    "                                (pl.col(\"SNP_Position\") >= (12242077 - 10000)) &\n",
    "                                (pl.col(\"SNP_Position\") <= (12242077 + 10000)) &\n",
    "                                (pl.col(\"Phenotype\") == \"Depression\"))\n",
    "                        .filter(~pl.col(\"SNP_Name_GWAS\").is_in(test.select(\"SNP_Name_GWAS\").to_series())))\n",
    "\n",
    "extra_points = (extra_points.with_columns(pl.when(pl.col(\"P-Value\") < 1e-50)\n",
    "                                            .then(1e-50)\n",
    "                                            .otherwise(pl.col(\"P-Value\"))\n",
    "                                            .alias(\"P-Value\"),\n",
    "                                            (-pl.col(\"P-Value\").log10()).alias(\"P-Value_log10\")))\n",
    "\n",
    "# # Create a list of labels for hovertext\n",
    "# df_lab = extra_points.to_dicts()\n",
    "\n",
    "# def label_maker(row):\n",
    "\n",
    "#     return [row[\"Phenotype\"],\n",
    "#             row[\"SNP_Name_GWAS\"],\n",
    "#             row[\"Chromosome\"],\n",
    "#             row[\"SNP_Position\"],\n",
    "#             row[\"P-Value\"]]\n",
    "\n",
    "\n",
    "# sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "#                                 y = extra_points.select(\"P-Value_log10\").to_series().to_list(),\n",
    "#                                 mode = \"markers\",\n",
    "#                                 marker = dict(size = 10,\n",
    "#                                                 color = \"rgba(168,168,168,0.5)\",\n",
    "#                                                 line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "#                                 hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "#                                                 '<br>SNP Name: %{customdata[1]}' + \n",
    "#                                                 '<br>Chromosome: %{customdata[2]}' + \n",
    "#                                                 '<br>SNP Position: %{customdata[3]}' + \n",
    "#                                                 '<br>P-Value: %{customdata[4]}',\n",
    "#                                 customdata = [label_maker(row) for row in df_lab],\n",
    "#                                 showlegend = False,\n",
    "#                                 name = \"No LD Data\"))\n",
    "extra_points"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
