{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dash_bootstrap_components as dbc\n",
    "import dash_daq as daq\n",
    "import heapq\n",
    "import numpy as np\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import polars as pl\n",
    "\n",
    "from dash import Dash, dcc, html, dash_table, Input, Output, State, ctx\n",
    "from jupyter_dash import JupyterDash\n",
    "from gtfparse import read_gtf # gtfparse==1.3.0\n",
    "\n",
    "\n",
    "app = JupyterDash(__name__, \n",
    "           external_stylesheets = [dbc.themes.LUX, dbc.icons.BOOTSTRAP],\n",
    "           title = \"GWAS SVatalog\")\n",
    "\n",
    "app._favicon = \"assets/favicon.ico\"\n",
    "\n",
    "server = app.server\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/l3/z7w_786570q78qsbk5sg6k2c0000gp/T/ipykernel_12796/61282847.py:6: DtypeWarning: Columns (9,23) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  DF_GWAS_FULL = pd.read_csv(\"data/gwas_catalog_v1.0-associations_e108.tsv\",\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>RISK ALLELE FREQUENCY</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2017-11-30</td>\n",
       "      <td>29059683</td>\n",
       "      <td>Michailidou K</td>\n",
       "      <td>2017-10-23</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>76,192 European ancestry cases, 63,082 Europea...</td>\n",
       "      <td>46,785 European ancestry cases, 42,892 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0867</td>\n",
       "      <td>7.000000e-06</td>\n",
       "      <td>5.154902</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>0.052000</td>\n",
       "      <td>[0.029-0.075] unit decrease</td>\n",
       "      <td>Illumina [~ 11800000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-11-30</td>\n",
       "      <td>29059683</td>\n",
       "      <td>Michailidou K</td>\n",
       "      <td>2017-10-23</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>76,192 European ancestry cases, 63,082 Europea...</td>\n",
       "      <td>46,785 European ancestry cases, 42,892 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0367</td>\n",
       "      <td>5.000000e-07</td>\n",
       "      <td>6.301030</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>0.087500</td>\n",
       "      <td>[0.053-0.122] unit decrease</td>\n",
       "      <td>Illumina [~ 11800000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2017-11-30</td>\n",
       "      <td>29059683</td>\n",
       "      <td>Michailidou K</td>\n",
       "      <td>2017-10-23</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>76,192 European ancestry cases, 63,082 Europea...</td>\n",
       "      <td>46,785 European ancestry cases, 42,892 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2459</td>\n",
       "      <td>4.000000e-13</td>\n",
       "      <td>12.397940</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>0.053200</td>\n",
       "      <td>[0.039-0.068] unit decrease</td>\n",
       "      <td>Illumina [~ 11800000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2017-11-30</td>\n",
       "      <td>29059683</td>\n",
       "      <td>Michailidou K</td>\n",
       "      <td>2017-10-23</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>76,192 European ancestry cases, 63,082 Europea...</td>\n",
       "      <td>46,785 European ancestry cases, 42,892 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.1018</td>\n",
       "      <td>4.000000e-06</td>\n",
       "      <td>5.397940</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>0.048300</td>\n",
       "      <td>[0.028-0.069] unit decrease</td>\n",
       "      <td>Illumina [~ 11800000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2017-11-30</td>\n",
       "      <td>29059683</td>\n",
       "      <td>Michailidou K</td>\n",
       "      <td>2017-10-23</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>76,192 European ancestry cases, 63,082 Europea...</td>\n",
       "      <td>46,785 European ancestry cases, 42,892 Europea...</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0639</td>\n",
       "      <td>6.000000e-06</td>\n",
       "      <td>5.221849</td>\n",
       "      <td>(EA)</td>\n",
       "      <td>0.071200</td>\n",
       "      <td>[0.04-0.102] unit decrease</td>\n",
       "      <td>Illumina [~ 11800000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471108</th>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>32916098</td>\n",
       "      <td>Verweij N</td>\n",
       "      <td>2020-09-02</td>\n",
       "      <td>Cell Syst</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>63,706 European and unknown ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>3.000000e-08</td>\n",
       "      <td>7.522879</td>\n",
       "      <td>(6 ms)</td>\n",
       "      <td>0.083091</td>\n",
       "      <td>[0.053-0.113] unit increase</td>\n",
       "      <td>Affymetrix [10000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471109</th>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>32916098</td>\n",
       "      <td>Verweij N</td>\n",
       "      <td>2020-09-02</td>\n",
       "      <td>Cell Syst</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>63,706 European and unknown ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>5.000000e-08</td>\n",
       "      <td>7.301030</td>\n",
       "      <td>(10 ms)</td>\n",
       "      <td>0.081970</td>\n",
       "      <td>[0.052-0.112] unit increase</td>\n",
       "      <td>Affymetrix [10000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471110</th>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>32916098</td>\n",
       "      <td>Verweij N</td>\n",
       "      <td>2020-09-02</td>\n",
       "      <td>Cell Syst</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>63,706 European and unknown ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>(8 ms)</td>\n",
       "      <td>0.085389</td>\n",
       "      <td>[0.055-0.116] unit increase</td>\n",
       "      <td>Affymetrix [10000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471111</th>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>32916098</td>\n",
       "      <td>Verweij N</td>\n",
       "      <td>2020-09-02</td>\n",
       "      <td>Cell Syst</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>63,706 European and unknown ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.791595</td>\n",
       "      <td>7.000000e-09</td>\n",
       "      <td>8.154902</td>\n",
       "      <td>(322 ms)</td>\n",
       "      <td>0.038484</td>\n",
       "      <td>[0.025-0.052] unit increase</td>\n",
       "      <td>Affymetrix [10000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471112</th>\n",
       "      <td>2020-11-30</td>\n",
       "      <td>32916098</td>\n",
       "      <td>Verweij N</td>\n",
       "      <td>2020-09-02</td>\n",
       "      <td>Cell Syst</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>63,706 European and unknown ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.791595</td>\n",
       "      <td>3.000000e-08</td>\n",
       "      <td>7.522879</td>\n",
       "      <td>(288 ms)</td>\n",
       "      <td>0.036731</td>\n",
       "      <td>[0.023-0.05] unit increase</td>\n",
       "      <td>Affymetrix [10000000] (imputed)</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>471113 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATE ADDED TO CATALOG  PUBMEDID   FIRST AUTHOR        DATE    JOURNAL   \n",
       "0                 2017-11-30  29059683  Michailidou K  2017-10-23     Nature  \\\n",
       "1                 2017-11-30  29059683  Michailidou K  2017-10-23     Nature   \n",
       "2                 2017-11-30  29059683  Michailidou K  2017-10-23     Nature   \n",
       "3                 2017-11-30  29059683  Michailidou K  2017-10-23     Nature   \n",
       "4                 2017-11-30  29059683  Michailidou K  2017-10-23     Nature   \n",
       "...                      ...       ...            ...         ...        ...   \n",
       "471108            2020-11-30  32916098      Verweij N  2020-09-02  Cell Syst   \n",
       "471109            2020-11-30  32916098      Verweij N  2020-09-02  Cell Syst   \n",
       "471110            2020-11-30  32916098      Verweij N  2020-09-02  Cell Syst   \n",
       "471111            2020-11-30  32916098      Verweij N  2020-09-02  Cell Syst   \n",
       "471112            2020-11-30  32916098      Verweij N  2020-09-02  Cell Syst   \n",
       "\n",
       "                                        LINK   \n",
       "0       www.ncbi.nlm.nih.gov/pubmed/29059683  \\\n",
       "1       www.ncbi.nlm.nih.gov/pubmed/29059683   \n",
       "2       www.ncbi.nlm.nih.gov/pubmed/29059683   \n",
       "3       www.ncbi.nlm.nih.gov/pubmed/29059683   \n",
       "4       www.ncbi.nlm.nih.gov/pubmed/29059683   \n",
       "...                                      ...   \n",
       "471108  www.ncbi.nlm.nih.gov/pubmed/32916098   \n",
       "471109  www.ncbi.nlm.nih.gov/pubmed/32916098   \n",
       "471110  www.ncbi.nlm.nih.gov/pubmed/32916098   \n",
       "471111  www.ncbi.nlm.nih.gov/pubmed/32916098   \n",
       "471112  www.ncbi.nlm.nih.gov/pubmed/32916098   \n",
       "\n",
       "                                                    STUDY   \n",
       "0       Association analysis identifies 65 new breast ...  \\\n",
       "1       Association analysis identifies 65 new breast ...   \n",
       "2       Association analysis identifies 65 new breast ...   \n",
       "3       Association analysis identifies 65 new breast ...   \n",
       "4       Association analysis identifies 65 new breast ...   \n",
       "...                                                   ...   \n",
       "471108       The Genetic Makeup of the Electrocardiogram.   \n",
       "471109       The Genetic Makeup of the Electrocardiogram.   \n",
       "471110       The Genetic Makeup of the Electrocardiogram.   \n",
       "471111       The Genetic Makeup of the Electrocardiogram.   \n",
       "471112       The Genetic Makeup of the Electrocardiogram.   \n",
       "\n",
       "                                            DISEASE/TRAIT   \n",
       "0                                           Breast cancer  \\\n",
       "1                                           Breast cancer   \n",
       "2                                           Breast cancer   \n",
       "3                                           Breast cancer   \n",
       "4                                           Breast cancer   \n",
       "...                                                   ...   \n",
       "471108  Electrocardiogram morphology (amplitude at tem...   \n",
       "471109  Electrocardiogram morphology (amplitude at tem...   \n",
       "471110  Electrocardiogram morphology (amplitude at tem...   \n",
       "471111  Electrocardiogram morphology (amplitude at tem...   \n",
       "471112  Electrocardiogram morphology (amplitude at tem...   \n",
       "\n",
       "                                      INITIAL SAMPLE SIZE   \n",
       "0       76,192 European ancestry cases, 63,082 Europea...  \\\n",
       "1       76,192 European ancestry cases, 63,082 Europea...   \n",
       "2       76,192 European ancestry cases, 63,082 Europea...   \n",
       "3       76,192 European ancestry cases, 63,082 Europea...   \n",
       "4       76,192 European ancestry cases, 63,082 Europea...   \n",
       "...                                                   ...   \n",
       "471108   63,706 European and unknown ancestry individuals   \n",
       "471109   63,706 European and unknown ancestry individuals   \n",
       "471110   63,706 European and unknown ancestry individuals   \n",
       "471111   63,706 European and unknown ancestry individuals   \n",
       "471112   63,706 European and unknown ancestry individuals   \n",
       "\n",
       "                                  REPLICATION SAMPLE SIZE  ...   \n",
       "0       46,785 European ancestry cases, 42,892 Europea...  ...  \\\n",
       "1       46,785 European ancestry cases, 42,892 Europea...  ...   \n",
       "2       46,785 European ancestry cases, 42,892 Europea...  ...   \n",
       "3       46,785 European ancestry cases, 42,892 Europea...  ...   \n",
       "4       46,785 European ancestry cases, 42,892 Europea...  ...   \n",
       "...                                                   ...  ...   \n",
       "471108                                                NaN  ...   \n",
       "471109                                                NaN  ...   \n",
       "471110                                                NaN  ...   \n",
       "471111                                                NaN  ...   \n",
       "471112                                                NaN  ...   \n",
       "\n",
       "                          CONTEXT INTERGENIC  RISK ALLELE FREQUENCY   \n",
       "0              intergenic_variant        1.0                 0.0867  \\\n",
       "1       regulatory_region_variant        1.0                 0.0367   \n",
       "2                  intron_variant        0.0                 0.2459   \n",
       "3              intergenic_variant        1.0                 0.1018   \n",
       "4              intergenic_variant        1.0                 0.0639   \n",
       "...                           ...        ...                    ...   \n",
       "471108             intron_variant        0.0               0.963504   \n",
       "471109             intron_variant        0.0               0.963504   \n",
       "471110             intron_variant        0.0               0.963504   \n",
       "471111         intergenic_variant        1.0               0.791595   \n",
       "471112         intergenic_variant        1.0               0.791595   \n",
       "\n",
       "             P-VALUE PVALUE_MLOG P-VALUE (TEXT) OR or BETA   \n",
       "0       7.000000e-06    5.154902           (EA)   0.052000  \\\n",
       "1       5.000000e-07    6.301030           (EA)   0.087500   \n",
       "2       4.000000e-13   12.397940           (EA)   0.053200   \n",
       "3       4.000000e-06    5.397940           (EA)   0.048300   \n",
       "4       6.000000e-06    5.221849           (EA)   0.071200   \n",
       "...              ...         ...            ...        ...   \n",
       "471108  3.000000e-08    7.522879         (6 ms)   0.083091   \n",
       "471109  5.000000e-08    7.301030        (10 ms)   0.081970   \n",
       "471110  1.000000e-08    8.000000         (8 ms)   0.085389   \n",
       "471111  7.000000e-09    8.154902       (322 ms)   0.038484   \n",
       "471112  3.000000e-08    7.522879       (288 ms)   0.036731   \n",
       "\n",
       "                      95% CI (TEXT)       PLATFORM [SNPS PASSING QC]  CNV  \n",
       "0       [0.029-0.075] unit decrease  Illumina [~ 11800000] (imputed)    N  \n",
       "1       [0.053-0.122] unit decrease  Illumina [~ 11800000] (imputed)    N  \n",
       "2       [0.039-0.068] unit decrease  Illumina [~ 11800000] (imputed)    N  \n",
       "3       [0.028-0.069] unit decrease  Illumina [~ 11800000] (imputed)    N  \n",
       "4        [0.04-0.102] unit decrease  Illumina [~ 11800000] (imputed)    N  \n",
       "...                             ...                              ...  ...  \n",
       "471108  [0.053-0.113] unit increase  Affymetrix [10000000] (imputed)    N  \n",
       "471109  [0.052-0.112] unit increase  Affymetrix [10000000] (imputed)    N  \n",
       "471110  [0.055-0.116] unit increase  Affymetrix [10000000] (imputed)    N  \n",
       "471111  [0.025-0.052] unit increase  Affymetrix [10000000] (imputed)    N  \n",
       "471112   [0.023-0.05] unit increase  Affymetrix [10000000] (imputed)    N  \n",
       "\n",
       "[471113 rows x 34 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### READ IN REQUIRED DATA ####\n",
    "\n",
    "## GWAS Catalog data v1.0 - downloaded on January 14, 2023 and edited by Dr. Zhuozhi Wang.\n",
    "    # Expanded SNP haplotypes to match the multiple chromosome locations\n",
    "    # Used dbSNP151 to correspond the correct rsID to the chromosome location when multiple rsIDs were provided\n",
    "DF_GWAS_FULL = pd.read_csv(\"data/gwas_catalog_v1.0-associations_e108.tsv\",\n",
    "                           sep = \"\\t\",\n",
    "                           header = 0,\n",
    "                           dtype = {\"PUBMEDID\" : \"object\",\n",
    "                                    \"LINK\" : \"object\",\n",
    "                                    \"P-VALUE\" : \"float64\"})\n",
    "DF_GWAS_FULL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_gwas = DF_GWAS_FULL[[\"CHR_ID\",\n",
    "                        \"CHR_POS\",\n",
    "                        \"SNPS\",\n",
    "                        \"STRONGEST SNP-RISK ALLELE\",\n",
    "                        \"RISK ALLELE FREQUENCY\",\n",
    "                        \"DISEASE/TRAIT\",\n",
    "                        \"STUDY\",\n",
    "                        \"PUBMEDID\",\n",
    "                        \"LINK\",\n",
    "                        \"P-VALUE\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the rows with empty chromosome location\n",
    "df_gwas = df_gwas.dropna(subset = \"CHR_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the first rsID only\n",
    "df_gwas[\"SNPS\"] = df_gwas[\"SNPS\"].str.split(\";\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change dtype of chromosome position - prevent decimals. Could not do before as there were ; in rows\n",
    "df_gwas[\"CHR_POS\"] = df_gwas[\"CHR_POS\"].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gwas.columns = [\"Chromosome\",\n",
    "                   \"SNP_Position\",\n",
    "                   \"SNP_Name_GWAS\",\n",
    "                   \"Risk_Allele\",\n",
    "                   \"Risk_Allele_Frequency\",\n",
    "                   \"Phenotype\",\n",
    "                   \"Study\",\n",
    "                   \"Pubmed_ID\",\n",
    "                   \"Link\",\n",
    "                   \"P-Value\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gwas[\"Chromosome\"] = \"chr\" + df_gwas[\"Chromosome\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gwas[\"Risk_Allele\"] = df_gwas[\"Risk_Allele\"].str.split(\"-\").str[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_Position</th>\n",
       "      <th>SNP_Name_GWAS</th>\n",
       "      <th>Risk_Allele</th>\n",
       "      <th>Risk_Allele_Frequency</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Study</th>\n",
       "      <th>Pubmed_ID</th>\n",
       "      <th>Link</th>\n",
       "      <th>P-Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>160772472</td>\n",
       "      <td>rs146071273</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0867</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>29059683</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>7.000000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr3</td>\n",
       "      <td>47364189</td>\n",
       "      <td>rs146331761</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0367</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>29059683</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>82217129</td>\n",
       "      <td>rs146817970</td>\n",
       "      <td>T</td>\n",
       "      <td>0.2459</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>29059683</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>4.000000e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr14</td>\n",
       "      <td>38968140</td>\n",
       "      <td>rs147032157</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1018</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>29059683</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>4.000000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>107031171</td>\n",
       "      <td>rs148731571</td>\n",
       "      <td>GCTTAGATATTAC</td>\n",
       "      <td>0.0639</td>\n",
       "      <td>Breast cancer</td>\n",
       "      <td>Association analysis identifies 65 new breast ...</td>\n",
       "      <td>29059683</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/29059683</td>\n",
       "      <td>6.000000e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471108</th>\n",
       "      <td>chr5</td>\n",
       "      <td>77842993</td>\n",
       "      <td>rs111778721</td>\n",
       "      <td>C</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>32916098</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>3.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471109</th>\n",
       "      <td>chr5</td>\n",
       "      <td>77842993</td>\n",
       "      <td>rs111778721</td>\n",
       "      <td>C</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>32916098</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>5.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471110</th>\n",
       "      <td>chr5</td>\n",
       "      <td>77842993</td>\n",
       "      <td>rs111778721</td>\n",
       "      <td>C</td>\n",
       "      <td>0.963504</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>32916098</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471111</th>\n",
       "      <td>chr5</td>\n",
       "      <td>45014943</td>\n",
       "      <td>rs111869282</td>\n",
       "      <td>C</td>\n",
       "      <td>0.791595</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>32916098</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>7.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471112</th>\n",
       "      <td>chr5</td>\n",
       "      <td>45014943</td>\n",
       "      <td>rs111869282</td>\n",
       "      <td>C</td>\n",
       "      <td>0.791595</td>\n",
       "      <td>Electrocardiogram morphology (amplitude at tem...</td>\n",
       "      <td>The Genetic Makeup of the Electrocardiogram.</td>\n",
       "      <td>32916098</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32916098</td>\n",
       "      <td>3.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>449339 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  SNP_Position SNP_Name_GWAS    Risk_Allele   \n",
       "0            chr2     160772472   rs146071273              A  \\\n",
       "1            chr3      47364189   rs146331761              A   \n",
       "2            chr5      82217129   rs146817970              T   \n",
       "3           chr14      38968140   rs147032157              A   \n",
       "4           chr13     107031171   rs148731571  GCTTAGATATTAC   \n",
       "...           ...           ...           ...            ...   \n",
       "471108       chr5      77842993   rs111778721              C   \n",
       "471109       chr5      77842993   rs111778721              C   \n",
       "471110       chr5      77842993   rs111778721              C   \n",
       "471111       chr5      45014943   rs111869282              C   \n",
       "471112       chr5      45014943   rs111869282              C   \n",
       "\n",
       "       Risk_Allele_Frequency   \n",
       "0                     0.0867  \\\n",
       "1                     0.0367   \n",
       "2                     0.2459   \n",
       "3                     0.1018   \n",
       "4                     0.0639   \n",
       "...                      ...   \n",
       "471108              0.963504   \n",
       "471109              0.963504   \n",
       "471110              0.963504   \n",
       "471111              0.791595   \n",
       "471112              0.791595   \n",
       "\n",
       "                                                Phenotype   \n",
       "0                                           Breast cancer  \\\n",
       "1                                           Breast cancer   \n",
       "2                                           Breast cancer   \n",
       "3                                           Breast cancer   \n",
       "4                                           Breast cancer   \n",
       "...                                                   ...   \n",
       "471108  Electrocardiogram morphology (amplitude at tem...   \n",
       "471109  Electrocardiogram morphology (amplitude at tem...   \n",
       "471110  Electrocardiogram morphology (amplitude at tem...   \n",
       "471111  Electrocardiogram morphology (amplitude at tem...   \n",
       "471112  Electrocardiogram morphology (amplitude at tem...   \n",
       "\n",
       "                                                    Study Pubmed_ID   \n",
       "0       Association analysis identifies 65 new breast ...  29059683  \\\n",
       "1       Association analysis identifies 65 new breast ...  29059683   \n",
       "2       Association analysis identifies 65 new breast ...  29059683   \n",
       "3       Association analysis identifies 65 new breast ...  29059683   \n",
       "4       Association analysis identifies 65 new breast ...  29059683   \n",
       "...                                                   ...       ...   \n",
       "471108       The Genetic Makeup of the Electrocardiogram.  32916098   \n",
       "471109       The Genetic Makeup of the Electrocardiogram.  32916098   \n",
       "471110       The Genetic Makeup of the Electrocardiogram.  32916098   \n",
       "471111       The Genetic Makeup of the Electrocardiogram.  32916098   \n",
       "471112       The Genetic Makeup of the Electrocardiogram.  32916098   \n",
       "\n",
       "                                        Link       P-Value  \n",
       "0       www.ncbi.nlm.nih.gov/pubmed/29059683  7.000000e-06  \n",
       "1       www.ncbi.nlm.nih.gov/pubmed/29059683  5.000000e-07  \n",
       "2       www.ncbi.nlm.nih.gov/pubmed/29059683  4.000000e-13  \n",
       "3       www.ncbi.nlm.nih.gov/pubmed/29059683  4.000000e-06  \n",
       "4       www.ncbi.nlm.nih.gov/pubmed/29059683  6.000000e-06  \n",
       "...                                      ...           ...  \n",
       "471108  www.ncbi.nlm.nih.gov/pubmed/32916098  3.000000e-08  \n",
       "471109  www.ncbi.nlm.nih.gov/pubmed/32916098  5.000000e-08  \n",
       "471110  www.ncbi.nlm.nih.gov/pubmed/32916098  1.000000e-08  \n",
       "471111  www.ncbi.nlm.nih.gov/pubmed/32916098  7.000000e-09  \n",
       "471112  www.ncbi.nlm.nih.gov/pubmed/32916098  3.000000e-08  \n",
       "\n",
       "[449339 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas = df_gwas[df_gwas[\"P-Value\"] <= 1]\n",
    "df_gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SV Annotation data - created by Dr. Zhuozhi Wang on March 13, 2023\n",
    "DF_ANNO_FULL = pd.read_csv(\"data/sv_annotations.tsv\",\n",
    "                           sep = \"\\t\",\n",
    "                           header = 0,\n",
    "                           dtype = {\"Start\" : \"Int64\",\n",
    "                                    \"End\" : \"Int64\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_anno = DF_ANNO_FULL[[\"Chromosome\",\n",
    "                        \"Start\",\n",
    "                        \"End\",\n",
    "                        \"ID\",\n",
    "                        \"Type\",\n",
    "                        \"Length\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SV_Start</th>\n",
       "      <th>SV_End</th>\n",
       "      <th>SV_Name</th>\n",
       "      <th>SV_Type</th>\n",
       "      <th>SV_Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11025</td>\n",
       "      <td>11068</td>\n",
       "      <td>P0000000001</td>\n",
       "      <td>del</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31973</td>\n",
       "      <td>32073</td>\n",
       "      <td>P0000000005</td>\n",
       "      <td>ins</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>50937</td>\n",
       "      <td>51053</td>\n",
       "      <td>P0000000006</td>\n",
       "      <td>del</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54656</td>\n",
       "      <td>55064</td>\n",
       "      <td>P0000000007</td>\n",
       "      <td>ins</td>\n",
       "      <td>409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>66162</td>\n",
       "      <td>66530</td>\n",
       "      <td>P0000000010</td>\n",
       "      <td>ins</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87178</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156029060</td>\n",
       "      <td>156029422</td>\n",
       "      <td>P0000225361</td>\n",
       "      <td>ins</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87179</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156029104</td>\n",
       "      <td>156029422</td>\n",
       "      <td>P0000225363</td>\n",
       "      <td>ins</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87180</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156029152</td>\n",
       "      <td>156029303</td>\n",
       "      <td>P0000225364</td>\n",
       "      <td>ins</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87181</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156029257</td>\n",
       "      <td>156029314</td>\n",
       "      <td>P0000225365</td>\n",
       "      <td>del</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87182</th>\n",
       "      <td>chrX</td>\n",
       "      <td>156029320</td>\n",
       "      <td>156029379</td>\n",
       "      <td>P0000225366</td>\n",
       "      <td>del</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87183 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome   SV_Start     SV_End      SV_Name SV_Type  SV_Length\n",
       "0           chr1      11025      11068  P0000000001     del         44\n",
       "1           chr1      31973      32073  P0000000005     ins        101\n",
       "2           chr1      50937      51053  P0000000006     del        117\n",
       "3           chr1      54656      55064  P0000000007     ins        409\n",
       "4           chr1      66162      66530  P0000000010     ins        369\n",
       "...          ...        ...        ...          ...     ...        ...\n",
       "87178       chrX  156029060  156029422  P0000225361     ins        363\n",
       "87179       chrX  156029104  156029422  P0000225363     ins        319\n",
       "87180       chrX  156029152  156029303  P0000225364     ins        152\n",
       "87181       chrX  156029257  156029314  P0000225365     del         58\n",
       "87182       chrX  156029320  156029379  P0000225366     del         60\n",
       "\n",
       "[87183 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_anno.columns =[\"Chromosome\",\n",
    "                  \"SV_Start\",\n",
    "                  \"SV_End\",\n",
    "                  \"SV_Name\",\n",
    "                  \"SV_Type\",\n",
    "                  \"SV_Length\"]\n",
    "df_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_Name</th>\n",
       "      <th>SNP_Position</th>\n",
       "      <th>SV_Name</th>\n",
       "      <th>SV_Position</th>\n",
       "      <th>R2</th>\n",
       "      <th>D'</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8_220692_T_C</td>\n",
       "      <td>220692</td>\n",
       "      <td>P0000096843</td>\n",
       "      <td>81301</td>\n",
       "      <td>0.000516</td>\n",
       "      <td>0.073408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8_482881_G_A</td>\n",
       "      <td>482881</td>\n",
       "      <td>P0000096843</td>\n",
       "      <td>81301</td>\n",
       "      <td>0.007804</td>\n",
       "      <td>0.100612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8_627101_A_G</td>\n",
       "      <td>627101</td>\n",
       "      <td>P0000096843</td>\n",
       "      <td>81301</td>\n",
       "      <td>0.029213</td>\n",
       "      <td>0.438748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8_685274_A_G</td>\n",
       "      <td>685274</td>\n",
       "      <td>P0000096843</td>\n",
       "      <td>81301</td>\n",
       "      <td>0.012024</td>\n",
       "      <td>0.180577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8_685275_C_T</td>\n",
       "      <td>685275</td>\n",
       "      <td>P0000096843</td>\n",
       "      <td>81301</td>\n",
       "      <td>0.010783</td>\n",
       "      <td>0.172748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14420</th>\n",
       "      <td>chrX_155225806_A_G</td>\n",
       "      <td>155225806</td>\n",
       "      <td>P0000225359</td>\n",
       "      <td>156029063</td>\n",
       "      <td>0.018419</td>\n",
       "      <td>0.157063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14421</th>\n",
       "      <td>chrX_155491696_G_C</td>\n",
       "      <td>155491696</td>\n",
       "      <td>P0000225359</td>\n",
       "      <td>156029063</td>\n",
       "      <td>0.033047</td>\n",
       "      <td>0.309470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14422</th>\n",
       "      <td>chrX_155757485_T_C</td>\n",
       "      <td>155757485</td>\n",
       "      <td>P0000225359</td>\n",
       "      <td>156029063</td>\n",
       "      <td>0.032094</td>\n",
       "      <td>0.316576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14423</th>\n",
       "      <td>chrX_155954788_C_T</td>\n",
       "      <td>155954788</td>\n",
       "      <td>P0000225359</td>\n",
       "      <td>156029063</td>\n",
       "      <td>0.066441</td>\n",
       "      <td>0.260534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14424</th>\n",
       "      <td>chrX_156000962_A_G</td>\n",
       "      <td>156000962</td>\n",
       "      <td>P0000225359</td>\n",
       "      <td>156029063</td>\n",
       "      <td>0.160441</td>\n",
       "      <td>0.588111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3100088 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SNP_Name  SNP_Position      SV_Name  SV_Position        R2   \n",
       "0         chr8_220692_T_C        220692  P0000096843        81301  0.000516  \\\n",
       "1         chr8_482881_G_A        482881  P0000096843        81301  0.007804   \n",
       "2         chr8_627101_A_G        627101  P0000096843        81301  0.029213   \n",
       "3         chr8_685274_A_G        685274  P0000096843        81301  0.012024   \n",
       "4         chr8_685275_C_T        685275  P0000096843        81301  0.010783   \n",
       "...                   ...           ...          ...          ...       ...   \n",
       "14420  chrX_155225806_A_G     155225806  P0000225359    156029063  0.018419   \n",
       "14421  chrX_155491696_G_C     155491696  P0000225359    156029063  0.033047   \n",
       "14422  chrX_155757485_T_C     155757485  P0000225359    156029063  0.032094   \n",
       "14423  chrX_155954788_C_T     155954788  P0000225359    156029063  0.066441   \n",
       "14424  chrX_156000962_A_G     156000962  P0000225359    156029063  0.160441   \n",
       "\n",
       "             D'  \n",
       "0      0.073408  \n",
       "1      0.100612  \n",
       "2      0.438748  \n",
       "3      0.180577  \n",
       "4      0.172748  \n",
       "...         ...  \n",
       "14420  0.157063  \n",
       "14421  0.309470  \n",
       "14422  0.316576  \n",
       "14423  0.260534  \n",
       "14424  0.588111  \n",
       "\n",
       "[3100088 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LD_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                       \"*_ld_stats.txt\"))\n",
    "LD_tables = [pd.read_table(file_path,\n",
    "                           header = 0,\n",
    "                           names = [\"SNP_Name\", \n",
    "                                    \"SNP_Position\",\n",
    "                                    \"SV_Name\",\n",
    "                                    \"SV_Position\",\n",
    "                                    \"R2\",\n",
    "                                    \"D'\"])\n",
    "             for file_path in LD_file_paths]\n",
    "\n",
    "DF_LD = pd.concat(LD_tables)\n",
    "DF_LD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_Position</th>\n",
       "      <th>SV_SNP_Name</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Alternate_Allele</th>\n",
       "      <th>Sample_AF</th>\n",
       "      <th>gnomAD_nfe_AF</th>\n",
       "      <th>dbSNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13</td>\n",
       "      <td>17767059</td>\n",
       "      <td>P0000147722</td>\n",
       "      <td>N</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.128713</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171311</td>\n",
       "      <td>chr13_18171311_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.589109</td>\n",
       "      <td>0.097408</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171393</td>\n",
       "      <td>chr13_18171393_T_C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.297030</td>\n",
       "      <td>0.473681</td>\n",
       "      <td>rs76065716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171453</td>\n",
       "      <td>chr13_18171453_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.623762</td>\n",
       "      <td>0.422176</td>\n",
       "      <td>rs9740229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171494</td>\n",
       "      <td>chr13_18171494_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.559406</td>\n",
       "      <td>0.427025</td>\n",
       "      <td>rs78165403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270599</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227085</td>\n",
       "      <td>chr16_90227085_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.341584</td>\n",
       "      <td>0.001972</td>\n",
       "      <td>rs200019921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270600</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227369</td>\n",
       "      <td>chr16_90227369_C_T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.470297</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270601</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227485</td>\n",
       "      <td>chr16_90227485_A_T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.103960</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270602</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227639</td>\n",
       "      <td>chr16_90227639_C_T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.410891</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270603</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90228078</td>\n",
       "      <td>chr16_90228078_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.193069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>rs367934211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8814835 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  SNP_Position         SV_SNP_Name Reference_Allele   \n",
       "0           chr13      17767059         P0000147722                N  \\\n",
       "1           chr13      18171311  chr13_18171311_A_G                A   \n",
       "2           chr13      18171393  chr13_18171393_T_C                T   \n",
       "3           chr13      18171453  chr13_18171453_G_A                G   \n",
       "4           chr13      18171494  chr13_18171494_A_G                A   \n",
       "...           ...           ...                 ...              ...   \n",
       "270599      chr16      90227085  chr16_90227085_A_G                A   \n",
       "270600      chr16      90227369  chr16_90227369_C_T                C   \n",
       "270601      chr16      90227485  chr16_90227485_A_T                A   \n",
       "270602      chr16      90227639  chr16_90227639_C_T                C   \n",
       "270603      chr16      90228078  chr16_90228078_C_A                C   \n",
       "\n",
       "       Alternate_Allele  Sample_AF  gnomAD_nfe_AF        dbSNP  \n",
       "0                 <INS>   0.128713       0.000000          NaN  \n",
       "1                     G   0.589109       0.097408          NaN  \n",
       "2                     C   0.297030       0.473681   rs76065716  \n",
       "3                     A   0.623762       0.422176    rs9740229  \n",
       "4                     G   0.559406       0.427025   rs78165403  \n",
       "...                 ...        ...            ...          ...  \n",
       "270599                G   0.341584       0.001972  rs200019921  \n",
       "270600                T   0.470297       0.000000          NaN  \n",
       "270601                T   0.103960       0.000293          NaN  \n",
       "270602                T   0.410891       0.000586          NaN  \n",
       "270603                A   0.193069       0.000000  rs367934211  \n",
       "\n",
       "[8814835 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## SV/SNP allele annotations done by Thomas Nalpathamkalam on April 10, 2023\n",
    "allele_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                            \"*_allele_freq.txt\"))\n",
    "allele_tables = [pd.read_table(file_path,\n",
    "                               header = 0,\n",
    "                               names = [\"Chromosome\",\n",
    "                                        \"SNP_Position\",\n",
    "                                        \"SV_SNP_Name\",\n",
    "                                        \"Reference_Allele\",\n",
    "                                        \"Alternate_Allele\",\n",
    "                                        \"Sample_AF\",\n",
    "                                        \"gnomAD_nfe_AF\",\n",
    "                                        \"dbSNP\"])\n",
    "                 for file_path in allele_file_paths]\n",
    "\n",
    "DF_ALLELE = pd.concat(allele_tables)\n",
    "DF_ALLELE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'transcript_id', 'transcript_type', 'transcript_name', 'tag', 'protein_id', 'db_xref', 'exon_number', 'exon_id']\n"
     ]
    }
   ],
   "source": [
    "## Gene/exon data from MANE.GRCh38.v1.0.ensembl_genomic.gtf extracted January 5, 2023\n",
    "\n",
    "DF_GTF = read_gtf(\"data/MANE.GRCh38.v1.0.ensembl_genomic.gtf\")\n",
    "DF_GTF = pd.DataFrame(data = DF_GTF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = DF_GTF[(DF_GTF[\"feature\"] == \"transcript\") | \\\n",
    "                 (DF_GTF[\"feature\"] == \"exon\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = df_gene[df_gene[\"tag\"].str.contains(\"MANE_Select\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = df_gene[[\"seqname\",\n",
    "                    \"start\", \n",
    "                    \"end\", \n",
    "                    \"strand\", \n",
    "                    \"feature\", \n",
    "                    \"gene_name\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "df_gene.columns = [\"chr\",\n",
    "                   \"start\",\n",
    "                   \"end\",\n",
    "                   \"strand\",\n",
    "                   \"feature\",\n",
    "                   \"gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### DATA WRANGLING ####\n",
    "## Subset allele data for only SV information\n",
    "df_sv_allele = DF_ALLELE[DF_ALLELE[\"SV_SNP_Name\"].str.startswith(\"P\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_allele = df_sv_allele.rename(columns = {\"SV_SNP_Name\" : \"SV_Name\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_allele = df_sv_allele.drop([\"Reference_Allele\",\n",
    "                                  \"SNP_Position\",\n",
    "                                  \"gnomAD_nfe_AF\",\n",
    "                                  \"dbSNP\"],\n",
    "                                 axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SV_Name</th>\n",
       "      <th>Alternate_Allele</th>\n",
       "      <th>Sample_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13</td>\n",
       "      <td>P0000147722</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.128713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>chr13</td>\n",
       "      <td>P0000147726</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.693069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chr13</td>\n",
       "      <td>P0000147727</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>0.846535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>chr13</td>\n",
       "      <td>P0000147729</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.495050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>chr13</td>\n",
       "      <td>P0000147734</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>0.113861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269874</th>\n",
       "      <td>chr16</td>\n",
       "      <td>P0000175630</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.178218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269900</th>\n",
       "      <td>chr16</td>\n",
       "      <td>P0000175632</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.435644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270213</th>\n",
       "      <td>chr16</td>\n",
       "      <td>P0000175638</td>\n",
       "      <td>&lt;DEL&gt;</td>\n",
       "      <td>0.173267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270309</th>\n",
       "      <td>chr16</td>\n",
       "      <td>P0000175646</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.103960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270572</th>\n",
       "      <td>chr16</td>\n",
       "      <td>P0000175650</td>\n",
       "      <td>&lt;INS&gt;</td>\n",
       "      <td>0.222772</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36295 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      SV_Name Alternate_Allele  Sample_AF\n",
       "0           chr13  P0000147722            <INS>   0.128713\n",
       "212         chr13  P0000147726            <INS>   0.693069\n",
       "225         chr13  P0000147727            <DEL>   0.846535\n",
       "315         chr13  P0000147729            <INS>   0.495050\n",
       "326         chr13  P0000147734            <DEL>   0.113861\n",
       "...           ...          ...              ...        ...\n",
       "269874      chr16  P0000175630            <INS>   0.178218\n",
       "269900      chr16  P0000175632            <INS>   0.435644\n",
       "270213      chr16  P0000175638            <DEL>   0.173267\n",
       "270309      chr16  P0000175646            <INS>   0.103960\n",
       "270572      chr16  P0000175650            <INS>   0.222772\n",
       "\n",
       "[36295 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sv_allele = df_sv_allele.drop_duplicates()\n",
    "df_sv_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset allele data for only SNP information\n",
    "df_snp_allele = DF_ALLELE[~DF_ALLELE[\"SV_SNP_Name\"].str.startswith(\"P\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snp_allele = df_snp_allele.rename(columns = {\"SV_SNP_Name\" : \"SNP_Name\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>SNP_Position</th>\n",
       "      <th>SNP_Name</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Alternate_Allele</th>\n",
       "      <th>Sample_AF</th>\n",
       "      <th>gnomAD_nfe_AF</th>\n",
       "      <th>dbSNP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171311</td>\n",
       "      <td>chr13_18171311_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.589109</td>\n",
       "      <td>0.097408</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171393</td>\n",
       "      <td>chr13_18171393_T_C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.297030</td>\n",
       "      <td>0.473681</td>\n",
       "      <td>rs76065716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171453</td>\n",
       "      <td>chr13_18171453_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.623762</td>\n",
       "      <td>0.422176</td>\n",
       "      <td>rs9740229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171494</td>\n",
       "      <td>chr13_18171494_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.559406</td>\n",
       "      <td>0.427025</td>\n",
       "      <td>rs78165403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr13</td>\n",
       "      <td>18171497</td>\n",
       "      <td>chr13_18171497_G_C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.430693</td>\n",
       "      <td>0.440541</td>\n",
       "      <td>rs75406701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270599</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227085</td>\n",
       "      <td>chr16_90227085_A_G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.341584</td>\n",
       "      <td>0.001972</td>\n",
       "      <td>rs200019921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270600</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227369</td>\n",
       "      <td>chr16_90227369_C_T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.470297</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270601</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227485</td>\n",
       "      <td>chr16_90227485_A_T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.103960</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270602</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90227639</td>\n",
       "      <td>chr16_90227639_C_T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.410891</td>\n",
       "      <td>0.000586</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270603</th>\n",
       "      <td>chr16</td>\n",
       "      <td>90228078</td>\n",
       "      <td>chr16_90228078_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.193069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>rs367934211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8518382 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome  SNP_Position            SNP_Name Reference_Allele   \n",
       "1           chr13      18171311  chr13_18171311_A_G                A  \\\n",
       "2           chr13      18171393  chr13_18171393_T_C                T   \n",
       "3           chr13      18171453  chr13_18171453_G_A                G   \n",
       "4           chr13      18171494  chr13_18171494_A_G                A   \n",
       "5           chr13      18171497  chr13_18171497_G_C                G   \n",
       "...           ...           ...                 ...              ...   \n",
       "270599      chr16      90227085  chr16_90227085_A_G                A   \n",
       "270600      chr16      90227369  chr16_90227369_C_T                C   \n",
       "270601      chr16      90227485  chr16_90227485_A_T                A   \n",
       "270602      chr16      90227639  chr16_90227639_C_T                C   \n",
       "270603      chr16      90228078  chr16_90228078_C_A                C   \n",
       "\n",
       "       Alternate_Allele  Sample_AF  gnomAD_nfe_AF        dbSNP  \n",
       "1                     G   0.589109       0.097408          NaN  \n",
       "2                     C   0.297030       0.473681   rs76065716  \n",
       "3                     A   0.623762       0.422176    rs9740229  \n",
       "4                     G   0.559406       0.427025   rs78165403  \n",
       "5                     C   0.430693       0.440541   rs75406701  \n",
       "...                 ...        ...            ...          ...  \n",
       "270599                G   0.341584       0.001972  rs200019921  \n",
       "270600                T   0.470297       0.000000          NaN  \n",
       "270601                T   0.103960       0.000293          NaN  \n",
       "270602                T   0.410891       0.000586          NaN  \n",
       "270603                A   0.193069       0.000000  rs367934211  \n",
       "\n",
       "[8518382 rows x 8 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_snp_allele = df_snp_allele.drop_duplicates()\n",
    "df_snp_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset DF_ANNO_FULL with columns that can be displayed to the public. Add SV AF to this table as well.\n",
    "df_sv_anno = DF_ANNO_FULL\n",
    "\n",
    "df_sv_anno = df_sv_anno.merge(df_sv_allele.iloc[:,[1,2,3]],\n",
    "                              left_on = \"ID\",\n",
    "                              right_on = \"SV_Name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_anno = df_sv_anno.drop(columns = \"Alternate_Allele\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_range = list(range(1,(df_sv_anno.shape[1]-2)))\n",
    "columns_range.insert(0, (df_sv_anno.shape[1]-2))\n",
    "columns_range.insert(6, (df_sv_anno.shape[1]-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_anno = df_sv_anno.iloc[:,columns_range]\n",
    "df_sv_anno = df_sv_anno.rename(columns = {\"SV_Name\" : \"SV Name\",\n",
    "                                          \"Sample_AF\" : \"SV Sample AF\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of SV table for easier access of data required - join the dataframes based on SV names\n",
    "df_sv_join = DF_LD.merge(df_anno,\n",
    "                         on = \"SV_Name\",\n",
    "                         how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_join = df_sv_join[[\"Chromosome\",\n",
    "                         \"SV_Name\",\n",
    "                         \"SV_Start\",\n",
    "                         \"SV_End\",\n",
    "                         \"SNP_Name\",\n",
    "                         \"SNP_Position\",\n",
    "                         \"R2\",\n",
    "                         \"D'\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merging SV allele information with the LD dataset\n",
    "df_sv_join = df_sv_join.merge(df_sv_allele,\n",
    "                              on = [\"Chromosome\",\n",
    "                                    \"SV_Name\"],\n",
    "                              how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_join = df_sv_join.rename(columns = {\"Alternate_Allele\" : \"SV_Type\",\n",
    "                                          \"Sample_AF\" : \"SV_AF\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_join.loc[df_sv_join[\"SV_Type\"] == \"<DEL>\", \"SV_Type\"] = \"Deletion\"\n",
    "df_sv_join.loc[df_sv_join[\"SV_Type\"] == \"<INS>\", \"SV_Type\"] = \"Insertion\"\n",
    "df_sv_join.loc[df_sv_join[\"SV_Type\"] == \"<DUP>\", \"SV_Type\"] = \"Duplication\"\n",
    "df_sv_join.loc[df_sv_join[\"SV_Type\"] == \"<INV>\", \"SV_Type\"] = \"Inversion\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_join = df_sv_join.iloc[:, [0,1,2,3,8,9,4,5,6,7]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merging SNP allele information with the LD dataset\n",
    "df_sv_snp_join = df_sv_join.merge(df_snp_allele,\n",
    "                                  on = [\"Chromosome\",\n",
    "                                        \"SNP_Name\",\n",
    "                                        \"SNP_Position\"],\n",
    "                                        how = \"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_snp_join = df_sv_snp_join.drop(\"SNP_Name\", axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_snp_join = df_sv_snp_join.iloc[:, [0,1,2,3,4,5,13,6,9,10,11,12,7,8]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sv_snp_join = df_sv_snp_join.rename(columns = {\"dbSNP\" : \"SNP_Name_dbSNP\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of a full dataframe to make data extraction much easier\n",
    "df_full_join = df_sv_snp_join.merge(df_gwas,\n",
    "                                    how = \"left\",\n",
    "                                    on = [\"Chromosome\",\n",
    "                                          \"SNP_Position\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows where SV AF is NA - these are SVs with AF < 0.1\n",
    "df_full_join = df_full_join[df_full_join[\"SV_AF\"].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change p-values to log10 scale\n",
    "    # Smallest pval was 1-e323. After looking at the distribution, all pvalues under 1e-50 were changed to 1e-50.\n",
    "df_full_join.loc[df_full_join[\"P-Value\"] < 1e-50, \"P-Value\"] = 1e-50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_full_join[\"P-Value_log10\"] = np.log10(df_full_join[\"P-Value\"])\n",
    "df_full_join[\"P-Value_log10\"] = -df_full_join[\"P-Value_log10\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset df_anno for the SVs in df_full_foin\n",
    "sv_names = df_full_join[\"SV_Name\"].unique()\n",
    "df_anno_subset = df_anno[df_anno[\"SV_Name\"].isin(sv_names)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of phenotype dictionary to encompass SV for each gene - computationally faster when creating SV table\n",
    "dict_pheno = {}\n",
    "\n",
    "df_dict = df_full_join[[\"SV_Name\", \"SV_Start\", \"SV_End\", \"Phenotype\"]]\n",
    "df_dict = df_dict.to_dict(\"records\")\n",
    "\n",
    "for row in df_dict:\n",
    "    if row[\"Phenotype\"] not in dict_pheno:\n",
    "        dict_pheno[row[\"Phenotype\"]] = []\n",
    "        dict_pheno[row[\"Phenotype\"]].append(row[\"SV_Name\"])\n",
    "    else:\n",
    "        dict_pheno[row[\"Phenotype\"]].append(row[\"SV_Name\"])\n",
    "\n",
    "for k, v in dict_pheno.items():\n",
    "    dict_pheno[k] = set(v)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of y-coordinates to prevent overlap in genes when plotted\n",
    "def assign_y_coordinates(df = df_gene,\n",
    "                         y_start = -0.3,\n",
    "                         y_increment = 0.5,\n",
    "                         col_name = \"y_coord\"):\n",
    "    # Priority queue to keep track of the end positions of genes for each y-coordinate\n",
    "    df = df[df[\"feature\"] == \"transcript\"]\n",
    "    df = df.sort_values(by = [\"chr\", \"start\"])\n",
    "    df = df.reset_index(drop = True) #reset index \n",
    "     # Dictionary to store priority queues for each chromosome\n",
    "    chromosome_heaps = {}\n",
    "\n",
    "    # List to store the y-coordinate for each gene annotation\n",
    "    y_coordinates = [y_start] * len(df)\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        chrom = row['chr']\n",
    "\n",
    "        # If the chromosome is not in the dictionary, initialize its heap\n",
    "        if chrom not in chromosome_heaps:\n",
    "            chromosome_heaps[chrom] = []\n",
    "\n",
    "        # Work with the heap for the current chromosome\n",
    "        heap = chromosome_heaps[chrom]\n",
    "\n",
    "        # If there's an available y-coordinate whose gene ends before the current gene starts,\n",
    "        # reuse that y-coordinate and update its gene end position\n",
    "        if heap and heap[0][0] <= row['start']:\n",
    "            previous_gene_end, y_coordinate = heapq.heappop(heap)\n",
    "            heapq.heappush(heap, (row['end'], y_coordinate))\n",
    "        else:\n",
    "            # If no y-coordinate is available, create a new one\n",
    "            if heap:\n",
    "                y_coordinate = min(y_coordinate for _, y_coordinate in heap) - y_increment\n",
    "            else:\n",
    "                y_coordinate = y_start\n",
    "            heapq.heappush(heap, (row['end'], y_coordinate))\n",
    "\n",
    "        # Assign the y-coordinate to the gene annotation\n",
    "        y_coordinates[index] = y_coordinate\n",
    "\n",
    "    # Add the y-coordinates to the dataframe\n",
    "    df[col_name] = y_coordinates\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_transcript = assign_y_coordinates(df = df_gene,\n",
    "                                     y_start = -0.3,\n",
    "                                     y_increment = 0.15,\n",
    "                                     col_name = \"y_coord_nopheno\")\n",
    "\n",
    "df_transcript = assign_y_coordinates(df = df_transcript,\n",
    "                                     y_start = -6.25,\n",
    "                                     y_increment = 4,\n",
    "                                     col_name = \"y_coord_pheno\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add icon for help popup\n",
    "question_icon = html.I(className = 'bi bi-question-diamond',\n",
    "                       style = {\"display\" : \"inline-block\",\n",
    "                                \"color\" : \"lightsalmon\",\n",
    "                                \"font-size\" : \"70%\"})\n",
    "                    #    id = 'question-icon')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### FILTERS ####\n",
    "\n",
    "# Make dropdown box for chromosomes and textbox for genomic range\n",
    "REGION_LABEL = html.P(\"Genomic Region\")\n",
    "\n",
    "chromosomes = [\"Any\", *range(1, 23), \"X\", \"Y\"]\n",
    "\n",
    "CHR_DD = dcc.Dropdown(id = 'chromosome-dropdown',\n",
    "                      placeholder = \"Select a chromosome\",\n",
    "                      options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in chromosomes])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "EMPTY_SPACE_REGION = html.Div(id = 'empty-space-region')\n",
    "\n",
    "RANGE_START_TB = dcc.Input(id = 'range-start-textbox',\n",
    "                           type = \"number\",\n",
    "                           placeholder = \"Start bp\",\n",
    "                           min = 1,\n",
    "                           debounce = True,\n",
    "                           className = 'class-range')\n",
    "\n",
    "RANGE_END_TB = dcc.Input(id = 'range-end-textbox',\n",
    "                         type = \"number\",\n",
    "                         placeholder = \"End bp\",\n",
    "                         min = 2,\n",
    "                         debounce = True,\n",
    "                         className = 'class-range')\n",
    "\n",
    "\n",
    "REGION_DIV = html.Div(id = 'chromosome-dropdown-div',\n",
    "                      children = [REGION_LABEL,\n",
    "                                  CHR_DD,\n",
    "                                  EMPTY_SPACE_REGION,\n",
    "                                  RANGE_START_TB,\n",
    "                                  RANGE_END_TB],\n",
    "                      className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for phenotypes\n",
    "# PHENO_LINE = html.Div(id = 'pheno-line-div')\n",
    "\n",
    "PHENO_LABEL = html.Div([html.P([\"Phenotype\",\n",
    "                                question_icon],\n",
    "                               id = 'pheno-label'),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Select phenotype of interest availible in GWAS Catalog.\\n\\nSVs are selected if they have an LD calculation with an associated SNP for the phenotype.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'pheno-label',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'pheno-label-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes = sorted(list(df_full_join[\"Phenotype\"].unique()))\n",
    "phenotypes.insert(0, \"Any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "PHENO_DD = dcc.Dropdown(id = 'phenotype-dropdown',\n",
    "                        placeholder = \"Select a phenotype of interest\",\n",
    "                        options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in phenotypes],\n",
    "                        optionHeight = 75)\n",
    "\n",
    "PHENO_DD_DIV = html.Div(id = 'phenotype-dropdown-div',\n",
    "                        children = [# PHENO_LINE,\n",
    "                                    PHENO_LABEL,\n",
    "                                    PHENO_DD],\n",
    "                        className = 'class-section-filter-div')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for genes\n",
    "GENE_LABEL = html.P(\"Gene\")\n",
    "\n",
    "genes = sorted(list(set(df_gene[\"gene\"])))\n",
    "\n",
    "GENE_DD = dcc.Dropdown(id = 'gene-dropdown',\n",
    "                       placeholder = \"Select a gene of interest\",\n",
    "                       options = [{\"label\" : str(i),\n",
    "                                   \"value\" : str(i)} for i in genes])\n",
    "\n",
    "GENE_DD_DIV = html.Div(id = 'gene-dropdown-div',\n",
    "                       children = [GENE_LABEL,\n",
    "                                   GENE_DD],\n",
    "                       className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create a tab section for filter between gene and genomic region\n",
    "TABS_FILTER_DIV = html.Div([dbc.Tabs([dbc.Tab(children = [GENE_DD_DIV],\n",
    "                                         label = \"by Gene\",\n",
    "                                         tab_id = \"tab-gene\"),\n",
    "                                     dbc.Tab(children = [REGION_DIV],\n",
    "                                             label = \"by Region\",\n",
    "                                             tab_id = \"tab-region\")],\n",
    "                                     id = \"tab-filters\",\n",
    "                                     active_tab = \"tab-gene\")],\n",
    "                           className = 'class-section-filter-div')\n",
    "\n",
    "\n",
    "FILTER_DIV = html.Div(id = 'filter-div',\n",
    "                      children = [TABS_FILTER_DIV,\n",
    "                                  PHENO_DD_DIV],\n",
    "                      className = 'class-filter-div')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### TABLE ####\n",
    "\n",
    "# Make DataTable\n",
    "def make_table(chrom = chromosomes,\n",
    "               range_start = 0,\n",
    "               range_end = 250000000,\n",
    "               gene = genes,\n",
    "               pheno = phenotypes):\n",
    "\n",
    "    global tab_show\n",
    "\n",
    "    # Filter by chromosome\n",
    "    if isinstance(chrom, list):\n",
    "        sv_list_chr = list(df_anno_subset[\"SV_Name\"])\n",
    "    else:\n",
    "        chrom = str(\"chr\" + chrom)\n",
    "        sv_list_chr = df_anno_subset[(df_anno_subset[\"Chromosome\"] == chrom)]\n",
    "        sv_list_chr = list(sv_list_chr[\"SV_Name\"])\n",
    "\n",
    "\n",
    "    # Filter by range\n",
    "    sv_list_range = df_anno_subset[(df_anno_subset[\"SV_Start\"] >= range_start) & \\\n",
    "                                   (df_anno_subset[\"SV_End\"] <= range_end)]\n",
    "\n",
    "    sv_list_range = list(sv_list_range[\"SV_Name\"])\n",
    "\n",
    "\n",
    "    # Filter by phenotype\n",
    "    if isinstance(pheno, list):\n",
    "        sv_list_pheno = list(df_anno_subset[\"SV_Name\"])\n",
    "    else:\n",
    "        sv_list_pheno = dict_pheno[pheno]\n",
    "\n",
    "\n",
    "    # Filter by gene\n",
    "    if isinstance(gene, list):\n",
    "        sv_list_gene = list(df_anno_subset[\"SV_Name\"])\n",
    "    else:\n",
    "        df_1gene = df_gene[(df_gene[\"feature\"] == \"transcript\") & \\\n",
    "                           (df_gene[\"gene\"] == gene)]\n",
    "        start = df_1gene[\"start\"].min()\n",
    "        end = df_1gene[\"end\"].max()\n",
    "        chrom = df_1gene.iat[0,0]\n",
    "        strand = df_1gene.iat[0,3]\n",
    "\n",
    "        sv_list_gene = df_sv_anno[(df_sv_anno[\"Chromosome\"] == chrom) & \\\n",
    "                                  (df_sv_anno[\"Start\"] >= (start - 100000)) & \\\n",
    "                                  (df_sv_anno[\"End\"] <= (end + 100000))]\n",
    "\n",
    "        sv_list_gene = list(sv_list_gene[\"SV Name\"])\n",
    "\n",
    "\n",
    "    # Find overlaps sv's from all lists\n",
    "    sv_list_complete = set(sv_list_pheno).intersection(sv_list_chr, sv_list_gene, sv_list_range)\n",
    "\n",
    "    tab_show = df_sv_anno[df_sv_anno[\"SV Name\"].isin(sv_list_complete)]\n",
    "    tab_show = tab_show[[\"SV Name\", \"Chromosome\", \"Start\", \"End\", \"Type\", \"Length\", \"SV Sample AF\"]]\n",
    "    tab_show = tab_show.rename(columns = {\"SV Name\" : \"ID\",\n",
    "                                          \"Chromosome\" : \"Chrom\",\n",
    "                                          \"Length\" : \"Size (bp)\",\n",
    "                                          \"SV Sample AF\" : \"AF\"})\n",
    "\n",
    "    return dash_table.DataTable(id ='strvar-table',\n",
    "                                data = tab_show.to_dict(\"records\"),\n",
    "                                columns = [{'id': c, 'name': c} for c in tab_show.loc[:,[ \"Chrom\", \"Start\", \"End\", \"Type\", \"Size (bp)\", \"AF\"]]],\n",
    "                                page_size = 100,\n",
    "                                fixed_rows = {'headers': True},\n",
    "                                style_table = {'height': '400px',\n",
    "                                               'overflowY': 'auto'},\n",
    "                                style_cell = {'textAlign': 'center',\n",
    "                                              'font-family': 'Nunito Sans',\n",
    "                                              'backgroundColor': 'whitesmoke',\n",
    "                                              'minWidth' : '100px'},\n",
    "                                style_header = {'backgroundColor': 'turquoise',\n",
    "                                                'fontWeight': 'bold'},\n",
    "                                row_selectable = 'single',\n",
    "                                selected_rows = [])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SV annotation table\n",
    "def make_annotation_table(sv = \"None\"):\n",
    "    if sv == \"None\":\n",
    "        df_anno_subset_sv = pd.DataFrame()\n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = df_anno_subset_sv.to_dict(\"records\"),\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_anno_subset_sv = df_sv_anno[df_sv_anno[\"SV Name\"] == sv]\n",
    "        df_anno_subset_sv = df_anno_subset_sv.transpose()\n",
    "        df_anno_subset_sv = df_anno_subset_sv.rename_axis(\" \").reset_index()\n",
    "        df_anno_subset_sv.columns = [\"Header\", \"Information\"]\n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = df_anno_subset_sv.to_dict(\"records\"),\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowX': 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell = {'textAlign': 'left',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': \"Header\"},\n",
    "                                                              'backgroundColor': 'turquoise',\n",
    "                                                              'fontWeight': 'bold',\n",
    "                                                              'maxWidth' : '200px',\n",
    "                                                              'whiteSpace' : 'normal',\n",
    "                                                              'height' : 'auto'}],\n",
    "                                    style_header = {'display' : 'none',\n",
    "                                                    'height' : '0px'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "DTABLE = html.Div(id = 'strvar-table-div',\n",
    "                  children = [make_table()],\n",
    "                  className = 'sv-table')\n",
    "\n",
    "# Create a header for the SV table\n",
    "DTABLE_HEADER = html.Div([html.H4([\"SV Information\",\n",
    "                                   question_icon],\n",
    "                                   id = 'dtable-header',\n",
    "                                   style = {'textAlign' : 'center',\n",
    "                                            'color' : '#2F2E2D',\n",
    "                                            'height' : '30 px',\n",
    "                                            'text-transform': 'none'}),\n",
    "                          dbc.Tooltip(dcc.Markdown(\"List of SVs subsetted by the filters applied.\\n\\nSelect an SV to visualize LD with GWAS-associated SNPs.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                      target = 'dtable-header',\n",
    "                                      placement = \"auto\",\n",
    "                                      delay = {\"show\" : 500,\n",
    "                                               \"hide\" : 50},\n",
    "                                      id = 'dtable-header-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "DTABLE_TEXTBOX = html.Div(id = 'dtable-textbox')\n",
    "\n",
    "\n",
    "DTABLE_DIV = html.Div(id = 'dtable-div',\n",
    "                      children = [DTABLE_HEADER,\n",
    "                                  DTABLE,\n",
    "                                  DTABLE_TEXTBOX],\n",
    "                      className = 'class-filter-div')\n",
    "\n",
    "\n",
    "ANNOTABLE = html.Div(id = 'anno-table-div',\n",
    "                     children = [make_annotation_table()],\n",
    "                     className = 'sv-anno-table')\n",
    "\n",
    "\n",
    "# Create the SV details collapse and table\n",
    "SV_TABLE_BUTTON = dbc.Button(\"SV Annotations\",\n",
    "                             id = \"sv-button-collapse\",\n",
    "                             color = \"secondary\",\n",
    "                             outline = True,\n",
    "                             size = \"sm\",\n",
    "                             n_clicks = 0)\n",
    "\n",
    "\n",
    "# Create the filter division\n",
    "ALL_FILTER_DIV = html.Div(id = 'all-filter-div',\n",
    "                          children = [FILTER_DIV,\n",
    "                                      DTABLE_DIV],\n",
    "                                      style = {'display' : 'flex',\n",
    "                                               'flexWrap' : \"wrap\",\n",
    "                                               'justifyContent' : \"space-evenly\",\n",
    "                                               'alignContent' : 'space-around'},\n",
    "                                               className = 'sections')\n",
    "\n",
    "SV_TABLE_BUTTON_DIV = html.Div(id = \"sv-button-div\",\n",
    "                               children = [SV_TABLE_BUTTON])\n",
    "\n",
    "\n",
    "SV_TABLE_COLLAPSE = dbc.Collapse(children = [ANNOTABLE],\n",
    "                                 id = \"sv-table-collapse\",\n",
    "                                 is_open = False)\n",
    "\n",
    "SV_TABLE = html.Div(id = 'sv-table-collapse-div',\n",
    "                    children = [SV_TABLE_COLLAPSE],\n",
    "                    style = {'display' : 'flex',\n",
    "                             'flexWrap' : \"wrap\",\n",
    "                             'justifyContent' : \"space-evenly\",\n",
    "                             'alignContent' : 'space-around'})\n",
    "\n",
    "\n",
    "SV_TABLE_DIV = html.Div(id = 'sv-table-div',\n",
    "                        children = [SV_TABLE,\n",
    "                                    SV_TABLE_BUTTON_DIV],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### PLOT ####\n",
    "\n",
    "# Make scatter plot to compare r2 and D' values as seen in DataTable\n",
    "def make_plot(sv = \"None\",\n",
    "              pheno = \"None\",\n",
    "              toggle = False,\n",
    "              toggle_pheno = False):\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage_value = \"D'\"\n",
    "        size_value = \"R2\"\n",
    "        legend_title = \"D'\"\n",
    "    else:\n",
    "        linkage_value = \"R2\"\n",
    "        size_value = \"D'\"\n",
    "        legend_title = \"r<sup>2</sup>\"\n",
    "\n",
    "    global tab, sv_start_plot, sv_end_plot, extra_points, extra_points_pheno\n",
    "\n",
    "    if sv == \"None\":\n",
    "        tab = pd.DataFrame(columns = [\"SNP Position\",\n",
    "                                      \"-log10(P-Value)\"])\n",
    "        sc_plot = px.scatter(data_frame = tab,\n",
    "                             x = \"SNP Position\",\n",
    "                             y = \"-log10(P-Value)\",\n",
    "                             labels = {\"-log10(P-Value)\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                       \"SNP Position\" : \"Position in hg38 (bp)\"},\n",
    "                             width = 850,\n",
    "                             height = 650,\n",
    "                            #  marginal_x = \"rug\",\n",
    "                             template = \"ggplot2\")\n",
    "\n",
    "        sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                              autosize = False,\n",
    "                              font_family = \"Nunito Sans\",\n",
    "                              font_size = 20)\n",
    "    else:\n",
    "        tab = df_full_join[df_full_join[\"SV_Name\"] == sv]\n",
    "\n",
    "        # Display +/- 1KB from SV position\n",
    "        sv_start = df_anno_subset[df_anno_subset[\"SV_Name\"] == sv][\"SV_Start\"].values[0]\n",
    "        sv_end = df_anno_subset[df_anno_subset[\"SV_Name\"] == sv][\"SV_End\"].values[0]\n",
    "\n",
    "        sv_start_plot = sv_start - 1000\n",
    "        sv_end_plot = sv_end + 1000\n",
    "\n",
    "\n",
    "        if pheno == \"None\":\n",
    "            tab_lab = tab.to_dict(\"records\")\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                label_pheno = row[\"Phenotype\"]\n",
    "                label_snp_name = row[\"SNP_Name_dbSNP\"]\n",
    "                label_chr = row[\"Chromosome\"]\n",
    "                label_snp_pos = row[\"SNP_Position\"]\n",
    "                label_pval = row[\"P-Value\"]\n",
    "                label_r2 = row[\"R2\"]\n",
    "                label_d = row[\"D'\"]\n",
    "                label_svname = row[\"SV_Name\"]\n",
    "                label_svstart = row[\"SV_Start\"]\n",
    "                label_svend = row[\"SV_End\"]\n",
    "                label_svtype = row[\"SV_Type\"]\n",
    "\n",
    "                label = [label_pheno, label_snp_name, label_chr, label_snp_pos, label_pval, label_r2, label_d, label_svname, label_svstart, label_svend, label_svtype]\n",
    "\n",
    "                return label\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab,\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = linkage_value,\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"R2\" : \"r<sup>2</sup>\"},\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                #  marginal_x = \"rug\",\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(marker = dict(color = \"rgba(57,201,187,0.5)\",\n",
    "                                                line = dict(color = \"rgba(57,201,187,1)\"),\n",
    "                                                size = 10),\n",
    "                                  hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + '<br>SNP Name: %{customdata[1]}' + '<br>Chromosome: %{customdata[2]}' + '<br>SNP Position: %{customdata[3]}' + '<br>P-Value: %{customdata[4]}' + '<br>r2: %{customdata[5]}' + \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab])\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                  autosize = False,\n",
    "                                  font_family = \"Nunito Sans\",\n",
    "                                  font_size = 20)\n",
    "                                #   xaxis_range = [sv_start_plot, sv_end_plot],\n",
    "                                #   yaxis_range = [-0.6, 1])\n",
    "\n",
    "             # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = min(tab[\"SNP_Position\"]) - 100000\n",
    "            max_SNP = max(tab[\"SNP_Position\"]) + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = min(tab[\"SNP_Position\"]) - 100000\n",
    "                max_SNP = max(tab[\"SNP_Position\"]) + 100000\n",
    "\n",
    "\n",
    "            chromo = list(set(tab[\"Chromosome\"]))[0]\n",
    "            svname = list(set(tab[\"SV_Name\"]))[0]\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -0.1,\n",
    "                              y1 = -0.15,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                           width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.01,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-0.07]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + '<br>Chromosome: %{customdata[2]}' + '<br>Start: %{customdata[8]}' + '<br>End: %{customdata[9]}' + '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Add extra points from df_sv with R2 and D' values\n",
    "            extra_points = df_sv_snp_join[df_sv_snp_join[\"SV_Name\"] == sv]\n",
    "\n",
    "            extra_points = extra_points[~extra_points[\"SNP_Name_dbSNP\"].isin(tab[\"SNP_Name_dbSNP\"])]\n",
    "\n",
    "             # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dict(\"records\")\n",
    "\n",
    "            def extra_label_maker(row):\n",
    "\n",
    "                label_snp_name = row[\"SNP_Name_dbSNP\"]\n",
    "                label_chr = row[\"Chromosome\"]\n",
    "                label_snp_pos = row[\"SNP_Position\"]\n",
    "                label_r2 = row[\"R2\"]\n",
    "                label_d = row[\"D'\"]\n",
    "\n",
    "                label = [label_snp_name, label_chr, label_snp_pos, label_r2, label_d]\n",
    "\n",
    "                return label\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points[\"SNP_Position\"],\n",
    "                                           y = extra_points[linkage_value],\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>SNP Name: %{customdata[0]}</b>' +'<br>Chromosome: %{customdata[1]}' + '<br>SNP Position: %{customdata[2]}' + '<br>r2: %{customdata[3]}' + \"<br>D': %{customdata[4]}\",\n",
    "                                           customdata = [extra_label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"Not in \\nGWAS Catalog\"))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene[(df_gene[\"chr\"] == chromo) & \\\n",
    "                            (df_gene[\"start\"] >= min_SNP) & \\\n",
    "                            (df_gene[\"end\"] <= max_SNP)]\n",
    "\n",
    "            exons_genes = exons.groupby([\"chr\", \"gene\", \"strand\"])[\"start\"].min()\n",
    "            exons_genes = exons_genes.to_frame().reset_index()\n",
    "\n",
    "            exons_end = exons.groupby([\"chr\", \"gene\", \"strand\"])[\"end\"].max()\n",
    "            exons_end = exons_end.to_frame().reset_index()\n",
    "\n",
    "            exons_genes[\"end\"] = exons_end[\"end\"]\n",
    "\n",
    "            # gene_line_position = -0.3\n",
    "\n",
    "            for i in range(0,len(exons_genes)):\n",
    "                row = exons_genes.loc[exons_genes.index[i]]\n",
    "                gene_line_position = float(df_transcript.loc[df_transcript[\"gene\"] == row[\"gene\"], \"y_coord_nopheno\"].values[0]) \n",
    "\n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                x0 = (row[\"start\"]),\n",
    "                                x1 = (row[\"end\"]),\n",
    "                                y0 = gene_line_position,\n",
    "                                y1 = gene_line_position,\n",
    "                                line_width = 3,\n",
    "                                line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = gene_line_position + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = gene_line_position + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons[exons[\"feature\"] == \"exon\"]\n",
    "\n",
    "            for i in range(0,len(exons_exons)):\n",
    "                row = exons_exons.loc[exons_exons.index[i]]\n",
    "                gene_line_position = float(df_transcript.loc[df_transcript[\"gene\"] == row[\"gene\"], \"y_coord_nopheno\"].values[0]) \n",
    "                sc_plot.add_shape(type = \"rect\",\n",
    "                                x0 = (row[\"start\"]),\n",
    "                                x1 = (row[\"end\"]),\n",
    "                                y0 = gene_line_position + 0.020,\n",
    "                                y1 = gene_line_position - 0.020,\n",
    "                                line = dict(color = \"mediumseagreen\",\n",
    "                                            width = 2),\n",
    "                                fillcolor = \"mediumseagreen\")\n",
    "\n",
    "\n",
    "        else:\n",
    "            tab = tab[tab[\"Phenotype\"] == pheno]\n",
    "\n",
    "            tab_lab = tab.to_dict(\"records\")\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                label_pheno = row[\"Phenotype\"]\n",
    "                label_snp_name = row[\"SNP_Name_dbSNP\"]\n",
    "                label_chr = row[\"Chromosome\"]\n",
    "                label_snp_pos = row[\"SNP_Position\"]\n",
    "                label_pval = row[\"P-Value\"]\n",
    "                label_r2 = row[\"R2\"]\n",
    "                label_d = row[\"D'\"]\n",
    "                label_svname = row[\"SV_Name\"]\n",
    "                label_svstart = row[\"SV_Start\"]\n",
    "                label_svend = row[\"SV_End\"]\n",
    "                label_svtype = row[\"SV_Type\"]\n",
    "\n",
    "                label = [label_pheno, label_snp_name, label_chr, label_snp_pos, label_pval, label_r2, label_d, label_svname, label_svstart, label_svend, label_svtype]\n",
    "\n",
    "                return label\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab,\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = \"P-Value_log10\",\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"P-Value_log10\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                           linkage_value : legend_title},\n",
    "                                 color = linkage_value,\n",
    "                                #  color_continuous_scale = px.colors.sequential.Viridis_r,\n",
    "                                color_continuous_scale = [(0, \"#4643CE\"),\n",
    "                                                        (0.2, \"#4643CE\"),\n",
    "                                                        (0.2, \"#90CAEE\"),\n",
    "                                                        (0.4, \"#90CAEE\"),\n",
    "                                                        (0.4, \"#46E17C\"),\n",
    "                                                        (0.6, \"#46E17C\"),\n",
    "                                                        (0.6, \"#EEC06B\"),\n",
    "                                                        (0.8, \"#EEC06B\"),\n",
    "                                                        (0.8, \"#CE5C5C\"),\n",
    "                                                        (1, \"#CE5C5C\")],\n",
    "                                 range_color = [0, 1],\n",
    "                                #  size = size_value,\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                #  marginal_x = \"rug\",\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + '<br>SNP Name: %{customdata[1]}' + '<br>Chromosome: %{customdata[2]}' + '<br>SNP Position: %{customdata[3]}' + '<br>P-Value: %{customdata[4]}' + '<br>r2: %{customdata[5]}' + \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab],\n",
    "                                  marker = dict(size = 10))\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                 autosize = True,\n",
    "                                 font_family = \"Nunito Sans\",\n",
    "                                 font_size = 20,\n",
    "                                #  xaxis_range = [sv_start_plot, sv_end_plot],\n",
    "                                #  yaxis_range = [-20, 50],\n",
    "                                 coloraxis_colorbar = dict(lenmode = \"pixels\",\n",
    "                                                           len = 150,\n",
    "                                                           yanchor = \"top\",\n",
    "                                                           y = 1,\n",
    "                                                           dtick = 0.2))\n",
    "\n",
    "             # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = min(tab[\"SNP_Position\"]) - 100000\n",
    "            max_SNP = max(tab[\"SNP_Position\"]) + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = min(tab[\"SNP_Position\"]) - 100000\n",
    "                max_SNP = max(tab[\"SNP_Position\"]) + 100000\n",
    "\n",
    "\n",
    "            chromo = list(set(tab[\"Chromosome\"]))[0]\n",
    "            svname = list(set(tab[\"SV_Name\"]))[0]\n",
    "\n",
    "            # Add points from gwas for pheno, without r2 and D' values\n",
    "            extra_points = df_gwas[(df_gwas[\"Chromosome\"] == chromo) & \\\n",
    "                                   (df_gwas[\"SNP_Position\"] >= min_SNP) & \\\n",
    "                                   (df_gwas[\"SNP_Position\"] <= max_SNP) & \\\n",
    "                                   (df_gwas[\"Phenotype\"] == pheno)]\n",
    "\n",
    "            extra_points = extra_points[~extra_points[\"SNP_Name_GWAS\"].isin(tab[\"SNP_Name_GWAS\"])]\n",
    "\n",
    "            extra_points.loc[extra_points[\"P-Value\"] < 1e-50, \"P-Value\"] = 1e-50\n",
    "\n",
    "            extra_points[\"P-Value_log10\"] = np.log10(extra_points[\"P-Value\"])\n",
    "            extra_points[\"P-Value_log10\"] = -extra_points[\"P-Value_log10\"]\n",
    "\n",
    "            # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dict(\"records\")\n",
    "\n",
    "            def label_maker(row):\n",
    "\n",
    "                label_pheno = row[\"Phenotype\"]\n",
    "                label_snp_name = row[\"SNP_Name_GWAS\"]\n",
    "                label_chr = row[\"Chromosome\"]\n",
    "                label_snp_pos = row[\"SNP_Position\"]\n",
    "                label_pval = row[\"P-Value\"]\n",
    "\n",
    "                label = [label_pheno, label_snp_name, label_chr, label_snp_pos, label_pval]\n",
    "\n",
    "                return label\n",
    "\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points[\"SNP_Position\"],\n",
    "                                           y = extra_points[\"P-Value_log10\"],\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + '<br>SNP Name: %{customdata[1]}' + '<br>Chromosome: %{customdata[2]}' + '<br>SNP Position: %{customdata[3]}' + '<br>P-Value: %{customdata[4]}',\n",
    "                                           customdata = [label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"No LD Data\"))\n",
    "\n",
    "            # Add points from other phenos in LD with SV\n",
    "            if toggle_pheno == True:\n",
    "                extra_points_pheno = df_full_join[(df_full_join[\"Chromosome\"] == chromo) & \\\n",
    "                                                  (df_full_join[\"SNP_Position\"] >= min_SNP) & \\\n",
    "                                                  (df_full_join[\"SNP_Position\"] <= max_SNP) & \\\n",
    "                                                  (df_full_join[\"SV_Name\"] == sv)]\n",
    "\n",
    "                extra_points_pheno = extra_points_pheno[~extra_points_pheno[\"SNP_Name_GWAS\"].isin(tab[\"SNP_Name_GWAS\"])]\n",
    "\n",
    "\n",
    "                # Create a list of labels for hovertext\n",
    "                df_lab_2 = extra_points_pheno.to_dict(\"records\")\n",
    "\n",
    "                def label_maker(row):\n",
    "\n",
    "                    label_pheno = row[\"Phenotype\"]\n",
    "                    label_snp_name = row[\"SNP_Name_GWAS\"]\n",
    "                    label_chr = row[\"Chromosome\"]\n",
    "                    label_snp_pos = row[\"SNP_Position\"]\n",
    "                    label_pval = row[\"P-Value\"]\n",
    "                    label_r2 = row[\"R2\"]\n",
    "                    label_d = row[\"D'\"]\n",
    "\n",
    "                    label = [label_pheno, label_snp_name, label_chr, label_snp_pos, label_pval, label_r2, label_d]\n",
    "\n",
    "                    return label\n",
    "\n",
    "\n",
    "                sc_plot.add_trace(go.Scattergl(x = extra_points_pheno[\"SNP_Position\"],\n",
    "                                               y = extra_points_pheno[\"P-Value_log10\"],\n",
    "                                               mode = \"markers\",\n",
    "                                               marker = dict(size = 10,\n",
    "                                                             color = \"rgba(168,168,168,0.5)\",\n",
    "                                                             line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                               hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + '<br>SNP Name: %{customdata[1]}' + '<br>Chromosome: %{customdata[2]}' + '<br>SNP Position: %{customdata[3]}' + '<br>P-Value: %{customdata[4]}' + '<br>r2: %{customdata[5]}' + \"<br>D': %{customdata[6]}\",\n",
    "                                               customdata = [label_maker(row) for row in df_lab_2],\n",
    "                                               showlegend = False,\n",
    "                                               name = \"Other Phenotypes\"))\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -2,\n",
    "                              y1 = -3,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                          width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.5,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-1.5]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + '<br>Chromosome: %{customdata[2]}' + '<br>Start: %{customdata[8]}' + '<br>End: %{customdata[9]}' + '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene[(df_gene[\"chr\"] == chromo) & \\\n",
    "                            (df_gene[\"start\"] >= min_SNP) & \\\n",
    "                            (df_gene[\"end\"] <= max_SNP)]\n",
    "\n",
    "            exons_genes = exons.groupby([\"chr\", \"gene\", \"strand\"])[\"start\"].min()\n",
    "            exons_genes = exons_genes.to_frame().reset_index()\n",
    "\n",
    "            exons_end = exons.groupby([\"chr\", \"gene\", \"strand\"])[\"end\"].max()\n",
    "            exons_end = exons_end.to_frame().reset_index()\n",
    "\n",
    "            exons_genes[\"end\"] = exons_end[\"end\"]\n",
    "\n",
    "            # gene_line_position = -6.25\n",
    "\n",
    "            for i in range(0,len(exons_genes)):\n",
    "                row = exons_genes.loc[exons_genes.index[i]]\n",
    "                gene_line_position = float(df_transcript.loc[df_transcript[\"gene\"] == row[\"gene\"], \"y_coord_pheno\"].values[0])  \n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                x0 = (row[\"start\"]),\n",
    "                                x1 = (row[\"end\"]),\n",
    "                                y0 = gene_line_position,\n",
    "                                y1 = gene_line_position,\n",
    "                                line_width = 3,\n",
    "                                line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    # sc_plot.add_trace(go.Scatter(x = np.array([text_anno_x]),\n",
    "                    #                             y = np.array([gene_line_position - 2.25]),\n",
    "                    #                             mode = \"markers+lines\",\n",
    "                    #                             marker = dict(symbol = \"triangle-right-open\",\n",
    "                    #                                         size = 12,\n",
    "                    #                                         color = \"darkslategrey\"),\n",
    "                    #                             showlegend = False,\n",
    "                    #                             hovertemplate = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                    #                             name = \"gene direction\"))\n",
    "\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = gene_line_position + 1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    # sc_plot.add_trace(go.Scatter(x = np.array([text_anno_x]),\n",
    "                    #                             y = np.array([gene_line_position - 1.25]),\n",
    "                    #                             mode = \"markers+lines\",\n",
    "                    #                             marker = dict(symbol = \"triangle-left-open\",\n",
    "                    #                                         size = 12,\n",
    "                    #                                         color = \"darkslategrey\"),\n",
    "                    #                             showlegend = False,\n",
    "                    #                             hovertemplate = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                    #                             name = \"gene direction\"))\n",
    "\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = gene_line_position + 1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons[exons[\"feature\"] == \"exon\"]\n",
    "\n",
    "            for i in range(0,len(exons_exons)):\n",
    "                row = exons_exons.loc[exons_exons.index[i]]\n",
    "                gene_line_position = float(df_transcript.loc[df_transcript[\"gene\"] == row[\"gene\"], \"y_coord_pheno\"].values[0]) \n",
    "                sc_plot.add_shape(type = \"rect\",\n",
    "                                x0 = (row[\"start\"]),\n",
    "                                x1 = (row[\"end\"]),\n",
    "                                y0 = gene_line_position + 0.75,\n",
    "                                y1 = gene_line_position - 0.75,\n",
    "                                line = dict(color = \"mediumseagreen\",\n",
    "                                            width = 2),\n",
    "                                fillcolor = \"mediumseagreen\")\n",
    "\n",
    "\n",
    "    return sc_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# SCATPLOT = dcc.Graph(id = \"scatter-plot\",\n",
    "#                      figure = make_plot())\n",
    "\n",
    "SCATPLOT = dcc.Loading(id = \"plot-loading\",\n",
    "                       children = [dcc.Graph(id = \"scatter-plot\",\n",
    "                                             figure = make_plot(),\n",
    "                                             config = {\"modeBarButtonsToRemove\": [\"select2d\",\n",
    "                                                                                  \"lasso2d\",\n",
    "                                                                                  \"zoomIn2d\",\n",
    "                                                                                  \"zoomOut2d\"],\n",
    "                                                        \"toImageButtonOptions\": {\"format\": \"svg\",\n",
    "                                                                                 \"filename\": \"gwas_svatalog_plot\",\n",
    "                                                                                 \"height\": 500,\n",
    "                                                                                 \"width\": 700,\n",
    "                                                                                 \"scale\": 1}})],\n",
    "                       type = \"dot\",\n",
    "                       color = \"turquoise\",\n",
    "                       parent_className = \"loading_wrapper\")\n",
    "\n",
    "\n",
    "TOGGLE_SWITCH = html.Div([dbc.Row([dbc.Col([\"D'\"],\n",
    "                                           id = 'toggle-d'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'toggle-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"r\", html.Sup(2,\n",
    "                                                          style = {'font-size' : '15px'})],\n",
    "                                            id = 'toggle-r')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region'),\n",
    "                          dbc.Row([html.P([\"Show Other Phenotypes\",\n",
    "                                           question_icon],\n",
    "                                          id = 'toggle-pheno-header'),\n",
    "                                   dbc.Tooltip(dcc.Markdown(\"Show GWAS-associated SNPs from other phenotypes in respect to selected SV.\",\n",
    "                                                            style = {'white-space':'pre-wrap'}),\n",
    "                                               target = 'toggle-pheno-header',\n",
    "                                               placement = \"auto\",\n",
    "                                               delay = {\"show\" : 500,\n",
    "                                                        \"hide\" : 50},\n",
    "                                               id = 'toggle-pheno-header-tooltip')],\n",
    "                                  id = 'toggle-pheno-header-row'),\n",
    "                          dbc.Row([dbc.Col([\"OFF\"],\n",
    "                                           id = 'toggle-off'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'other-pheno-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"ON\"],\n",
    "                                            id = 'toggle-on')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region-pheno'),\n",
    "                          dbc.Row([dbc.Button(\"Export SNP Data to CSV\",\n",
    "                                               id = 'download-button',\n",
    "                                               color = \"secondary\",\n",
    "                                               outline = True,\n",
    "                                               size = \"sm\",),\n",
    "                                   dcc.Download(id = 'download-csv')],\n",
    "                                   id = 'download-region')])\n",
    "\n",
    "\n",
    "\n",
    "SCATPLOT_DIV = html.Div(id = 'scatter-plot-div',\n",
    "                        children = [SCATPLOT,\n",
    "                                    TOGGLE_SWITCH],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SNP table from GWAS data based on clickData from plot\n",
    "def make_SNP_table(snp = \"None\"):\n",
    "    if snp == \"None\":\n",
    "        df_snp = pd.DataFrame()\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = df_snp.to_dict(\"records\"),\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_snp.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '500px', 'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_snp = df_full_join[(df_full_join[\"SNP_Name_dbSNP\"] == snp) | (df_full_join[\"SNP_Name_GWAS\"] == snp)]\n",
    "\n",
    "        df_snp.loc[:, \"Link\"] = \"[\" + df_snp[\"Pubmed_ID\"] + \"]\" + \"(https://\" + df_snp[\"Link\"] + \")\"\n",
    "\n",
    "        df_snp.loc[:, \"SNP_Name_dbSNP\"] = \"[\" + df_snp[\"SNP_Name_dbSNP\"] + \"]\" + \"(https://www.ncbi.nlm.nih.gov/snp/\" + df_snp[\"SNP_Name_dbSNP\"] + \")\"\n",
    "\n",
    "        df_snp.loc[:, \"SNP_Name_GWAS\"] = \"[\" + df_snp[\"SNP_Name_GWAS\"] + \"]\" + \"(https://www.ebi.ac.uk/gwas/variants/\" + df_snp[\"SNP_Name_GWAS\"] + \")\"\n",
    "\n",
    "        df_snp = df_snp[[\"Chromosome\",\n",
    "                         \"SNP_Position\",\n",
    "                         \"SNP_Name_dbSNP\",\n",
    "                         \"SNP_Name_GWAS\",\n",
    "                         \"Reference_Allele\",\n",
    "                         \"Alternate_Allele\",\n",
    "                         \"Risk_Allele\",\n",
    "                         \"Risk_Allele_Frequency\",\n",
    "                         \"Sample_AF\",\n",
    "                         \"gnomAD_nfe_AF\",\n",
    "                         \"Phenotype\",\n",
    "                         \"P-Value\",\n",
    "                         \"Study\",\n",
    "                         \"Link\"]]\n",
    "        df_snp = df_snp.rename(columns = {\"Link\" : \"Pubmed Link\",\n",
    "                                          \"Chromosome\" : \"Chrom\",\n",
    "                                          \"SNP_Position\" : \"SNP Position\",\n",
    "                                          \"SNP_Name_dbSNP\" : \"SNP Name: dbSNP\",\n",
    "                                          \"SNP_Name_GWAS\" : \"SNP Name: GWAS\",\n",
    "                                          \"Risk_Allele\" : \"Risk Allele\",\n",
    "                                          \"Risk_Allele_Frequency\" : \"Risk AF\",\n",
    "                                          \"Reference_Allele\" : \"Reference Allele\",\n",
    "                                          \"Alternate_Allele\" : \"Alternate Allele\",\n",
    "                                          \"Sample_AF\" : \"Sample AF\",\n",
    "                                          \"gnomAD_nfe_AF\" : \"gnomAD NFE AF\"})\n",
    "\n",
    "        df_snp = df_snp.drop_duplicates()\n",
    "\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = df_snp.to_dict(\"records\"),\n",
    "                                    columns = [{'id': c, 'name': c, 'presentation': 'markdown'} if (c == \"Pubmed Link\" or c == \"SNP Name: dbSNP\" or c == \"SNP Name: GWAS\") else ({'id': c, 'name': c, 'type':'numeric', 'format': {'specifier': '.3f'}} if (c == \"Risk AF\" or c == \"Sample AF\" or c == \"gnomAD NFE AF\") else {'id': c, 'name': c}) for c in df_snp.columns],\n",
    "                                    markdown_options = {\"html\" : True,\n",
    "                                                        \"link_target\": \"_blank\"},\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    # fixed_columns = {'headers' : True,\n",
    "                                    #                  'data' : 4},\n",
    "                                    style_table = {'height': '500px',\n",
    "                                                   'minWidth': '100%',\n",
    "                                                   'width' : '100%',\n",
    "                                                   'maxWidth' : '100%',\n",
    "                                                   'overflowX' : 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': [\"Chrom\",\n",
    "                                                                                    \"SNP Position\",\n",
    "                                                                                    \"SNP Name: dbSNP\",\n",
    "                                                                                    \"SNP Name: GWAS\",\n",
    "                                                                                    \"Reference Allele\",\n",
    "                                                                                    \"Alternate Allele\",\n",
    "                                                                                    \"Risk Allele\",\n",
    "                                                                                    \"Risk AF\",\n",
    "                                                                                    \"Sample AF\",\n",
    "                                                                                    \"gnomAD NFE AF\",\n",
    "                                                                                    \"P-Value\",\n",
    "                                                                                    \"Pubmed Link\"]},\n",
    "                                                               'minWidth' : '100px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke'},\n",
    "                                                               {'if': {'column_id': [\"Study\",\n",
    "                                                                                     \"Phenotype\"]},\n",
    "                                                               'maxWidth' : '200px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke',\n",
    "                                                               'whiteSpace' : 'normal',\n",
    "                                                               'height' : 'auto'}],\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold',\n",
    "                                                    'textAlign' : 'left',\n",
    "                                                    'whiteSpace' : 'normal',\n",
    "                                                    'height' : 'auto'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SNP_TABLE = html.Div([html.Div(id = 'snp-table-div',\n",
    "                               children = [make_SNP_table()],\n",
    "                               className = 'sections'),\n",
    "                      dbc.Modal([dbc.ModalHeader(),\n",
    "                                 dbc.ModalBody(\"SNP table populated below.\")],\n",
    "                                 id = 'modal-snp-table',\n",
    "                                 is_open = False)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### HEADERS ####\n",
    "\n",
    "## Navigation Bar to match LocusFocus\n",
    "NAVBAR = dbc.NavbarSimple(id = 'navbar-lf',\n",
    "                          brand = \"LocusFocus\",\n",
    "                          brand_href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                          brand_external_link = \"_blank\",\n",
    "                          color = \"#40e0d0\",\n",
    "                          links_left = True,\n",
    "                          expand = \"lg\",\n",
    "                          children = [dbc.NavItem(dbc.NavLink(\"Colocalization\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Set-Based Test\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/setbasedtest\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"GWAS SVatalog\",\n",
    "                                                              href = \"https://svatalog.research.sickkids.ca/\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Documentation\",\n",
    "                                                              href = \"https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Contact Us\",\n",
    "                                                              id = 'button-modal-contact-us',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Contact Us\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         <h5 children=\"GWAS SVatalog\"  />\n",
    "                                                                                         Shalvi Chirmade <span style=\"color: #2ba089\" children=\"shalvi.chirmade@sickkids.ca\" />\n",
    "\n",
    "                                                                                         &nbsp;\n",
    "\n",
    "                                                                                         <h5 children=\"LocusFocus\" />\n",
    "                                                                                         Mackenzie Frew <span style=\"color: #2ba089\" children=\"mackenzie.frew@sickkids.ca\" />\n",
    "                                                                                        ''',\n",
    "                                                                                        dangerously_allow_html = True,\n",
    "                                                                                        id = 'contact-us-mrkdwn'))],\n",
    "                                                              id = 'modal-contact-us',\n",
    "                                                              is_open = False)]),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Subscribe\",\n",
    "                                                              href = \"https://mailchi.mp/752ab1c4d516/locusfocus\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Citation\",\n",
    "                                                              id = 'button-modal-citation',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Citation\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         *to be published*\n",
    "                                                                                        '''))],\n",
    "                                                              id = 'modal-citation',\n",
    "                                                              is_open = False)])])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Logo for the top of the webpage, disclaimer and info\n",
    "GSV_LOGO = html.Img(src = \"assets/gwas-svatalog-name.png\",\n",
    "                    id = 'gsv-logo-image')\n",
    "\n",
    "TCAG_LOGO = html.Img(src = \"assets/tcaglogo.png\",\n",
    "                     id = 'tcag-logo-image')\n",
    "\n",
    "SK_LOGO = html.Img(src = \"assets/SKlogo.svg\",\n",
    "                   id = 'sk-logo-image')\n",
    "\n",
    "DISCLAIMER_BUTTON = dbc.Button(\"Disclaimer\",\n",
    "                               color = \"secondary\",\n",
    "                               outline = True,\n",
    "                               size = \"sm\",\n",
    "                               id = 'disclaimer-button')\n",
    "\n",
    "DISCLAIMER_COLLAPSE = dbc.Collapse(dbc.Card(dbc.CardBody(\"Database constructed from predominantly European population of 101 individuals with Cystic Fibrosis (CF). The alleles affected by CF aside, the remainder of the genome is comparable to a healthy population of European descent (citation TBD). Genomic location is referenced against GRCh38.\")),\n",
    "                                   id = 'disclaimer-collapse',\n",
    "                                   is_open = False)\n",
    "\n",
    "DISCLAIMER_DIV = html.Div(id = 'disclaimer-div',\n",
    "                          children = [DISCLAIMER_BUTTON,\n",
    "                                      DISCLAIMER_COLLAPSE])\n",
    "\n",
    "# DOCUMENTATION = dcc.Markdown('''\n",
    "# Documentation for GWAS SVatalog can be found [here](https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html).\n",
    "# Colocalization testing of GWAS loci across various datasets can be conducted at [LocusFocus] (https://locusfocus.research.sickkids.ca/).\n",
    "# ''',\n",
    "#                              id = 'doc-markdown')\n",
    "\n",
    "# DISC_DOC_DIV = html.Div(id = 'disc-doc-div',\n",
    "#                         children = [DOCUMENTATION,\n",
    "#                                     DISCLAIMER_DIV])\n",
    "\n",
    "GSV_LOGO_DIV = html.Div(id = 'gsv-logo-image-div',\n",
    "                        children = [GSV_LOGO])\n",
    "\n",
    "COMP_LOGO_DIV = html.Div(id = 'comp-logo-image-div',\n",
    "                         children = [TCAG_LOGO,\n",
    "                                     SK_LOGO,\n",
    "                                     DISCLAIMER_DIV])\n",
    "\n",
    "LOGO_DIV = html.Div(id = 'logo-image-div',\n",
    "                    children = [GSV_LOGO_DIV,\n",
    "                                COMP_LOGO_DIV])\n",
    "\n",
    "\n",
    "HEADER_DIV = html.Div(id = 'header-line-div')\n",
    "\n",
    "\n",
    "# Header for filters\n",
    "FILTER_HEADER = html.H3(\"Search for Structural Variants\",\n",
    "                        id = 'filter-header',\n",
    "                        style = {'textAlign' : 'center',\n",
    "                                 'color' : '#2F2E2D',\n",
    "                                 'height' : '70 px'})\n",
    "\n",
    "FILTER_HEADER_DIV = html.Div(id = 'filter-header-div',\n",
    "                             children = [FILTER_HEADER],\n",
    "                             className = 'section-headers')\n",
    "\n",
    "\n",
    "\n",
    "# Header for plot\n",
    "PLOT_HEADER = html.Div([html.H3([\"SVs and GWAS Hits Linkage Disequilibirum Plot\",\n",
    "                                 question_icon],\n",
    "                                 id = 'plot-header',\n",
    "                                 style = {'textAlign' : 'center',\n",
    "                                          'color' : '#2F2E2D',\n",
    "                                          'height' : '70 px'}),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Interactive plot visualizing linkage disequilibrium (LD) between the selected SV and GWAS-associated SNPs from GWAS Catalog.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'plot-header',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'plot-header-tooltip')])\n",
    "\n",
    "PLOT_HEADER_DIV = html.Div(id = 'plot-header-div',\n",
    "                           children = [PLOT_HEADER],\n",
    "                           className = 'section-headers')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#### CALLBACKS ####\n",
    "\n",
    "# Incorporate together to create interactive webpage\n",
    "app.layout = html.Div([NAVBAR,\n",
    "                       LOGO_DIV,\n",
    "                       HEADER_DIV,\n",
    "                       FILTER_HEADER_DIV,\n",
    "                       ALL_FILTER_DIV,\n",
    "                       SV_TABLE_DIV,\n",
    "                       PLOT_HEADER_DIV,\n",
    "                       SCATPLOT_DIV,\n",
    "                       SNP_TABLE])\n",
    "\n",
    "\n",
    "# Contact Us modal\n",
    "@app.callback(\n",
    "    Output('modal-contact-us', 'is_open'),\n",
    "    [Input('button-modal-contact-us', 'n_clicks')],\n",
    "    [State('modal-contact-us', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_contact, is_open):\n",
    "    if modal_contact:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Citation modal\n",
    "@app.callback(\n",
    "    Output('modal-citation', 'is_open'),\n",
    "    [Input('button-modal-citation', 'n_clicks')],\n",
    "    [State('modal-citation', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_citation, is_open):\n",
    "    if modal_citation:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Disclaimer cutton opens collapsed text\n",
    "@app.callback(\n",
    "    Output('disclaimer-collapse', 'is_open'),\n",
    "    [Input('disclaimer-button', 'n_clicks')],\n",
    "    [State('disclaimer-collapse', 'is_open')]\n",
    ")\n",
    "def dislcaimer_collapse(clicks, is_open):\n",
    "    if clicks:\n",
    "        return not is_open\n",
    "    return is_open\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Chromosome dropdown changes the genes and phenotypes displayed\n",
    "@app.callback(\n",
    "    Output('gene-dropdown', 'options'),\n",
    "    Output('phenotype-dropdown', 'options'),\n",
    "    Input('chromosome-dropdown', 'value')\n",
    ")\n",
    "def update_gene_dropdown(chrom):\n",
    "\n",
    "    if chrom is None:\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    elif chrom == \"Any\":\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    else:\n",
    "        chromosome = str(\"chr\" + str(chrom))\n",
    "        new_genes = df_gene[df_gene[\"chr\"] == chromosome]\n",
    "        new_genes = sorted(list(new_genes[\"gene\"]))\n",
    "        new_pheno = df_full_join[df_full_join[\"Chromosome\"] == chromosome]\n",
    "        new_pheno = sorted(list(new_pheno[\"Phenotype\"].unique().astype(str)))\n",
    "\n",
    "    return [{'label': str(i), 'value': str(i)} for i in new_genes], [{'label': str(i), 'value': str(i)} for i in new_pheno]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Filters alter the table displayed\n",
    "@app.callback(\n",
    "    Output('strvar-table-div', 'children'),\n",
    "    Input('chromosome-dropdown', 'value'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('gene-dropdown', 'value'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_table(chrom, pheno, gene, range_start, range_end):\n",
    "\n",
    "    chrom = chromosomes if chrom is None else chrom\n",
    "    pheno = phenotypes if pheno is None else pheno\n",
    "    gene = genes if gene is None else gene\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if chrom == \"Any\":\n",
    "        chrom = chromosomes\n",
    "    else:\n",
    "        chrom = chrom\n",
    "    \n",
    "    if pheno == \"Any\":\n",
    "        pheno = phenotypes\n",
    "    else:\n",
    "        pheno = pheno\n",
    "\n",
    "    new_df = make_table(chrom = chrom,\n",
    "                        range_start = range_start,\n",
    "                        range_end = range_end,\n",
    "                        pheno = pheno,\n",
    "                        gene = gene)\n",
    "\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('dtable-textbox', 'children'),\n",
    "    Input('strvar-table', 'derived_virtual_data'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_sv_textbox(data, selected_rows, range_start, range_end):\n",
    "\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if range_end <= range_start:\n",
    "        return html.P(\"Fix range values.\")\n",
    "\n",
    "    if data == []:\n",
    "        return html.P(\"No SVs in selected filters.\")\n",
    "    \n",
    "    elif selected_rows:\n",
    "        return html.P(\"Plot generated below.\")\n",
    "    \n",
    "    else:\n",
    "        return html.P(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('anno-table-div', 'children'),\n",
    "    Input('strvar-table', 'selected_rows')\n",
    ")\n",
    "def update_anno_table(selected_rows):\n",
    "\n",
    "    if selected_rows:\n",
    "        # sv_selected = tab_show.iloc[selected_rows[0],[0]][0]\n",
    "        row_label = tab_show.index[selected_rows[0]]\n",
    "        col_label = tab_show.columns[0]\n",
    "        sv_selected = tab_show.loc[row_label, col_label]\n",
    "        new_anno_table = make_annotation_table(sv = sv_selected)\n",
    "\n",
    "    else:\n",
    "        new_anno_table = make_annotation_table(sv = \"None\")\n",
    "\n",
    "    return new_anno_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Collapse SV annotation table\n",
    "@app.callback(\n",
    "    Output(\"sv-table-collapse\", \"is_open\"),\n",
    "    [Input(\"sv-button-collapse\", \"n_clicks\")],\n",
    "    [State(\"sv-table-collapse\", \"is_open\")],\n",
    ")\n",
    "def toggle_collapse(n, is_open):\n",
    "    if n:\n",
    "        return not is_open\n",
    "    return is_open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Only row selection on table changes plot, SV div textbox and button enabling\n",
    "@app.callback(\n",
    "    Output('scatter-plot', 'figure'),\n",
    "    Output('scatter-plot', 'clickData'),\n",
    "    Output('download-button', 'disabled'),\n",
    "    Output('other-pheno-switch', 'disabled'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value')\n",
    ")\n",
    "def update_plot(selected_rows, pheno, toggle, toggle_pheno):\n",
    "\n",
    "    pheno = \"None\" if pheno is None else pheno\n",
    "    toggle = False if toggle is None else toggle\n",
    "    toggle_pheno = False if toggle_pheno is None else toggle_pheno\n",
    "\n",
    "    if selected_rows:\n",
    "        # sv_name = tab_show.iloc[selected_rows[0],[0]][0]\n",
    "        row_label = tab_show.index[selected_rows[0]]\n",
    "        col_label = tab_show.columns[0]\n",
    "        sv_name = tab_show.loc[row_label, col_label]\n",
    "        button = False\n",
    "\n",
    "    else:\n",
    "        sv_name = \"None\"\n",
    "        button = True\n",
    "\n",
    "    if pheno == \"None\":\n",
    "        pheno_toggle_disable = True\n",
    "\n",
    "    else:\n",
    "        pheno_toggle_disable = False\n",
    "\n",
    "    new_plot = make_plot(sv = sv_name,\n",
    "                         pheno = pheno,\n",
    "                         toggle = toggle,\n",
    "                         toggle_pheno = toggle_pheno)\n",
    "\n",
    "    return new_plot, {}, button, pheno_toggle_disable\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Update SNP table based on data point clicked on plot\n",
    "@app.callback(\n",
    "        Output('snp-table-div', 'children'),\n",
    "        Output('modal-snp-table', 'is_open'),\n",
    "        Input('scatter-plot', 'clickData'),\n",
    "        Input('strvar-table', 'selected_rows'),\n",
    "        State('modal-snp-table', 'is_open')\n",
    ")\n",
    "def update_SNP_table(clickData, selected_rows, is_open):\n",
    "\n",
    "    if clickData:\n",
    "        if clickData == {}:\n",
    "            rsID = \"None\"\n",
    "            is_open = False\n",
    "        else:\n",
    "            rsID = clickData[\"points\"][0][\"customdata\"][1]\n",
    "            is_open = True\n",
    "\n",
    "    else:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    if not selected_rows:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    new_snp_table = make_SNP_table(snp = rsID)\n",
    "\n",
    "    return new_snp_table, is_open\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash is running on http://127.0.0.1:4321/\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:dash.dash:Dash is running on http://127.0.0.1:4321/\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dash app running on http://127.0.0.1:4321/\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Download SNP data seen in plot\n",
    "@app.callback(\n",
    "    Output('download-csv', 'data'),\n",
    "    Input('download-button', 'n_clicks'),\n",
    "    Input('scatter-plot', 'relayoutData'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value'),\n",
    "    prevent_initial_call = True\n",
    ")\n",
    "def download_plot_data(n_clicks, relayoutData, toggle, toggle_pheno):\n",
    "    # Example from https://towardsdatascience.com/building-a-dashboard-in-plotly-dash-c748588e2920\n",
    "    global df_download\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage = \"D'\"\n",
    "    elif toggle == True:\n",
    "        linkage = \"R2\"\n",
    "    else:\n",
    "        linkage = \"D'\"\n",
    "\n",
    "\n",
    "    # If button was triggered\n",
    "    if ctx.triggered[0]['prop_id'] == 'download-button.n_clicks':\n",
    "\n",
    "        if (relayoutData == None or \"xaxis.autorange\" in relayoutData):\n",
    "            # This means the plot is in the default starting position\n",
    "            df_download = tab\n",
    "            df_extra = extra_points\n",
    "\n",
    "            if toggle_pheno == True:\n",
    "                df_extra_pheno = extra_points_pheno\n",
    "                df_download = pd.concat([df_download, df_extra, df_extra_pheno])\n",
    "            elif toggle_pheno == False:\n",
    "                df_download = pd.concat([df_download, df_extra])\n",
    "            else:\n",
    "                df_download = pd.concat([df_download, df_extra])\n",
    "\n",
    "\n",
    "        elif \"xaxis.range[0]\" in relayoutData:\n",
    "\n",
    "            if \"yaxis.range[0]\" in relayoutData:\n",
    "\n",
    "                if \"P-Value\" in extra_points.columns:\n",
    "                    df_download = tab[(tab[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"]) & \\\n",
    "                                      (tab[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"]) &\n",
    "                                            (extra_points[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra_pheno = extra_points_pheno[(extra_points_pheno[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                                        (extra_points_pheno[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"]) &\n",
    "                                                        (extra_points_pheno[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                                        (extra_points_pheno[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                else:\n",
    "                    df_download = tab[(tab[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"]) & \\\n",
    "                                      (tab[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                      (tab[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"]) &\n",
    "                                            (extra_points[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "            else:\n",
    "\n",
    "                if toggle_pheno == True:\n",
    "\n",
    "                    df_download = tab[(tab[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra_pheno = extra_points_pheno[(extra_points_pheno[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                                        (extra_points_pheno[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra, df_extra_pheno])\n",
    "\n",
    "                elif toggle_pheno == False:\n",
    "\n",
    "                    df_download = tab[(tab[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra])\n",
    "\n",
    "                else:\n",
    "\n",
    "                    df_download = tab[(tab[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"SNP_Position\"] >= relayoutData[\"xaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"SNP_Position\"] <= relayoutData[\"xaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra])\n",
    "\n",
    "\n",
    "\n",
    "        else: # yaxis.range[0] only\n",
    "\n",
    "            if toggle_pheno == True:\n",
    "\n",
    "                    df_download = tab[(tab[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                      (tab[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra_pheno = extra_points_pheno[(extra_points_pheno[\"P-Value_log10\"] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                                        (extra_points_pheno[\"P-Value_log10\"] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra, df_extra_pheno])\n",
    "\n",
    "            elif toggle_pheno == False:\n",
    "\n",
    "                    df_download = tab[(tab[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                      (tab[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra])\n",
    "\n",
    "            else:\n",
    "\n",
    "                    df_download = tab[(tab[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                      (tab[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_extra = extra_points[(extra_points[linkage] >= relayoutData[\"yaxis.range[0]\"]) & \\\n",
    "                                            (extra_points[linkage] <= relayoutData[\"yaxis.range[1]\"])]\n",
    "\n",
    "                    df_download = pd.concat([df_download, df_extra])\n",
    "            #raise PreventUpdate\n",
    "\n",
    "        return dcc.send_data_frame(df_download.to_csv, \"gwas_svatalog_snp_data.csv\")\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug = True,\n",
    "                   host = \"127.0.0.1\",\n",
    "                   port = 4321)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
