{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 499,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dash_bootstrap_components as dbc\n",
    "import dash_daq as daq\n",
    "import heapq\n",
    "import numpy as np\n",
    "import glob\n",
    "import os\n",
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import polars as pl\n",
    "import pyarrow\n",
    "\n",
    "from dash import Dash, dcc, html, dash_table, Input, Output, State, ctx\n",
    "from jupyter_dash import JupyterDash\n",
    "from gtfparse import read_gtf # gtfparse==1.3.0\n",
    "\n",
    "\n",
    "app = JupyterDash(__name__, \n",
    "           external_stylesheets = [dbc.themes.LUX, dbc.icons.BOOTSTRAP],\n",
    "           title = \"GWAS SVatalog\")\n",
    "\n",
    "app._favicon = \"assets/favicon.ico\"\n",
    "\n",
    "server = app.server\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "DF_GWAS_FULL = pl.read_csv(\"data/gwas_catalog_v1.0-associations_e108.tsv\",\n",
    "                           separator = \"\\t\",\n",
    "                           quote_char = '',\n",
    "                           dtypes = {\"PUBMEDID\": pl.Utf8, \n",
    "                                     \"LINK\": pl.Utf8,\n",
    "                                     \"P-VALUE\": pl.Float64},\n",
    "                           infer_schema_length = 1_000_000)\n",
    "\n",
    "# Remove quotes from all string columns\n",
    "# DF_GWAS_FULL = DF_GWAS_FULL.with_columns([pl.col(column).str.replace(r'^\"|\"$', \"\")\n",
    "#                                           for column in DF_GWAS_FULL.columns])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 501,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gwas = DF_GWAS_FULL.select([\n",
    "    pl.col(\"CHR_ID\"),\n",
    "    pl.col(\"CHR_POS\").cast(pl.Int32),\n",
    "    pl.col(\"SNPS\"),\n",
    "    pl.col(\"STRONGEST SNP-RISK ALLELE\"),\n",
    "    pl.col(\"RISK ALLELE FREQUENCY\"),\n",
    "    pl.col(\"DISEASE/TRAIT\"),\n",
    "    pl.col(\"STUDY\"),\n",
    "    pl.col(\"PUBMEDID\"),\n",
    "    pl.col(\"LINK\"),\n",
    "    pl.col(\"P-VALUE\")\n",
    "]).drop_nulls([\"CHR_ID\"]).with_columns([\n",
    "    (pl.lit(\"chr\") + pl.col(\"CHR_ID\")).alias(\"CHR_ID\"),\n",
    "    pl.col(\"SNPS\").str.split(\";\").arr.first(),\n",
    "    pl.col(\"STRONGEST SNP-RISK ALLELE\").str.split(\"-\").arr.last()\n",
    "]).filter(pl.col(\"P-VALUE\") <= 1)\n",
    "\n",
    "# Rename columns in the output\n",
    "df_gwas = df_gwas.rename({\n",
    "    \"CHR_ID\": \"Chromosome\",\n",
    "    \"CHR_POS\" : \"SNP_Position\",\n",
    "    \"SNPS\": \"SNP_Name_GWAS\",\n",
    "    \"STRONGEST SNP-RISK ALLELE\": \"Risk_Allele\",\n",
    "    \"RISK ALLELE FREQUENCY\": \"Risk_Allele_Frequency\",\n",
    "    \"DISEASE/TRAIT\": \"Phenotype\",\n",
    "    \"STUDY\": \"Study\",\n",
    "    \"PUBMEDID\": \"Pubmed_ID\",\n",
    "    \"LINK\": \"Link\",\n",
    "    \"P-VALUE\": \"P-Value\"\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 502,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SV Annotation data - created by Dr. Zhuozhi Wang on March 13, 2023\n",
    "DF_ANNO_FULL = pl.read_csv(\"data/sv_annotations.tsv\",\n",
    "                           separator = \"\\t\",\n",
    "                           quote_char = '',\n",
    "                           infer_schema_length = 1_000_000)\n",
    "\n",
    "# Select and rename specific columns\n",
    "df_anno = DF_ANNO_FULL.select([\n",
    "    pl.col(\"Chromosome\"),\n",
    "    pl.col(\"Start\").alias(\"SV_Start\"),\n",
    "    pl.col(\"End\").cast(pl.Int64).alias(\"SV_End\"),\n",
    "    pl.col(\"ID\").alias(\"SV_Name\"),\n",
    "    pl.col(\"Type\").alias(\"SV_Type\"),\n",
    "    pl.col(\"Length\").alias(\"SV_Length\")\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 503,
   "metadata": {},
   "outputs": [],
   "source": [
    "LD_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                       \"*_ld_stats.txt\"))\n",
    "LD_tables = [pl.read_csv(file_path,\n",
    "                         separator = \"\\t\",\n",
    "                         has_header = True,\n",
    "                         new_columns = [\"SNP_Name\", \n",
    "                                        \"SNP_Position\",\n",
    "                                        \"SV_Name\",\n",
    "                                        \"SV_Position\",\n",
    "                                        \"R2\",\n",
    "                                        \"D'\"])\n",
    "             for file_path in LD_file_paths]\n",
    "\n",
    "DF_LD = pl.concat(LD_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SV/SNP allele annotations done by Thomas Nalpathamkalam on April 10, 2023\n",
    "allele_file_paths = glob.glob(os.path.join(\"data/LD\",\n",
    "                                            \"*_allele_freq.txt\"))\n",
    "allele_tables = [pl.read_csv(file_path,\n",
    "                             separator = \"\\t\",\n",
    "                             has_header = True,\n",
    "                             new_columns = [\"Chromosome\",\n",
    "                                            \"SNP_Position\",\n",
    "                                            \"SV_SNP_Name\",\n",
    "                                            \"Reference_Allele\",\n",
    "                                            \"Alternate_Allele\",\n",
    "                                            \"Sample_AF\",\n",
    "                                            \"gnomAD_nfe_AF\",\n",
    "                                            \"dbSNP\"])\n",
    "                 for file_path in allele_file_paths]\n",
    "\n",
    "DF_ALLELE = pl.concat(allele_tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Gene/exon data from MANE.GRCh38.v1.0.ensembl_genomic.gtf extracted January 5, 2023\n",
    "\n",
    "DF_GTF = read_gtf(\"data/MANE.GRCh38.v1.0.ensembl_genomic.gtf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 506,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = pl.DataFrame(DF_GTF).filter(\n",
    "    (pl.col(\"feature\") == \"transcript\") | (pl.col(\"feature\") == \"exon\")\n",
    ").filter(pl.col(\"tag\").str.contains(\"MANE_Select\")).select([\n",
    "    pl.col(\"seqname\").alias(\"chr\"),\n",
    "    pl.col(\"start\").cast(pl.Int64),\n",
    "    pl.col(\"end\").cast(pl.Int64),\n",
    "    pl.col(\"strand\"),\n",
    "    pl.col(\"feature\"),\n",
    "    pl.col(\"gene_name\").alias(\"gene\")\n",
    "])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 507,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### DATA WRANGLING ####\n",
    "## Subset allele data for only SV information\n",
    "df_sv_allele = (DF_ALLELE\n",
    "                .filter(pl.col(\"SV_SNP_Name\").str.starts_with(\"P\"))\n",
    "                .rename({\"SV_SNP_Name\": \"SV_Name\"})\n",
    "                .drop([\"Reference_Allele\", \"SNP_Position\", \"gnomAD_nfe_AF\", \"dbSNP\"])\n",
    "                .unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 508,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset allele data for only SNP information\n",
    "df_snp_allele = (DF_ALLELE.filter(~pl.col(\"SV_SNP_Name\").str.starts_with(\"P\"))\n",
    "                 .rename({\"SV_SNP_Name\": \"SNP_Name\"})\n",
    "                 .unique())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset DF_ANNO_FULL with columns that can be displayed to the public. Add SV AF to this table as well.\n",
    "df_sv_anno = (DF_ANNO_FULL\n",
    "              .join(\n",
    "                  df_sv_allele.select([\"SV_Name\",\n",
    "                                       \"Alternate_Allele\", \n",
    "                                       \"Sample_AF\"]),\n",
    "                                       left_on = \"ID\",\n",
    "                                       right_on = \"SV_Name\",\n",
    "                                       how = \"inner\")\n",
    "              .drop(\"Alternate_Allele\")\n",
    "              .select([\"ID\",\n",
    "                       *DF_ANNO_FULL.columns[1:6],\n",
    "                       \"Sample_AF\",\n",
    "                       *DF_ANNO_FULL.columns[6:]])\n",
    "              .rename({\"ID\": \"SV Name\",\n",
    "                       \"Sample_AF\": \"SV Sample AF\"}))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 510,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of SV table for easier access of data required - join the dataframes based on SV names\n",
    "df_sv_join = (DF_LD\n",
    "              .join(df_anno, \n",
    "                    on = \"SV_Name\", \n",
    "                    how = \"left\")\n",
    "              .select([\"Chromosome\", \n",
    "                       \"SV_Name\", \n",
    "                       \"SV_Start\", \n",
    "                       \"SV_End\",\n",
    "                       \"SNP_Name\", \n",
    "                       \"SNP_Position\", \n",
    "                       \"R2\", \n",
    "                       \"D'\"])\n",
    "              .join(df_sv_allele, \n",
    "                    on = [\"Chromosome\", \n",
    "                          \"SV_Name\"], \n",
    "                    how=\"left\")\n",
    "              .rename({\"Alternate_Allele\": \"SV_Type\",\n",
    "                       \"Sample_AF\": \"SV_AF\"})\n",
    "              .with_columns([pl.when(pl.col(\"SV_Type\") == \"<DEL>\").then(\"Deletion\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<INS>\").then(\"Insertion\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<DUP>\").then(\"Duplication\")\n",
    "                               .when(pl.col(\"SV_Type\") == \"<INV>\").then(\"Inversion\")\n",
    "                               .otherwise(pl.col(\"SV_Type\"))\n",
    "                               .alias(\"SV_Type\")])\n",
    "              .select([\"Chromosome\", \n",
    "                       \"SV_Name\", \n",
    "                       \"SV_Start\", \n",
    "                       \"SV_End\",\n",
    "                       \"SV_Type\", \n",
    "                       \"SV_AF\", \n",
    "                       \"SNP_Name\", \n",
    "                       \"SNP_Position\", \n",
    "                       \"R2\", \n",
    "                       \"D'\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 511,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Merging SNP allele information with the LD dataset\n",
    "df_sv_snp_join = (\n",
    "    df_sv_join\n",
    "    .join(\n",
    "        df_snp_allele, \n",
    "        on=[\"Chromosome\", \"SNP_Name\", \"SNP_Position\"], \n",
    "        how=\"left\"\n",
    "    )\n",
    "    .drop(\"SNP_Name\")\n",
    "    .with_columns(\n",
    "        pl.col(\"SNP_Position\").cast(pl.Int32))\n",
    "    .select([\n",
    "        \"Chromosome\", \"SV_Name\", \"SV_Start\", \"SV_End\", \"SV_Type\", \"SV_AF\",\n",
    "        \"dbSNP\", \"SNP_Position\", \n",
    "        \"Reference_Allele\", \"Alternate_Allele\", \"Sample_AF\", \"gnomAD_nfe_AF\", \"R2\", \"D'\"\n",
    "    ])  # Reorder columns\n",
    "    .rename({\"dbSNP\": \"SNP_Name_dbSNP\"})  # Rename column\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 512,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of a full dataframe to make data extraction much easier\n",
    "df_full_join = (df_sv_snp_join\n",
    "                .join(df_gwas, \n",
    "                      on = [\"Chromosome\", \n",
    "                            \"SNP_Position\"], \n",
    "                      how = \"left\")\n",
    "                .filter(pl.col(\"SV_AF\").is_not_null())\n",
    "                .with_columns(pl.when(pl.col(\"P-Value\") < 1e-50)\n",
    "                                .then(1e-50)\n",
    "                                .otherwise(pl.col(\"P-Value\"))\n",
    "                                .alias(\"P-Value\"))\n",
    "                .with_columns((-pl.col(\"P-Value\")\n",
    "                                  .log10())\n",
    "                                  .alias(\"P-Value_log10\")))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset df_anno for the SVs in df_full_foin\n",
    "sv_names = df_full_join.select(\"SV_Name\").unique()\n",
    "df_anno_subset = df_anno.filter(pl.col(\"SV_Name\")\n",
    "                                  .is_in(sv_names[\"SV_Name\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 514,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of phenotype dictionary to encompass SV for each gene - computationally faster when creating SV table\n",
    "df_phenotype = (\n",
    "    df_full_join\n",
    "    .groupby(\"Phenotype\")\n",
    "    .agg(pl.col(\"SV_Name\").unique())  # Ensure SV_Name is unique within each group\n",
    ")\n",
    "\n",
    "dict_pheno = {\n",
    "    row[0]: set(row[1]) for row in df_phenotype.iter_rows()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 515,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Creation of y-coordinates to prevent overlap in genes when plotted\n",
    "def assign_y_coordinates(df: pl.DataFrame,\n",
    "                         y_start: float = -0.3,\n",
    "                         y_increment: float = 0.5,\n",
    "                         col_name: str = \"y_coord\") -> pl.DataFrame:\n",
    "    \n",
    "    # Priority queue to keep track of the end positions of genes for each y-coordinate\n",
    "    df_filtered = (df.filter(pl.col(\"feature\") == \"transcript\")\n",
    "                     .sort([\"chr\", \"start\"]))\n",
    "    \n",
    "    rows = df_filtered.iter_rows()\n",
    "    \n",
    "     # Dictionary to store priority queues for each chromosome\n",
    "    chromosome_heaps = {}\n",
    "\n",
    "    # List to store the y-coordinate for each gene annotation\n",
    "    y_coordinates = []\n",
    "\n",
    "    for row in rows:\n",
    "        chrom, start, end = row[0], row[1], row[2]\n",
    "\n",
    "        # Initialize heap for new chromosomes\n",
    "        if chrom not in chromosome_heaps:\n",
    "            chromosome_heaps[chrom] = []\n",
    "\n",
    "        heap = chromosome_heaps[chrom]\n",
    "\n",
    "        # Check for available y-coordinates (no overlap)\n",
    "        if heap and heap[0][0] <= start:\n",
    "            _, y_coord = heapq.heappop(heap)\n",
    "            heapq.heappush(heap, (end, y_coord))\n",
    "        else:\n",
    "            # Create a new y-coordinate if none available\n",
    "            y_coord = heap[0][1] - y_increment if heap else y_start\n",
    "            heapq.heappush(heap, (end, y_coord))\n",
    "\n",
    "        y_coordinates.append(y_coord)\n",
    "\n",
    "    # Add the y-coordinates as a new column\n",
    "    return df_filtered.with_columns(pl.Series(col_name, y_coordinates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 516,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_transcript = assign_y_coordinates(df = df_gene,\n",
    "                                     y_start = -0.3,\n",
    "                                     y_increment = 0.15,\n",
    "                                     col_name = \"y_coord_nopheno\")\n",
    "\n",
    "df_transcript = assign_y_coordinates(df = df_transcript,\n",
    "                                     y_start = -6.25,\n",
    "                                     y_increment = 4,\n",
    "                                     col_name = \"y_coord_pheno\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 517,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add icon for help popup\n",
    "question_icon = html.I(className = 'bi bi-question-diamond',\n",
    "                       style = {\"display\" : \"inline-block\",\n",
    "                                \"color\" : \"lightsalmon\",\n",
    "                                \"font-size\" : \"70%\"})\n",
    "                    #    id = 'question-icon')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 518,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### FILTERS ####\n",
    "\n",
    "# Make dropdown box for chromosomes and textbox for genomic range\n",
    "REGION_LABEL = html.P(\"Genomic Region\")\n",
    "\n",
    "chromosomes = [\"Any\", *range(1, 23), \"X\", \"Y\"]\n",
    "\n",
    "CHR_DD = dcc.Dropdown(id = 'chromosome-dropdown',\n",
    "                      placeholder = \"Select a chromosome\",\n",
    "                      options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in chromosomes])\n",
    "\n",
    "\n",
    "EMPTY_SPACE_REGION = html.Div(id = 'empty-space-region')\n",
    "\n",
    "RANGE_START_TB = dcc.Input(id = 'range-start-textbox',\n",
    "                           type = \"number\",\n",
    "                           placeholder = \"Start bp\",\n",
    "                           min = 1,\n",
    "                           debounce = True,\n",
    "                           className = 'class-range')\n",
    "\n",
    "RANGE_END_TB = dcc.Input(id = 'range-end-textbox',\n",
    "                         type = \"number\",\n",
    "                         placeholder = \"End bp\",\n",
    "                         min = 2,\n",
    "                         debounce = True,\n",
    "                         className = 'class-range')\n",
    "\n",
    "\n",
    "REGION_DIV = html.Div(id = 'chromosome-dropdown-div',\n",
    "                      children = [REGION_LABEL,\n",
    "                                  CHR_DD,\n",
    "                                  EMPTY_SPACE_REGION,\n",
    "                                  RANGE_START_TB,\n",
    "                                  RANGE_END_TB],\n",
    "                      className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 519,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for phenotypes\n",
    "# PHENO_LINE = html.Div(id = 'pheno-line-div')\n",
    "\n",
    "PHENO_LABEL = html.Div([html.P([\"Phenotype\",\n",
    "                                question_icon],\n",
    "                               id = 'pheno-label'),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Select phenotype of interest availible in GWAS Catalog.\\n\\nSVs are selected if they have an LD calculation with an associated SNP for the phenotype.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'pheno-label',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'pheno-label-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 520,
   "metadata": {},
   "outputs": [],
   "source": [
    "phenotypes = sorted(df_full_join.select(\"Phenotype\")\n",
    "                                .unique()\n",
    "                                .to_series()\n",
    "                                .to_list())\n",
    "phenotypes.insert(0, \"Any\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 521,
   "metadata": {},
   "outputs": [],
   "source": [
    "PHENO_DD = dcc.Dropdown(id = 'phenotype-dropdown',\n",
    "                        placeholder = \"Select a phenotype of interest\",\n",
    "                        options = [{\"label\" : str(i),\n",
    "                                  \"value\" : str(i)} for i in phenotypes],\n",
    "                        optionHeight = 75)\n",
    "\n",
    "PHENO_DD_DIV = html.Div(id = 'phenotype-dropdown-div',\n",
    "                        children = [# PHENO_LINE,\n",
    "                                    PHENO_LABEL,\n",
    "                                    PHENO_DD],\n",
    "                        className = 'class-section-filter-div')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 522,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make dropdown box for genes\n",
    "GENE_LABEL = html.P(\"Gene\")\n",
    "\n",
    "genes = sorted(list(set(df_gene[\"gene\"])))\n",
    "\n",
    "GENE_DD = dcc.Dropdown(id = 'gene-dropdown',\n",
    "                       placeholder = \"Select a gene of interest\",\n",
    "                       options = [{\"label\" : str(i),\n",
    "                                   \"value\" : str(i)} for i in genes])\n",
    "\n",
    "GENE_DD_DIV = html.Div(id = 'gene-dropdown-div',\n",
    "                       children = [GENE_LABEL,\n",
    "                                   GENE_DD],\n",
    "                       className = 'class-filter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 523,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create a tab section for filter between gene and genomic region\n",
    "TABS_FILTER_DIV = html.Div([dbc.Tabs([dbc.Tab(children = [GENE_DD_DIV],\n",
    "                                         label = \"by Gene\",\n",
    "                                         tab_id = \"tab-gene\"),\n",
    "                                     dbc.Tab(children = [REGION_DIV],\n",
    "                                             label = \"by Region\",\n",
    "                                             tab_id = \"tab-region\")],\n",
    "                                     id = \"tab-filters\",\n",
    "                                     active_tab = \"tab-gene\")],\n",
    "                           className = 'class-section-filter-div')\n",
    "\n",
    "\n",
    "FILTER_DIV = html.Div(id = 'filter-div',\n",
    "                      children = [TABS_FILTER_DIV,\n",
    "                                  PHENO_DD_DIV],\n",
    "                      className = 'class-filter-div')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 524,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### TABLE ####\n",
    "\n",
    "# Make DataTable\n",
    "def make_table(chrom = chromosomes,\n",
    "               range_start = 0,\n",
    "               range_end = 250000000,\n",
    "               gene = genes,\n",
    "               pheno = phenotypes):\n",
    "\n",
    "    global tab_show\n",
    "\n",
    "    # Filter by chromosome\n",
    "    if isinstance(chrom, list):\n",
    "        sv_list_chr = df_anno_subset[\"SV_Name\"].to_list()\n",
    "    else:\n",
    "        chrom = f\"chr{chrom}\"\n",
    "        sv_list_chr = df_anno_subset.filter(pl.col(\"Chromosome\") == chrom)[\"SV_Name\"].to_list()\n",
    "\n",
    "    # Filter by range\n",
    "    sv_list_range = df_anno_subset.filter((pl.col(\"SV_Start\") >= range_start) & \\\n",
    "                                          (pl.col(\"SV_End\") <= range_end))[\"SV_Name\"].to_list()\n",
    "\n",
    "    # Filter by phenotype\n",
    "    if isinstance(pheno, list):\n",
    "        sv_list_pheno = df_anno_subset[\"SV_Name\"].to_list()\n",
    "    else:\n",
    "        sv_list_pheno = dict_pheno[pheno]\n",
    "\n",
    "    # Filter by gene\n",
    "    if isinstance(gene, list):\n",
    "        sv_list_gene = list(df_anno_subset[\"SV_Name\"])\n",
    "    else:\n",
    "        df_1gene = df_gene.filter((pl.col(\"feature\") == \"transcript\") & \\\n",
    "                                  (pl.col(\"gene\") == gene))\n",
    "        start = df_1gene[\"start\"].min()\n",
    "        end = df_1gene[\"end\"].max()\n",
    "        chrom = df_1gene.select(\"chr\").to_series()[0]\n",
    "        \n",
    "        sv_list_gene = df_sv_anno.filter((pl.col(\"Chromosome\") == chrom) & \\\n",
    "                                         (pl.col(\"Start\") >= (start - 100000)) & \\\n",
    "                                         (pl.col(\"End\") <= (end + 100000)))[\"SV Name\"].to_list()\n",
    "\n",
    "    # Find overlaps sv's from all lists\n",
    "    sv_list_complete = set(sv_list_pheno).intersection(sv_list_chr, sv_list_gene, sv_list_range)\n",
    "\n",
    "    tab_show = df_sv_anno.filter(pl.col(\"SV Name\").is_in(sv_list_complete)).select([\n",
    "        pl.col(\"SV Name\").alias(\"ID\"),\n",
    "        pl.col(\"Chromosome\").alias(\"Chrom\"),\n",
    "        pl.col(\"Start\"),\n",
    "        pl.col(\"End\"),\n",
    "        pl.col(\"Type\"),\n",
    "        pl.col(\"Length\").alias(\"Size (bp)\"),\n",
    "        pl.col(\"SV Sample AF\").alias(\"AF\")])\n",
    "\n",
    "    return dash_table.DataTable(id ='strvar-table',\n",
    "                                data = tab_show.to_dicts(),\n",
    "                                columns = [{'id': c, 'name': c} for c in [\"Chrom\", \"Start\", \"End\", \"Type\", \"Size (bp)\", \"AF\"]],\n",
    "                                page_size = 100,\n",
    "                                fixed_rows = {'headers': True},\n",
    "                                style_table = {'height': '400px',\n",
    "                                               'overflowY': 'auto'},\n",
    "                                style_cell = {'textAlign': 'center',\n",
    "                                              'font-family': 'Nunito Sans',\n",
    "                                              'backgroundColor': 'whitesmoke',\n",
    "                                              'minWidth' : '100px'},\n",
    "                                style_header = {'backgroundColor': 'turquoise',\n",
    "                                                'fontWeight': 'bold'},\n",
    "                                row_selectable = 'single',\n",
    "                                selected_rows = [])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 525,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SV annotation table\n",
    "def make_annotation_table(sv = \"None\"):\n",
    "    if sv == \"None\":\n",
    "        df_anno_subset_sv = pl.DataFrame()\n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = [],\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_anno_subset_sv = df_sv_anno.filter(pl.col(\"SV Name\") == sv)\n",
    "        df_anno_subset_sv = (df_anno_subset_sv.melt()\n",
    "                                              .rename({\"variable\": \"Header\",\n",
    "                                                       \"value\": \"Information\"}))\n",
    "        \n",
    "        return dash_table.DataTable(id ='anno-table',\n",
    "                                    data = df_anno_subset_sv.to_dicts(),\n",
    "                                    columns = [{'id': c, 'name': c} for c in df_anno_subset_sv.columns],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '400px',\n",
    "                                                   'overflowX': 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell = {'textAlign': 'left',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': \"Header\"},\n",
    "                                                              'backgroundColor': 'turquoise',\n",
    "                                                              'fontWeight': 'bold',\n",
    "                                                              'maxWidth' : '200px',\n",
    "                                                              'whiteSpace' : 'normal',\n",
    "                                                              'height' : 'auto'}],\n",
    "                                    style_header = {'display' : 'none',\n",
    "                                                    'height' : '0px'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 526,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "DTABLE = html.Div(id = 'strvar-table-div',\n",
    "                  children = [make_table()],\n",
    "                  className = 'sv-table')\n",
    "\n",
    "# Create a header for the SV table\n",
    "DTABLE_HEADER = html.Div([html.H4([\"SV Information\",\n",
    "                                   question_icon],\n",
    "                                   id = 'dtable-header',\n",
    "                                   style = {'textAlign' : 'center',\n",
    "                                            'color' : '#2F2E2D',\n",
    "                                            'height' : '30 px',\n",
    "                                            'text-transform': 'none'}),\n",
    "                          dbc.Tooltip(dcc.Markdown(\"List of SVs subsetted by the filters applied.\\n\\nSelect an SV to visualize LD with GWAS-associated SNPs.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                      target = 'dtable-header',\n",
    "                                      placement = \"auto\",\n",
    "                                      delay = {\"show\" : 500,\n",
    "                                               \"hide\" : 50},\n",
    "                                      id = 'dtable-header-tooltip')])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 527,
   "metadata": {},
   "outputs": [],
   "source": [
    "DTABLE_TEXTBOX = html.Div(id = 'dtable-textbox')\n",
    "\n",
    "\n",
    "DTABLE_DIV = html.Div(id = 'dtable-div',\n",
    "                      children = [DTABLE_HEADER,\n",
    "                                  DTABLE,\n",
    "                                  DTABLE_TEXTBOX],\n",
    "                      className = 'class-filter-div')\n",
    "\n",
    "\n",
    "ANNOTABLE = html.Div(id = 'anno-table-div',\n",
    "                     children = [make_annotation_table()],\n",
    "                     className = 'sv-anno-table')\n",
    "\n",
    "\n",
    "# Create the SV details collapse and table\n",
    "SV_TABLE_BUTTON = dbc.Button(\"SV Annotations\",\n",
    "                             id = \"sv-button-collapse\",\n",
    "                             color = \"secondary\",\n",
    "                             outline = True,\n",
    "                             size = \"sm\",\n",
    "                             n_clicks = 0)\n",
    "\n",
    "\n",
    "# Create the filter division\n",
    "ALL_FILTER_DIV = html.Div(id = 'all-filter-div',\n",
    "                          children = [FILTER_DIV,\n",
    "                                      DTABLE_DIV],\n",
    "                                      style = {'display' : 'flex',\n",
    "                                               'flexWrap' : \"wrap\",\n",
    "                                               'justifyContent' : \"space-evenly\",\n",
    "                                               'alignContent' : 'space-around'},\n",
    "                                               className = 'sections')\n",
    "\n",
    "SV_TABLE_BUTTON_DIV = html.Div(id = \"sv-button-div\",\n",
    "                               children = [SV_TABLE_BUTTON])\n",
    "\n",
    "\n",
    "SV_TABLE_COLLAPSE = dbc.Collapse(children = [ANNOTABLE],\n",
    "                                 id = \"sv-table-collapse\",\n",
    "                                 is_open = False)\n",
    "\n",
    "SV_TABLE = html.Div(id = 'sv-table-collapse-div',\n",
    "                    children = [SV_TABLE_COLLAPSE],\n",
    "                    style = {'display' : 'flex',\n",
    "                             'flexWrap' : \"wrap\",\n",
    "                             'justifyContent' : \"space-evenly\",\n",
    "                             'alignContent' : 'space-around'})\n",
    "\n",
    "\n",
    "SV_TABLE_DIV = html.Div(id = 'sv-table-div',\n",
    "                        children = [SV_TABLE,\n",
    "                                    SV_TABLE_BUTTON_DIV],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 528,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### PLOT ####\n",
    "\n",
    "# Make scatter plot to compare r2 and D' values as seen in DataTable\n",
    "def make_plot(sv = \"None\",\n",
    "              pheno = \"None\",\n",
    "              toggle = False,\n",
    "              toggle_pheno = False):\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage_value = \"D'\"\n",
    "        size_value = \"R2\"\n",
    "        legend_title = \"D'\"\n",
    "    else:\n",
    "        linkage_value = \"R2\"\n",
    "        size_value = \"D'\"\n",
    "        legend_title = \"r<sup>2</sup>\"\n",
    "\n",
    "    global tab, sv_start_plot, sv_end_plot, extra_points, extra_points_pheno\n",
    "\n",
    "    if sv == \"None\":\n",
    "        tab = pl.DataFrame({\"SNP_Position\": [],\n",
    "                            \"-log10(P-Value)\": []})\n",
    "        \n",
    "        sc_plot = px.scatter(data_frame = tab.to_pandas(),\n",
    "                             x = \"SNP_Position\",\n",
    "                             y = \"-log10(P-Value)\",\n",
    "                             labels = {\"-log10(P-Value)\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                       \"SNP_Position\" : \"Position in hg38 (bp)\"},\n",
    "                             width = 850,\n",
    "                             height = 650,\n",
    "                             template = \"ggplot2\")\n",
    "\n",
    "        sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                              autosize = False,\n",
    "                              font_family = \"Nunito Sans\",\n",
    "                              font_size = 20)\n",
    "    else:\n",
    "        tab = df_full_join.filter(pl.col(\"SV_Name\") == sv)\n",
    "\n",
    "        # Display +/- 1KB from SV position\n",
    "        sv_start = df_anno_subset.filter(pl.col(\"SV_Name\") == sv).select(\"SV_Start\").to_numpy()[0, 0]\n",
    "        sv_end = df_anno_subset.filter(pl.col(\"SV_Name\") == sv).select(\"SV_End\").to_numpy()[0, 0]\n",
    "\n",
    "        sv_start_plot = sv_start - 1000\n",
    "        sv_end_plot = sv_end + 1000\n",
    "\n",
    "\n",
    "        if pheno == \"None\":\n",
    "            tab_lab = tab.to_dicts()\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"],\n",
    "                        row[\"SV_Name\"],\n",
    "                        row[\"SV_Start\"],\n",
    "                        row[\"SV_End\"],\n",
    "                        row[\"SV_Type\"]]\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab.to_pandas(),\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = linkage_value,\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"R2\" : \"r<sup>2</sup>\"},\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                #  marginal_x = \"rug\",\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(marker = dict(color = \"rgba(57,201,187,0.5)\",\n",
    "                                                line = dict(color = \"rgba(57,201,187,1)\"),\n",
    "                                                size = 10),\n",
    "                                  hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                  '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                  '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                  '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                  '<br>P-Value: %{customdata[4]}' + \n",
    "                                                  '<br>r2: %{customdata[5]}' + \n",
    "                                                  \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab])\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                  autosize = False,\n",
    "                                  font_family = \"Nunito Sans\",\n",
    "                                  font_size = 20)\n",
    "                                #   xaxis_range = [sv_start_plot, sv_end_plot],\n",
    "                                #   yaxis_range = [-0.6, 1])\n",
    "\n",
    "             # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = tab.select(\"SNP_Position\").min().item() - 100000\n",
    "            max_SNP = tab.select(\"SNP_Position\").max().item() + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = tab.select(\"SNP_Position\").min()[0, 0] - 100000\n",
    "                max_SNP = tab.select(\"SNP_Position\").max()[0, 0] + 100000\n",
    "\n",
    "\n",
    "            chromo = tab.select(\"Chromosome\").unique().item()\n",
    "            svname = tab.select(\"SV_Name\").unique().item()\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -0.1,\n",
    "                              y1 = -0.15,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                           width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.01,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-0.07]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + \n",
    "                                                         '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                         '<br>Start: %{customdata[8]}' + \n",
    "                                                         '<br>End: %{customdata[9]}' + \n",
    "                                                         '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Add extra points from df_sv with R2 and D' values\n",
    "            extra_points = df_sv_snp_join.filter(pl.col(\"SV_Name\") == sv)\n",
    "\n",
    "            extra_points = extra_points.filter(~pl.col(\"SNP_Name_dbSNP\").is_in(tab.select(\"SNP_Name_dbSNP\").to_series()))\n",
    "\n",
    "             # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dicts()\n",
    "\n",
    "            def extra_label_maker(row):\n",
    "\n",
    "                return [row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"]]\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                           y = extra_points.select(linkage_value).to_series().to_list(),\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>SNP Name: %{customdata[0]}</b>' +\n",
    "                                                           '<br>Chromosome: %{customdata[1]}' + \n",
    "                                                           '<br>SNP Position: %{customdata[2]}' + \n",
    "                                                           '<br>r2: %{customdata[3]}' + \n",
    "                                                           \"<br>D': %{customdata[4]}\",\n",
    "                                           customdata = [extra_label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"Not in \\nGWAS Catalog\"))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene.filter((pl.col(\"chr\") == chromo) &\n",
    "                                   (pl.col(\"start\") >= min_SNP) &\n",
    "                                   (pl.col(\"end\") <= max_SNP))\n",
    "\n",
    "            exons_genes = (exons.groupby([\"chr\", \"gene\", \"strand\"])\n",
    "                                .agg([pl.col(\"start\").min().alias(\"start\"),\n",
    "                                      pl.col(\"end\").max().alias(\"end\")]))\n",
    "            \n",
    "            # Compute `gene_line_position` for each row in `exons_genes`\n",
    "            exons_genes = exons_genes.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene).select(\"y_coord_nopheno\").item()))\n",
    "                                     .alias(\"gene_line_position\"))\n",
    "\n",
    "\n",
    "            for row in exons_genes.iter_rows(named = True):\n",
    "\n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = row[\"gene_line_position\"],\n",
    "                                  y1 = row[\"gene_line_position\"],\n",
    "                                  line_width = 3,\n",
    "                                  line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 0.060,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons.filter(pl.col(\"feature\") == \"exon\")\n",
    "\n",
    "            # Draw exons in the plot\n",
    "            for row in exons_exons.iter_rows(named=True):\n",
    "                gene_line_position = float(df_transcript.filter(pl.col(\"gene\") == row[\"gene\"]).select(\"y_coord_nopheno\").item())\n",
    "                sc_plot.add_shape(type = \"rect\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = gene_line_position + 0.02,\n",
    "                                  y1 = gene_line_position - 0.02,\n",
    "                                  line = dict(color = \"mediumseagreen\", \n",
    "                                            width = 2),\n",
    "                                  fillcolor = \"mediumseagreen\")\n",
    " \n",
    "\n",
    "        else:\n",
    "            tab = tab.filter(pl.col(\"Phenotype\") == pheno)\n",
    "\n",
    "            tab_lab = tab.to_dicts()\n",
    "\n",
    "            def tab_label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_dbSNP\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"],\n",
    "                        row[\"R2\"],\n",
    "                        row[\"D'\"],\n",
    "                        row[\"SV_Name\"],\n",
    "                        row[\"SV_Start\"],\n",
    "                        row[\"SV_End\"],\n",
    "                        row[\"SV_Type\"]]\n",
    "\n",
    "            sc_plot = px.scatter(data_frame = tab.to_pandas(),\n",
    "                                 x = \"SNP_Position\",\n",
    "                                 y = \"P-Value_log10\",\n",
    "                                 labels = {\"SNP_Position\" : \"Position in hg38 (bp)\",\n",
    "                                           \"P-Value_log10\" : \"-log<sub>10</sub> (P-Value)\",\n",
    "                                           linkage_value : legend_title},\n",
    "                                 color = linkage_value,\n",
    "                                 color_continuous_scale = [(0, \"#4643CE\"),\n",
    "                                                           (0.2, \"#4643CE\"),\n",
    "                                                           (0.2, \"#90CAEE\"),\n",
    "                                                           (0.4, \"#90CAEE\"),\n",
    "                                                           (0.4, \"#46E17C\"),\n",
    "                                                           (0.6, \"#46E17C\"),\n",
    "                                                           (0.6, \"#EEC06B\"),\n",
    "                                                           (0.8, \"#EEC06B\"),\n",
    "                                                           (0.8, \"#CE5C5C\"),\n",
    "                                                           (1, \"#CE5C5C\")],\n",
    "                                 range_color = [0, 1],\n",
    "                                 width = 850,\n",
    "                                 height = 650,\n",
    "                                 template = \"ggplot2\")\n",
    "\n",
    "            sc_plot.update_traces(hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                  '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                  '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                  '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                  '<br>P-Value: %{customdata[4]}' + \n",
    "                                                  '<br>r2: %{customdata[5]}' + \n",
    "                                                  \"<br>D': %{customdata[6]}\",\n",
    "                                  customdata = [tab_label_maker(row) for row in tab_lab],\n",
    "                                  marker = dict(size = 10))\n",
    "\n",
    "            sc_plot.update_xaxes(showspikes = True,\n",
    "                                 spikemode = \"across\")\n",
    "\n",
    "            sc_plot.update_layout(margin = dict(l = 20, r = 20, t = 50, b = 50),\n",
    "                                 autosize = True,\n",
    "                                 font_family = \"Nunito Sans\",\n",
    "                                 font_size = 20,\n",
    "                                 coloraxis_colorbar = dict(lenmode = \"pixels\",\n",
    "                                                           len = 150,\n",
    "                                                           yanchor = \"top\",\n",
    "                                                           y = 1,\n",
    "                                                           dtick = 0.2))\n",
    "\n",
    "            # Extract min and max SNP positions in chosen SV\n",
    "            min_SNP = tab.select(\"SNP_Position\").min().item() - 100000\n",
    "            max_SNP = tab.select(\"SNP_Position\").max().item() + 100000\n",
    "\n",
    "            if sv_start < min_SNP:\n",
    "                min_SNP = sv_start - 100000\n",
    "\n",
    "            elif sv_start > max_SNP:\n",
    "                max_SNP = sv_end  + 100000\n",
    "\n",
    "            else:\n",
    "                min_SNP = tab.select(\"SNP_Position\").min()[0, 0] - 100000\n",
    "                max_SNP = tab.select(\"SNP_Position\").max()[0, 0] + 100000\n",
    "\n",
    "\n",
    "            chromo = tab.select(\"Chromosome\").unique().item()\n",
    "            svname = tab.select(\"SV_Name\").unique().item()\n",
    "\n",
    "            # Add points from gwas for pheno, without r2 and D' values\n",
    "            extra_points = (df_gwas.filter((pl.col(\"Chromosome\") == chromo) &\n",
    "                                           (pl.col(\"SNP_Position\") >= min_SNP) &\n",
    "                                           (pl.col(\"SNP_Position\") <= max_SNP) &\n",
    "                                           (pl.col(\"Phenotype\") == pheno))\n",
    "                                   .filter(~pl.col(\"SNP_Name_GWAS\").is_in(tab.select(\"SNP_Name_GWAS\").to_series())))\n",
    "\n",
    "            extra_points = (extra_points.with_columns(pl.when(pl.col(\"P-Value\") < 1e-50)\n",
    "                                                        .then(1e-50)\n",
    "                                                        .otherwise(pl.col(\"P-Value\"))\n",
    "                                                        .alias(\"P-Value\"),\n",
    "                                                        (-pl.col(\"P-Value\").log10()).alias(\"P-Value_log10\")))\n",
    "\n",
    "            # Create a list of labels for hovertext\n",
    "            df_lab = extra_points.to_dicts()\n",
    "\n",
    "            def label_maker(row):\n",
    "\n",
    "                return [row[\"Phenotype\"],\n",
    "                        row[\"SNP_Name_GWAS\"],\n",
    "                        row[\"Chromosome\"],\n",
    "                        row[\"SNP_Position\"],\n",
    "                        row[\"P-Value\"]]\n",
    "\n",
    "\n",
    "            sc_plot.add_trace(go.Scattergl(x = extra_points.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                           y = extra_points.select(\"P-Value_log10\").to_series().to_list(),\n",
    "                                           mode = \"markers\",\n",
    "                                           marker = dict(size = 10,\n",
    "                                                         color = \"rgba(168,168,168,0.5)\",\n",
    "                                                         line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                           hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                           '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                           '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                           '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                           '<br>P-Value: %{customdata[4]}',\n",
    "                                           customdata = [label_maker(row) for row in df_lab],\n",
    "                                           showlegend = False,\n",
    "                                           name = \"No LD Data\"))\n",
    "\n",
    "            # Add points from other phenos in LD with SV\n",
    "            if toggle_pheno == True:\n",
    "                extra_points_pheno = (df_full_join.filter((pl.col(\"Chromosome\") == chromo) &\n",
    "                                                          (pl.col(\"SNP_Position\") >= min_SNP) &\n",
    "                                                          (pl.col(\"SNP_Position\") <= max_SNP) &\n",
    "                                                          (pl.col(\"SV_Name\") == sv))\n",
    "                                                  .filter(~pl.col(\"SNP_Name_GWAS\").is_in(tab.select(\"SNP_Name_GWAS\").to_series())))\n",
    "\n",
    "\n",
    "                # Create a list of labels for hovertext\n",
    "                df_lab_2 = extra_points_pheno.to_dicts()\n",
    "\n",
    "                def label_maker(row):\n",
    "\n",
    "                    return [row[\"Phenotype\"],\n",
    "                            row[\"SNP_Name_GWAS\"],\n",
    "                            row[\"Chromosome\"],\n",
    "                            row[\"SNP_Position\"],\n",
    "                            row[\"P-Value\"],\n",
    "                            row[\"R2\"],\n",
    "                            row[\"D'\"]]\n",
    "\n",
    "\n",
    "                sc_plot.add_trace(go.Scattergl(x = extra_points_pheno.select(\"SNP_Position\").to_series().to_list(),\n",
    "                                               y = extra_points_pheno.select(\"P-Value_log10\").to_series().to_list(),\n",
    "                                               mode = \"markers\",\n",
    "                                               marker = dict(size = 10,\n",
    "                                                             color = \"rgba(168,168,168,0.5)\",\n",
    "                                                             line = dict(color = \"rgba(168,168,168,1)\")),\n",
    "                                               hovertemplate = '<br><b>Phenotype:</b> %{customdata[0]}<br>' + \n",
    "                                                               '<br>SNP Name: %{customdata[1]}' + \n",
    "                                                               '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                               '<br>SNP Position: %{customdata[3]}' + \n",
    "                                                               '<br>P-Value: %{customdata[4]}' + \n",
    "                                                               '<br>r2: %{customdata[5]}' + \n",
    "                                                               \"<br>D': %{customdata[6]}\",\n",
    "                                               customdata = [label_maker(row) for row in df_lab_2],\n",
    "                                               showlegend = False,\n",
    "                                               name = \"Other Phenotypes\"))\n",
    "            else:\n",
    "                pass\n",
    "\n",
    "\n",
    "            sc_plot.add_shape(type = \"rect\",\n",
    "                              x0 = sv_start,\n",
    "                              x1 = sv_end,\n",
    "                              y0 = -2,\n",
    "                              y1 = -3,\n",
    "                              line = dict(color = \"darkslategrey\",\n",
    "                                          width = 1),\n",
    "                              fillcolor = \"lightsalmon\")\n",
    "\n",
    "            sc_plot.add_annotation(x = ((sv_end - sv_start) / 2) + sv_start,\n",
    "                                   y = -0.5,\n",
    "                                   text = svname,\n",
    "                                   font = dict(size = 12),\n",
    "                                   showarrow = False)\n",
    "\n",
    "            sc_plot.add_trace(go.Scatter(x = np.array([((sv_end - sv_start) / 2) + sv_start]),\n",
    "                                         y = np.array([-1.5]),\n",
    "                                         mode = \"markers\",\n",
    "                                         marker = dict(symbol = \"triangle-down\",\n",
    "                                                       size = 15,\n",
    "                                                       color = \"lightsalmon\"),\n",
    "                                         showlegend = False,\n",
    "                                         name = svname,\n",
    "                                         hovertemplate = '<br><b>SV Name:</b> %{customdata[7]}<br>' + \n",
    "                                                         '<br>Chromosome: %{customdata[2]}' + \n",
    "                                                         '<br>Start: %{customdata[8]}' + \n",
    "                                                         '<br>End: %{customdata[9]}' + \n",
    "                                                         '<br>Type: %{customdata[10]}',\n",
    "                                         customdata = [tab_label_maker(tab_lab[0])]))\n",
    "\n",
    "            # Use min and max SNP positions to extract required rows from df_exons to obtain gene information\n",
    "            exons = df_gene.filter((pl.col(\"chr\") == chromo) &\n",
    "                                   (pl.col(\"start\") >= min_SNP) &\n",
    "                                   (pl.col(\"end\") <= max_SNP))\n",
    "\n",
    "            exons_genes = (exons.groupby([\"chr\", \"gene\", \"strand\"])\n",
    "                                .agg([pl.col(\"start\").min().alias(\"start\"),\n",
    "                                      pl.col(\"end\").max().alias(\"end\")]))\n",
    "            \n",
    "            # Compute `gene_line_position` for each row in `exons_genes`\n",
    "            exons_genes = exons_genes.with_columns(pl.col(\"gene\")\n",
    "                                     .apply(lambda gene: float(df_transcript.filter(pl.col(\"gene\") == gene).select(\"y_coord_pheno\").item()))\n",
    "                                     .alias(\"gene_line_position\"))\n",
    "\n",
    "\n",
    "            for row in exons_genes.iter_rows(named = True):\n",
    "\n",
    "                sc_plot.add_shape(type = \"line\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = row[\"gene_line_position\"],\n",
    "                                  y1 = row[\"gene_line_position\"],\n",
    "                                  line_width = 3,\n",
    "                                  line_color = \"darkslategrey\")\n",
    "\n",
    "                text_anno_x = round((row[\"end\"] - row[\"start\"]) / 2) + row[\"start\"]\n",
    "\n",
    "                if row[\"strand\"] == \"+\":\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] +  1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2192</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> forward',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "                else:\n",
    "                    sc_plot.add_annotation(x = text_anno_x,\n",
    "                                           y = row[\"gene_line_position\"] + 1.75,\n",
    "                                           text = row[\"gene\"] + '<span style=\"font-size:15px\">\\u2190</span>',\n",
    "                                           font = dict(size = 10),\n",
    "                                           hovertext = row[\"gene\"] + '<br><b>Gene direction:</b> reverse',\n",
    "                                           name = \"Gene\",\n",
    "                                           showarrow = False)\n",
    "\n",
    "\n",
    "            exons_exons = exons.filter(pl.col(\"feature\") == \"exon\")\n",
    "\n",
    "            # Draw exons in the plot\n",
    "            for row in exons_exons.iter_rows(named=True):\n",
    "                gene_line_position = float(df_transcript.filter(pl.col(\"gene\") == row[\"gene\"]).select(\"y_coord_pheno\").item())\n",
    "                sc_plot.add_shape(type = \"rect\",\n",
    "                                  x0 = row[\"start\"],\n",
    "                                  x1 = row[\"end\"],\n",
    "                                  y0 = gene_line_position + 0.75,\n",
    "                                  y1 = gene_line_position - 0.75,\n",
    "                                  line = dict(color = \"mediumseagreen\", \n",
    "                                            width = 2),\n",
    "                                  fillcolor = \"mediumseagreen\")\n",
    "\n",
    "\n",
    "    return sc_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 529,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SCATPLOT = dcc.Loading(id = \"plot-loading\",\n",
    "                       children = [dcc.Graph(id = \"scatter-plot\",\n",
    "                                             figure = make_plot(),\n",
    "                                             config = {\"modeBarButtonsToRemove\": [\"select2d\",\n",
    "                                                                                  \"lasso2d\",\n",
    "                                                                                  \"zoomIn2d\",\n",
    "                                                                                  \"zoomOut2d\"],\n",
    "                                                        \"toImageButtonOptions\": {\"format\": \"svg\",\n",
    "                                                                                 \"filename\": \"gwas_svatalog_plot\",\n",
    "                                                                                 \"height\": 500,\n",
    "                                                                                 \"width\": 700,\n",
    "                                                                                 \"scale\": 1}})],\n",
    "                       type = \"dot\",\n",
    "                       color = \"turquoise\",\n",
    "                       parent_className = \"loading_wrapper\")\n",
    "\n",
    "\n",
    "TOGGLE_SWITCH = html.Div([dbc.Row([dbc.Col([\"D'\"],\n",
    "                                           id = 'toggle-d'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'toggle-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"r\", html.Sup(2,\n",
    "                                                          style = {'font-size' : '15px'})],\n",
    "                                            id = 'toggle-r')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region'),\n",
    "                          dbc.Row([html.P([\"Show Other Phenotypes\",\n",
    "                                           question_icon],\n",
    "                                          id = 'toggle-pheno-header'),\n",
    "                                   dbc.Tooltip(dcc.Markdown(\"Show GWAS-associated SNPs from other phenotypes in respect to selected SV.\",\n",
    "                                                            style = {'white-space':'pre-wrap'}),\n",
    "                                               target = 'toggle-pheno-header',\n",
    "                                               placement = \"auto\",\n",
    "                                               delay = {\"show\" : 500,\n",
    "                                                        \"hide\" : 50},\n",
    "                                               id = 'toggle-pheno-header-tooltip')],\n",
    "                                  id = 'toggle-pheno-header-row'),\n",
    "                          dbc.Row([dbc.Col([\"OFF\"],\n",
    "                                           id = 'toggle-off'),\n",
    "                                   dbc.Col(daq.ToggleSwitch(id = 'other-pheno-switch',\n",
    "                                                            color = \"lightsalmon\")),\n",
    "                                   dbc.Col([\"ON\"],\n",
    "                                            id = 'toggle-on')],\n",
    "                                   align = \"center\",\n",
    "                                   id = 'toggle-region-pheno'),\n",
    "                          dbc.Row([dbc.Button(\"Export SNP Data to CSV\",\n",
    "                                               id = 'download-button',\n",
    "                                               color = \"secondary\",\n",
    "                                               outline = True,\n",
    "                                               size = \"sm\",),\n",
    "                                   dcc.Download(id = 'download-csv')],\n",
    "                                   id = 'download-region')])\n",
    "\n",
    "\n",
    "\n",
    "SCATPLOT_DIV = html.Div(id = 'scatter-plot-div',\n",
    "                        children = [SCATPLOT,\n",
    "                                    TOGGLE_SWITCH],\n",
    "                        className = 'sections')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 530,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Make SNP table from GWAS data based on clickData from plot\n",
    "def make_SNP_table(snp = \"None\"):\n",
    "    if snp == \"None\":\n",
    "        df_snp = pl.DataFrame()\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = [],\n",
    "                                    columns = [],\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '500px', 'overflowY': 'auto'},\n",
    "                                    style_cell = {'textAlign': 'center',\n",
    "                                                  'font-family': 'Nunito Sans',\n",
    "                                                  'backgroundColor': 'whitesmoke'},\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold'})\n",
    "\n",
    "    else:\n",
    "        df_snp = (df_full_join.filter((pl.col(\"SNP_Name_dbSNP\") == snp) | (pl.col(\"SNP_Name_GWAS\") == snp))\n",
    "                              .with_columns([\n",
    "                pl.format(\"[{}](https://{})\", pl.col(\"Pubmed_ID\"), pl.col(\"Link\")).alias(\"Link\"),\n",
    "                pl.format(\"[{}](https://www.ncbi.nlm.nih.gov/snp/{})\", pl.col(\"SNP_Name_dbSNP\"), pl.col(\"SNP_Name_dbSNP\")).alias(\"SNP_Name_dbSNP\"),\n",
    "                pl.format(\"[{}](https://www.ebi.ac.uk/gwas/variants/{})\", pl.col(\"SNP_Name_GWAS\"), pl.col(\"SNP_Name_GWAS\")).alias(\"SNP_Name_GWAS\")])\n",
    "                              .select([\n",
    "                pl.col(\"Chromosome\").alias(\"Chrom\"),\n",
    "                pl.col(\"SNP_Position\").alias(\"SNP Position\"),\n",
    "                pl.col(\"SNP_Name_dbSNP\").alias(\"SNP Name: dbSNP\"),\n",
    "                pl.col(\"SNP_Name_GWAS\").alias(\"SNP Name: GWAS\"),\n",
    "                pl.col(\"Reference_Allele\").alias(\"Reference Allele\"),\n",
    "                pl.col(\"Alternate_Allele\").alias(\"Alternate Allele\"),\n",
    "                pl.col(\"Risk_Allele\").alias(\"Risk Allele\"),\n",
    "                pl.col(\"Risk_Allele_Frequency\").alias(\"Risk AF\"),\n",
    "                pl.col(\"Sample_AF\").alias(\"Sample AF\"),\n",
    "                pl.col(\"gnomAD_nfe_AF\").alias(\"gnomAD NFE AF\"),\n",
    "                pl.col(\"Phenotype\"),\n",
    "                pl.col(\"P-Value\"),\n",
    "                pl.col(\"Study\"),\n",
    "                pl.col(\"Link\").alias(\"Pubmed Link\")])\n",
    "                               .unique())\n",
    "\n",
    "        return dash_table.DataTable(id ='snp-table',\n",
    "                                    data = df_snp.to_dicts(),\n",
    "                                    columns = [{'id': c, 'name': c, 'presentation': 'markdown'} \n",
    "                                               if (c == \"Pubmed Link\" or c == \"SNP Name: dbSNP\" or c == \"SNP Name: GWAS\") \n",
    "                                               else ({'id': c, 'name': c, 'type':'numeric', 'format': {'specifier': '.3f'}} \n",
    "                                                     if (c == \"Risk AF\" or c == \"Sample AF\" or c == \"gnomAD NFE AF\") \n",
    "                                                     else {'id': c, 'name': c}) for c in df_snp.columns],\n",
    "                                    markdown_options = {\"html\" : True,\n",
    "                                                        \"link_target\": \"_blank\"},\n",
    "                                    page_size = 400,\n",
    "                                    fixed_rows = {'headers': True},\n",
    "                                    style_table = {'height': '500px',\n",
    "                                                   'minWidth': '100%',\n",
    "                                                   'width' : '100%',\n",
    "                                                   'maxWidth' : '100%',\n",
    "                                                   'overflowX' : 'auto',\n",
    "                                                   'overflowY': 'auto'},\n",
    "                                    style_data = {'whiteSpace' : 'normal',\n",
    "                                                  'height' : 'auto'},\n",
    "                                    style_cell_conditional = [{'if': {'column_id': [\"Chrom\",\n",
    "                                                                                    \"SNP Position\",\n",
    "                                                                                    \"SNP Name: dbSNP\",\n",
    "                                                                                    \"SNP Name: GWAS\",\n",
    "                                                                                    \"Reference Allele\",\n",
    "                                                                                    \"Alternate Allele\",\n",
    "                                                                                    \"Risk Allele\",\n",
    "                                                                                    \"Risk AF\",\n",
    "                                                                                    \"Sample AF\",\n",
    "                                                                                    \"gnomAD NFE AF\",\n",
    "                                                                                    \"P-Value\",\n",
    "                                                                                    \"Pubmed Link\"]},\n",
    "                                                               'minWidth' : '100px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke'},\n",
    "                                                               {'if': {'column_id': [\"Study\",\n",
    "                                                                                     \"Phenotype\"]},\n",
    "                                                               'maxWidth' : '200px',\n",
    "                                                               'textAlign': 'left',\n",
    "                                                               'font-family': 'Nunito Sans',\n",
    "                                                               'backgroundColor': 'whitesmoke',\n",
    "                                                               'whiteSpace' : 'normal',\n",
    "                                                               'height' : 'auto'}],\n",
    "                                    style_header = {'backgroundColor': 'turquoise',\n",
    "                                                    'fontWeight': 'bold',\n",
    "                                                    'textAlign' : 'left',\n",
    "                                                    'whiteSpace' : 'normal',\n",
    "                                                    'height' : 'auto'})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 531,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SNP_TABLE = html.Div([html.Div(id = 'snp-table-div',\n",
    "                               children = [make_SNP_table()],\n",
    "                               className = 'sections'),\n",
    "                      dbc.Modal([dbc.ModalHeader(),\n",
    "                                 dbc.ModalBody(\"SNP table populated below.\")],\n",
    "                                 id = 'modal-snp-table',\n",
    "                                 is_open = False)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 532,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### HEADERS ####\n",
    "\n",
    "## Navigation Bar to match LocusFocus\n",
    "NAVBAR = dbc.NavbarSimple(id = 'navbar-lf',\n",
    "                          brand = \"LocusFocus\",\n",
    "                          brand_href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                          brand_external_link = \"_blank\",\n",
    "                          color = \"#40e0d0\",\n",
    "                          links_left = True,\n",
    "                          expand = \"lg\",\n",
    "                          children = [dbc.NavItem(dbc.NavLink(\"Colocalization\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Set-Based Test\",\n",
    "                                                              href = \"https://locusfocus.research.sickkids.ca/setbasedtest\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"GWAS SVatalog\",\n",
    "                                                              href = \"https://svatalog.research.sickkids.ca/\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Documentation\",\n",
    "                                                              href = \"https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Contact Us\",\n",
    "                                                              id = 'button-modal-contact-us',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Contact Us\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         <h5 children=\"GWAS SVatalog\"  />\n",
    "                                                                                         Shalvi Chirmade <span style=\"color: #2ba089\" children=\"shalvi.chirmade@sickkids.ca\" />\n",
    "\n",
    "                                                                                         &nbsp;\n",
    "\n",
    "                                                                                         <h5 children=\"LocusFocus\" />\n",
    "                                                                                         Mackenzie Frew <span style=\"color: #2ba089\" children=\"mackenzie.frew@sickkids.ca\" />\n",
    "                                                                                        ''',\n",
    "                                                                                        dangerously_allow_html = True,\n",
    "                                                                                        id = 'contact-us-mrkdwn'))],\n",
    "                                                              id = 'modal-contact-us',\n",
    "                                                              is_open = False)]),\n",
    "                                      dbc.NavItem(dbc.NavLink(\"Subscribe\",\n",
    "                                                              href = \"https://mailchi.mp/752ab1c4d516/locusfocus\",\n",
    "                                                              target = \"_blank\",\n",
    "                                                              class_name = 'navbar-items')),\n",
    "                                      dbc.NavItem([dbc.Button(\"Citation\",\n",
    "                                                              id = 'button-modal-citation',\n",
    "                                                              n_clicks = 0,\n",
    "                                                              class_name = 'navbar-items'),\n",
    "                                                   dbc.Modal([dbc.ModalHeader(dbc.ModalTitle(\"Citation\")),\n",
    "                                                              dbc.ModalBody(dcc.Markdown('''\n",
    "                                                                                         *to be published*\n",
    "                                                                                        '''))],\n",
    "                                                              id = 'modal-citation',\n",
    "                                                              is_open = False)])])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Logo for the top of the webpage, disclaimer and info\n",
    "GSV_LOGO = html.Img(src = \"assets/gwas-svatalog-name.png\",\n",
    "                    id = 'gsv-logo-image')\n",
    "\n",
    "TCAG_LOGO = html.Img(src = \"assets/tcaglogo.png\",\n",
    "                     id = 'tcag-logo-image')\n",
    "\n",
    "SK_LOGO = html.Img(src = \"assets/SKlogo.svg\",\n",
    "                   id = 'sk-logo-image')\n",
    "\n",
    "DISCLAIMER_BUTTON = dbc.Button(\"Disclaimer\",\n",
    "                               color = \"secondary\",\n",
    "                               outline = True,\n",
    "                               size = \"sm\",\n",
    "                               id = 'disclaimer-button')\n",
    "\n",
    "DISCLAIMER_COLLAPSE = dbc.Collapse(dbc.Card(dbc.CardBody(\"Database constructed from predominantly European population of 101 individuals with Cystic Fibrosis (CF). The alleles affected by CF aside, the remainder of the genome is comparable to a healthy population of European descent (citation TBD). Genomic location is referenced against GRCh38.\")),\n",
    "                                   id = 'disclaimer-collapse',\n",
    "                                   is_open = False)\n",
    "\n",
    "DISCLAIMER_DIV = html.Div(id = 'disclaimer-div',\n",
    "                          children = [DISCLAIMER_BUTTON,\n",
    "                                      DISCLAIMER_COLLAPSE])\n",
    "\n",
    "# DOCUMENTATION = dcc.Markdown('''\n",
    "# Documentation for GWAS SVatalog can be found [here](https://gwas-svatalog-docs.readthedocs.io/en/latest/index.html).\n",
    "# Colocalization testing of GWAS loci across various datasets can be conducted at [LocusFocus] (https://locusfocus.research.sickkids.ca/).\n",
    "# ''',\n",
    "#                              id = 'doc-markdown')\n",
    "\n",
    "# DISC_DOC_DIV = html.Div(id = 'disc-doc-div',\n",
    "#                         children = [DOCUMENTATION,\n",
    "#                                     DISCLAIMER_DIV])\n",
    "\n",
    "GSV_LOGO_DIV = html.Div(id = 'gsv-logo-image-div',\n",
    "                        children = [GSV_LOGO])\n",
    "\n",
    "COMP_LOGO_DIV = html.Div(id = 'comp-logo-image-div',\n",
    "                         children = [TCAG_LOGO,\n",
    "                                     SK_LOGO,\n",
    "                                     DISCLAIMER_DIV])\n",
    "\n",
    "LOGO_DIV = html.Div(id = 'logo-image-div',\n",
    "                    children = [GSV_LOGO_DIV,\n",
    "                                COMP_LOGO_DIV])\n",
    "\n",
    "\n",
    "HEADER_DIV = html.Div(id = 'header-line-div')\n",
    "\n",
    "\n",
    "# Header for filters\n",
    "FILTER_HEADER = html.H3(\"Search for Structural Variants\",\n",
    "                        id = 'filter-header',\n",
    "                        style = {'textAlign' : 'center',\n",
    "                                 'color' : '#2F2E2D',\n",
    "                                 'height' : '70 px'})\n",
    "\n",
    "FILTER_HEADER_DIV = html.Div(id = 'filter-header-div',\n",
    "                             children = [FILTER_HEADER],\n",
    "                             className = 'section-headers')\n",
    "\n",
    "\n",
    "\n",
    "# Header for plot\n",
    "PLOT_HEADER = html.Div([html.H3([\"SVs and GWAS Hits Linkage Disequilibirum Plot\",\n",
    "                                 question_icon],\n",
    "                                 id = 'plot-header',\n",
    "                                 style = {'textAlign' : 'center',\n",
    "                                          'color' : '#2F2E2D',\n",
    "                                          'height' : '70 px'}),\n",
    "                        dbc.Tooltip(dcc.Markdown(\"Interactive plot visualizing linkage disequilibrium (LD) between the selected SV and GWAS-associated SNPs from GWAS Catalog.\",\n",
    "                                                 style = {'white-space':'pre-wrap'}),\n",
    "                                    target = 'plot-header',\n",
    "                                    placement = \"auto\",\n",
    "                                    delay = {\"show\" : 500,\n",
    "                                             \"hide\" : 50},\n",
    "                                    id = 'plot-header-tooltip')])\n",
    "\n",
    "PLOT_HEADER_DIV = html.Div(id = 'plot-header-div',\n",
    "                           children = [PLOT_HEADER],\n",
    "                           className = 'section-headers')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 534,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#### CALLBACKS ####\n",
    "\n",
    "# Incorporate together to create interactive webpage\n",
    "app.layout = html.Div([NAVBAR,\n",
    "                       LOGO_DIV,\n",
    "                       HEADER_DIV,\n",
    "                       FILTER_HEADER_DIV,\n",
    "                       ALL_FILTER_DIV,\n",
    "                       SV_TABLE_DIV,\n",
    "                       PLOT_HEADER_DIV,\n",
    "                       SCATPLOT_DIV,\n",
    "                       SNP_TABLE])\n",
    "\n",
    "\n",
    "# Contact Us modal\n",
    "@app.callback(\n",
    "    Output('modal-contact-us', 'is_open'),\n",
    "    [Input('button-modal-contact-us', 'n_clicks')],\n",
    "    [State('modal-contact-us', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_contact, is_open):\n",
    "    if modal_contact:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Citation modal\n",
    "@app.callback(\n",
    "    Output('modal-citation', 'is_open'),\n",
    "    [Input('button-modal-citation', 'n_clicks')],\n",
    "    [State('modal-citation', 'is_open')],\n",
    ")\n",
    "def toggle_modal(modal_citation, is_open):\n",
    "    if modal_citation:\n",
    "        return not is_open\n",
    "    return is_open\n",
    "\n",
    "\n",
    "# Disclaimer cutton opens collapsed text\n",
    "@app.callback(\n",
    "    Output('disclaimer-collapse', 'is_open'),\n",
    "    [Input('disclaimer-button', 'n_clicks')],\n",
    "    [State('disclaimer-collapse', 'is_open')]\n",
    ")\n",
    "def dislcaimer_collapse(clicks, is_open):\n",
    "    if clicks:\n",
    "        return not is_open\n",
    "    return is_open\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Chromosome dropdown changes the genes and phenotypes displayed\n",
    "@app.callback(\n",
    "    Output('gene-dropdown', 'options'),\n",
    "    Output('phenotype-dropdown', 'options'),\n",
    "    Input('chromosome-dropdown', 'value')\n",
    ")\n",
    "def update_gene_dropdown(chrom):\n",
    "\n",
    "    if chrom is None:\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    elif chrom == \"Any\":\n",
    "        new_genes = genes\n",
    "        new_pheno = phenotypes\n",
    "    else:\n",
    "        chromosome = f\"chr{chrom}\"\n",
    "        new_genes = df_gene.filter(pl.col(\"chr\") == chromosome).select(\"gene\").unique().to_series().to_list()\n",
    "        new_genes = sorted(new_genes)\n",
    "        new_pheno = df_full_join.filter(pl.col(\"Chromosome\") == chromosome).select(\"Phenotype\").unique().to_series().to_list()\n",
    "        new_pheno = sorted(map(str, new_pheno))\n",
    "\n",
    "    return [{'label': str(i), 'value': str(i)} for i in new_genes], [{'label': str(i), 'value': str(i)} for i in new_pheno]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Filters alter the table displayed\n",
    "@app.callback(\n",
    "    Output('strvar-table-div', 'children'),\n",
    "    Input('chromosome-dropdown', 'value'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('gene-dropdown', 'value'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_table(chrom, pheno, gene, range_start, range_end):\n",
    "\n",
    "    chrom = chromosomes if chrom is None else chrom\n",
    "    pheno = phenotypes if pheno is None else pheno\n",
    "    gene = genes if gene is None else gene\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if chrom == \"Any\":\n",
    "        chrom = chromosomes\n",
    "    else:\n",
    "        chrom = chrom\n",
    "    \n",
    "    if pheno == \"Any\":\n",
    "        pheno = phenotypes\n",
    "    else:\n",
    "        pheno = pheno\n",
    "\n",
    "    new_df = make_table(chrom = chrom,\n",
    "                        range_start = range_start,\n",
    "                        range_end = range_end,\n",
    "                        pheno = pheno,\n",
    "                        gene = gene)\n",
    "\n",
    "    return new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('dtable-textbox', 'children'),\n",
    "    Input('strvar-table', 'derived_virtual_data'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('range-start-textbox', 'value'),\n",
    "    Input('range-end-textbox', 'value')\n",
    ")\n",
    "def update_sv_textbox(data, selected_rows, range_start, range_end):\n",
    "\n",
    "    range_start = 0 if range_start is None else range_start\n",
    "    range_end = 250000000 if range_end is None else range_end\n",
    "\n",
    "    if range_end <= range_start:\n",
    "        return html.P(\"Fix range values.\")\n",
    "\n",
    "    if data == []:\n",
    "        return html.P(\"No SVs in selected filters.\")\n",
    "    \n",
    "    elif selected_rows:\n",
    "        return html.P(\"Plot generated below.\")\n",
    "    \n",
    "    else:\n",
    "        return html.P(\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 538,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "@app.callback(\n",
    "    Output('anno-table-div', 'children'),\n",
    "    Input('strvar-table', 'selected_rows')\n",
    ")\n",
    "def update_anno_table(selected_rows):\n",
    "\n",
    "    if selected_rows:\n",
    "        sv_selected = tab_show[selected_rows[0], 0]\n",
    "        new_anno_table = make_annotation_table(sv = sv_selected)\n",
    "\n",
    "    else:\n",
    "        new_anno_table = make_annotation_table(sv = \"None\")\n",
    "\n",
    "    return new_anno_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 539,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Collapse SV annotation table\n",
    "@app.callback(\n",
    "    Output(\"sv-table-collapse\", \"is_open\"),\n",
    "    [Input(\"sv-button-collapse\", \"n_clicks\")],\n",
    "    [State(\"sv-table-collapse\", \"is_open\")],\n",
    ")\n",
    "def toggle_collapse(n, is_open):\n",
    "    if n:\n",
    "        return not is_open\n",
    "    return is_open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Only row selection on table changes plot, SV div textbox and button enabling\n",
    "@app.callback(\n",
    "    Output('scatter-plot', 'figure'),\n",
    "    Output('scatter-plot', 'clickData'),\n",
    "    Output('download-button', 'disabled'),\n",
    "    Output('other-pheno-switch', 'disabled'),\n",
    "    Input('strvar-table', 'selected_rows'),\n",
    "    Input('phenotype-dropdown', 'value'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value')\n",
    ")\n",
    "def update_plot(selected_rows, pheno, toggle, toggle_pheno):\n",
    "\n",
    "    pheno = \"None\" if pheno is None else pheno\n",
    "    toggle = False if toggle is None else toggle\n",
    "    toggle_pheno = False if toggle_pheno is None else toggle_pheno\n",
    "\n",
    "    if selected_rows:\n",
    "        sv_name = tab_show[selected_rows[0], 0]\n",
    "        button = False\n",
    "\n",
    "    else:\n",
    "        sv_name = \"None\"\n",
    "        button = True\n",
    "\n",
    "    if pheno == \"None\":\n",
    "        pheno_toggle_disable = True\n",
    "\n",
    "    else:\n",
    "        pheno_toggle_disable = False\n",
    "\n",
    "    new_plot = make_plot(sv = sv_name,\n",
    "                         pheno = pheno,\n",
    "                         toggle = toggle,\n",
    "                         toggle_pheno = toggle_pheno)\n",
    "\n",
    "    return new_plot, {}, button, pheno_toggle_disable\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 541,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Update SNP table based on data point clicked on plot\n",
    "@app.callback(\n",
    "        Output('snp-table-div', 'children'),\n",
    "        Output('modal-snp-table', 'is_open'),\n",
    "        Input('scatter-plot', 'clickData'),\n",
    "        Input('strvar-table', 'selected_rows'),\n",
    "        State('modal-snp-table', 'is_open'),\n",
    "    prevent_initial_call = True\n",
    ")\n",
    "def update_SNP_table(clickData, selected_rows, is_open):\n",
    "\n",
    "    if clickData:\n",
    "        if clickData == {}:\n",
    "            rsID = \"None\"\n",
    "            is_open = False\n",
    "        else:\n",
    "            rsID = clickData[\"points\"][0][\"customdata\"][1]\n",
    "            is_open = True\n",
    "\n",
    "    else:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    if not selected_rows:\n",
    "        rsID = \"None\"\n",
    "        is_open = False\n",
    "\n",
    "    new_snp_table = make_SNP_table(snp = rsID)\n",
    "\n",
    "    return new_snp_table, is_open\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Download SNP data seen in plot\n",
    "@app.callback(\n",
    "    Output('download-csv', 'data'),\n",
    "    Input('download-button', 'n_clicks'),\n",
    "    Input('scatter-plot', 'relayoutData'),\n",
    "    Input('toggle-switch', 'value'),\n",
    "    Input('other-pheno-switch', 'value'),\n",
    "    prevent_initial_call = True\n",
    ")\n",
    "def download_plot_data(n_clicks, relayoutData, toggle, toggle_pheno):\n",
    "    # Example from https://towardsdatascience.com/building-a-dashboard-in-plotly-dash-c748588e2920\n",
    "    global df_download\n",
    "\n",
    "    if toggle == False:\n",
    "        linkage = \"D'\"\n",
    "    elif toggle == True:\n",
    "        linkage = \"R2\"\n",
    "    else:\n",
    "        linkage = \"D'\"\n",
    "        \n",
    "    # Function to concat all dfs using polars\n",
    "    def align_and_reorder_columns(dfs, reference_df):\n",
    "    \n",
    "        reference_columns = reference_df.columns\n",
    "        reference_schema = reference_df.schema\n",
    "\n",
    "        aligned_dfs = []\n",
    "        for df in dfs:\n",
    "            missing_columns = [col for col in reference_columns if col not in df.columns]\n",
    "            df_with_all_columns = df.with_columns([pl.lit(None, dtype = reference_schema[col]).alias(col)\n",
    "                                                   for col in missing_columns])\n",
    "            \n",
    "            df_with_all_columns = df_with_all_columns.select(['Chromosome',\n",
    "                                                                'SV_Name',\n",
    "                                                                'SV_Start',\n",
    "                                                                'SV_End',\n",
    "                                                                'SV_Type',\n",
    "                                                                'SV_AF',\n",
    "                                                                'SNP_Name_GWAS',\n",
    "                                                                'SNP_Name_dbSNP',\n",
    "                                                                'SNP_Position',\n",
    "                                                                'Reference_Allele',\n",
    "                                                                'Alternate_Allele',\n",
    "                                                                'Sample_AF',\n",
    "                                                                'gnomAD_nfe_AF',\n",
    "                                                                'Risk_Allele',\n",
    "                                                                'Risk_Allele_Frequency',\n",
    "                                                                \"D'\",\n",
    "                                                                'R2',\n",
    "                                                                'Phenotype',\n",
    "                                                                'P-Value',\n",
    "                                                                'P-Value_log10',\n",
    "                                                                'Pubmed_ID',\n",
    "                                                                'Study',\n",
    "                                                                'Link'])\n",
    "            aligned_dfs.append(df_with_all_columns)\n",
    "\n",
    "        concatenated_df = pl.concat(aligned_dfs)\n",
    "\n",
    "        return concatenated_df\n",
    "\n",
    "        \n",
    "    # If button was triggered\n",
    "    if ctx.triggered[0]['prop_id'] == 'download-button.n_clicks':\n",
    "        # Default range (no zoom)\n",
    "        if relayoutData is None or \"xaxis.autorange\" in relayoutData:\n",
    "            df_download = tab\n",
    "            df_extra = extra_points\n",
    "\n",
    "            if toggle_pheno:\n",
    "                df_extra_pheno = extra_points_pheno\n",
    "                df_download = align_and_reorder_columns([df_download, df_extra, df_extra_pheno], \n",
    "                                                        reference_df = df_download)\n",
    "            else:\n",
    "                df_download = align_and_reorder_columns([df_download, df_extra], \n",
    "                                                        reference_df = df_download)\n",
    "\n",
    "        # Filter by x-axis and y-axis ranges\n",
    "        elif \"xaxis.range[0]\" in relayoutData:\n",
    "            x_min, x_max = relayoutData[\"xaxis.range[0]\"], relayoutData[\"xaxis.range[1]\"]\n",
    "\n",
    "            if \"yaxis.range[0]\" in relayoutData:\n",
    "                y_min, y_max = relayoutData[\"yaxis.range[0]\"], relayoutData[\"yaxis.range[1]\"]\n",
    "\n",
    "                if toggle_pheno:\n",
    "                    df_download = tab.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                             (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                             (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                             (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                    \n",
    "                    df_extra = extra_points.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                   (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                   (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                   (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                    \n",
    "                    df_extra_pheno = extra_points_pheno.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                               (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                               (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                               (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                else:\n",
    "                    \n",
    "                    if \"D'\" in extra_points.columns:\n",
    "                        df_download = tab.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                (pl.col(linkage) >= y_min) &\n",
    "                                                (pl.col(linkage) <= y_max))\n",
    "                        \n",
    "                        df_extra = extra_points.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                    (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                    (pl.col(linkage) >= y_min) &\n",
    "                                                    (pl.col(linkage) <= y_max))\n",
    "                    \n",
    "                    else:\n",
    "                        df_download = tab.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                        \n",
    "                        df_extra = extra_points.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                    (pl.col(\"SNP_Position\") <= x_max) &\n",
    "                                                    (pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                    (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                    \n",
    "            else:  # Only x-axis range\n",
    "                df_download = tab.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                         (pl.col(\"SNP_Position\") <= x_max))\n",
    "                \n",
    "                df_extra = extra_points.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                               (pl.col(\"SNP_Position\") <= x_max))\n",
    "                \n",
    "                if toggle_pheno:\n",
    "                    df_extra_pheno = extra_points_pheno.filter((pl.col(\"SNP_Position\") >= x_min) &\n",
    "                                                               (pl.col(\"SNP_Position\") <= x_max))\n",
    "                    \n",
    "                    df_download = align_and_reorder_columns([df_download, df_extra, df_extra_pheno], \n",
    "                                                             reference_df = df_download)\n",
    "                else:\n",
    "                    df_download = align_and_reorder_columns([df_download, df_extra], \n",
    "                                                             reference_df = df_download)\n",
    "        else:  # Only y-axis range\n",
    "            y_min, y_max = relayoutData[\"yaxis.range[0]\"], relayoutData[\"yaxis.range[1]\"]\n",
    "            \n",
    "            if toggle_pheno:\n",
    "                df_download = tab.filter((pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                         (pl.col(\"P-Value_log10\") <= y_max))\n",
    "            \n",
    "                df_extra = extra_points.filter((pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                               (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                \n",
    "                df_extra_pheno = extra_points_pheno.filter((pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                           (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                \n",
    "                df_download = align_and_reorder_columns([df_download, df_extra, df_extra_pheno], \n",
    "                                                         reference_df = df_download)\n",
    "            else:\n",
    "                \n",
    "                if \"D'\" in extra_points.columns:\n",
    "                    df_download = tab.filter((pl.col(linkage) >= y_min) &\n",
    "                                                (pl.col(linkage) <= y_max))\n",
    "                \n",
    "                    df_extra = extra_points.filter((pl.col(linkage) >= y_min) &\n",
    "                                                    (pl.col(linkage) <= y_max))\n",
    "                else:\n",
    "                    df_download = tab.filter((pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                             (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                \n",
    "                    df_extra = extra_points.filter((pl.col(\"P-Value_log10\") >= y_min) &\n",
    "                                                   (pl.col(\"P-Value_log10\") <= y_max))\n",
    "                \n",
    "                df_download = align_and_reorder_columns([df_download, df_extra], \n",
    "                                                         reference_df = df_download)\n",
    "            #raise PreventUpdate\n",
    "\n",
    "        return dcc.send_data_frame(df_download.to_pandas().to_csv, \n",
    "                                   \"gwas_svatalog_snp_data.csv\")\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    app.run_server(debug = True,\n",
    "                   host = \"127.0.0.1\",\n",
    "                   port = 4321)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
